
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../rnaseq_2/">
      
      
        <link rel="next" href="../../SCGLUE/scglue_1/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.1.2">
    
    
      
        <title>2-3. 通路富集分析(GSEA) - Bioinformatic Tutorial (Daxing Research)</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.7bf56d0a.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.a0c5b2b5.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#rna-seq-gsea" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bioinformatic Tutorial (Daxing Research)" class="md-header__button md-logo" aria-label="Bioinformatic Tutorial (Daxing Research)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bioinformatic Tutorial (Daxing Research)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              2-3. 通路富集分析(GSEA)
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
            </label>
          
        
      </form>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/Starlitnightly/bioinformatic_tutorial" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bioinformatic Tutorial (Daxing Research)" class="md-nav__button md-logo" aria-label="Bioinformatic Tutorial (Daxing Research)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Bioinformatic Tutorial (Daxing Research)
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Starlitnightly/bioinformatic_tutorial" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        写在前面
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          基础环境配置
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          基础环境配置
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_1/" class="md-nav__link">
        配置1. Python环境
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_2/" class="md-nav__link">
        配置2. Linux环境
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_3/" class="md-nav__link">
        配置3. R语言环境
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          RNA-seq上游分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          RNA-seq上游分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQUP/rnasequp_1/" class="md-nav__link">
        1-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQUP/rnasequp_2/" class="md-nav__link">
        1-2. 上游分析全教程
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          RNA-seq下游分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          RNA-seq下游分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../rnaseq_1/" class="md-nav__link">
        2-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../rnaseq_2/" class="md-nav__link">
        2-2. 差异表达基因(DEG)分析
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          2-3. 通路富集分析(GSEA)
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        2-3. 通路富集分析(GSEA)
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1" class="md-nav__link">
    1. 数据预处理
  </a>
  
    <nav class="md-nav" aria-label="1. 数据预处理">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11" class="md-nav__link">
    1.1 加载包
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12" class="md-nav__link">
    1.2 提取数据
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13" class="md-nav__link">
    1.3 数据排序
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-rnk" class="md-nav__link">
    1.4 rnk矩阵提取
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-gsea" class="md-nav__link">
    2. GSEA通路富集
  </a>
  
    <nav class="md-nav" aria-label="2. GSEA通路富集">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-index" class="md-nav__link">
    2.1 基因index转大写
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-gsea" class="md-nav__link">
    2.2 GSEA通路富集
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23" class="md-nav__link">
    2.3 单通路可视化
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24" class="md-nav__link">
    2.4 多通路可视化
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          单细胞样本对齐
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          单细胞样本对齐
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_1/" class="md-nav__link">
        3-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_2/" class="md-nav__link">
        3-2. 数据预处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_3/" class="md-nav__link">
        3-3. 模型准备
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_4/" class="md-nav__link">
        3-4. 模型训练
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_5/" class="md-nav__link">
        3-5. 对齐细胞
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          MOFA单细胞多组学因子分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          MOFA单细胞多组学因子分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_1/" class="md-nav__link">
        4-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_2/" class="md-nav__link">
        4-2. 单细胞样本对齐
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_3/" class="md-nav__link">
        4-3. scRNA-seq数据处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_4/" class="md-nav__link">
        4-4. scATAC-seq数据处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_5/" class="md-nav__link">
        4-5. MOFA模型构建
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_6/" class="md-nav__link">
        4-6. MOFA下游分析-R语言环节
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_7/" class="md-nav__link">
        4-7. MOFA下游分析-Python环节
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          TMT蛋白组学分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          TMT蛋白组学分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_1/" class="md-nav__link">
        5-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_2/" class="md-nav__link">
        5-2. TMT蛋白组学上游分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_3/" class="md-nav__link">
        5-3. TMT蛋白组学下游分析
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          多组学分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          多组学分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-1/" class="md-nav__link">
        6-1. 数据准备
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-2/" class="md-nav__link">
        6-2. GLUE多组学整合
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-3/" class="md-nav__link">
        6-3. 细胞亚群分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-4/" class="md-nav__link">
        6-4. 差异表达分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-5/" class="md-nav__link">
        6-5. 细胞通讯分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-6/" class="md-nav__link">
        6-6. 动态调控分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MultiOMIC/book-7/" class="md-nav__link">
        6-7. 转录因子网络
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
          Python可视化
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Python可视化
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_1/" class="md-nav__link">
        1-1. 火山图
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_2/" class="md-nav__link">
        1-2. 箱线图
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_3/" class="md-nav__link">
        1-3. 热图(聚类)
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1" class="md-nav__link">
    1. 数据预处理
  </a>
  
    <nav class="md-nav" aria-label="1. 数据预处理">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11" class="md-nav__link">
    1.1 加载包
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12" class="md-nav__link">
    1.2 提取数据
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13" class="md-nav__link">
    1.3 数据排序
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-rnk" class="md-nav__link">
    1.4 rnk矩阵提取
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-gsea" class="md-nav__link">
    2. GSEA通路富集
  </a>
  
    <nav class="md-nav" aria-label="2. GSEA通路富集">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#21-index" class="md-nav__link">
    2.1 基因index转大写
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#22-gsea" class="md-nav__link">
    2.2 GSEA通路富集
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#23" class="md-nav__link">
    2.3 单通路可视化
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#24" class="md-nav__link">
    2.4 多通路可视化
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
  
                  

  
  


<h1 id="rna-seq-gsea">RNA-seq分析： 通路富集分析(GSEA)</h1>
<p>在得到差异表达分析的结果后，我们往往还会对差异表达基因相关的功能感兴趣，在这里，我们介绍GSEA方法，这是一种比较好用的通路富集方法，可以根据基因表达的倍数情况，来分析基因相关的通路。</p>
<blockquote>
<p>一般来说，做GSEA分析其实是可以不用做提取差异表达基因分析的结果的，直接用DESeq2的结果就可以了，但是我这里更关注差异表达基因相关的功能，所以就用了差异表达基因来做GSEA分析</p>
</blockquote>
<h2 id="1">1. 数据预处理</h2>
<h3 id="11">1.1 加载包</h3>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="kn">import</span> <span class="nn">gseapy</span> <span class="k">as</span> <span class="nn">gp</span>
</code></pre></div>
<h3 id="12">1.2 提取数据</h3>
<p>由于我们需要同时考虑log2FoldChange与p-value，所以我们这里用一个新的公式来计算基因倍数</p>
<p>metric=-sign(log2FoldChange)/-lg10(pvalue)</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">gseada</span><span class="o">=</span><span class="n">data1</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">data1</span><span class="p">[</span><span class="s1">&#39;sig&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s1">&#39;normal&#39;</span><span class="p">]</span>
<span class="linenos" data-linenos="2 "></span><span class="c1">#倍数变化规则</span>
<span class="linenos" data-linenos="3 "></span><span class="n">gseada</span><span class="p">[</span><span class="s1">&#39;fcsign&#39;</span><span class="p">]</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">sign</span><span class="p">(</span><span class="n">gseada</span><span class="p">[</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">])</span>
<span class="linenos" data-linenos="4 "></span><span class="n">gseada</span><span class="p">[</span><span class="s1">&#39;logp&#39;</span><span class="p">]</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">gseada</span><span class="p">[</span><span class="s1">&#39;pvalue&#39;</span><span class="p">])</span>
<span class="linenos" data-linenos="5 "></span><span class="n">gseada</span><span class="p">[</span><span class="s1">&#39;metric&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">gseada</span><span class="p">[</span><span class="s1">&#39;logp&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">gseada</span><span class="p">[</span><span class="s1">&#39;fcsign&#39;</span><span class="p">]</span>
<span class="linenos" data-linenos="6 "></span><span class="n">gseada</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<p>得到结果如下</p>
<table>
<thead>
<tr>
<th align="right"></th>
<th align="right">baseMean</th>
<th align="right">log2FoldChange</th>
<th align="right">lfcSE</th>
<th align="right">stat</th>
<th align="right">pvalue</th>
<th align="right">padj</th>
<th align="right">sig</th>
<th align="right">log(padj)</th>
<th align="right">fcsign</th>
<th align="right">logp</th>
<th align="right">metric</th>
</tr>
</thead>
<tbody>
<tr>
<td align="right">ID3</td>
<td align="right">5442.355360</td>
<td align="right">-5.710849</td>
<td align="right">0.145082</td>
<td align="right">-39.362892</td>
<td align="right">0.000000e+00</td>
<td align="right">0.000000e+00</td>
<td align="right">down</td>
<td align="right">inf</td>
<td align="right">1.0</td>
<td align="right">inf</td>
<td align="right">inf</td>
</tr>
<tr>
<td align="right">TXNIP</td>
<td align="right">3147.115496</td>
<td align="right">4.041551</td>
<td align="right">0.103068</td>
<td align="right">39.212444</td>
<td align="right">0.000000e+00</td>
<td align="right">0.000000e+00</td>
<td align="right">up</td>
<td align="right">inf</td>
<td align="right">-1.0</td>
<td align="right">inf</td>
<td align="right">-inf</td>
</tr>
<tr>
<td align="right">MMP1</td>
<td align="right">4344.077177</td>
<td align="right">4.476829</td>
<td align="right">0.108971</td>
<td align="right">41.082873</td>
<td align="right">0.000000e+00</td>
<td align="right">0.000000e+00</td>
<td align="right">up</td>
<td align="right">inf</td>
<td align="right">-1.0</td>
<td align="right">inf</td>
<td align="right">-inf</td>
</tr>
<tr>
<td align="right">ID1</td>
<td align="right">5526.691818</td>
<td align="right">-5.280627</td>
<td align="right">0.112286</td>
<td align="right">-47.028177</td>
<td align="right">0.000000e+00</td>
<td align="right">0.000000e+00</td>
<td align="right">down</td>
<td align="right">inf</td>
<td align="right">1.0</td>
<td align="right">inf</td>
<td align="right">inf</td>
</tr>
<tr>
<td align="right">PTGIS</td>
<td align="right">4187.607799</td>
<td align="right">3.441545</td>
<td align="right">0.092530</td>
<td align="right">37.193624</td>
<td align="right">8.651991e-303</td>
<td align="right">3.367009e-299</td>
<td align="right">up</td>
<td align="right">298.472756</td>
<td align="right">-1.0</td>
<td align="right">302.062884</td>
<td align="right">-302.062884</td>
</tr>
</tbody>
</table>
<h3 id="13">1.3 数据排序</h3>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">gseada</span><span class="o">=</span><span class="n">gseada</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;metric&#39;</span><span class="p">],</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">gseada</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<table>
<thead>
<tr>
<th align="right"></th>
<th align="right">baseMean</th>
<th align="right">log2FoldChange</th>
<th align="right">lfcSE</th>
<th align="right">stat</th>
<th align="right">pvalue</th>
<th align="right">padj</th>
<th align="right">sig</th>
<th align="right">log(padj)</th>
<th align="right">fcsign</th>
<th align="right">logp</th>
<th align="right">metric</th>
</tr>
</thead>
<tbody>
<tr>
<td align="right">ID3</td>
<td align="right">5442.355360</td>
<td align="right">-5.710849</td>
<td align="right">0.145082</td>
<td align="right">-39.362892</td>
<td align="right">0.000000e+00</td>
<td align="right">0.000000e+00</td>
<td align="right">down</td>
<td align="right">inf</td>
<td align="right">1.0</td>
<td align="right">inf</td>
<td align="right">inf</td>
</tr>
<tr>
<td align="right">ID1</td>
<td align="right">5526.691818</td>
<td align="right">-5.280627</td>
<td align="right">0.112286</td>
<td align="right">-47.028177</td>
<td align="right">0.000000e+00</td>
<td align="right">0.000000e+00</td>
<td align="right">down</td>
<td align="right">inf</td>
<td align="right">1.0</td>
<td align="right">inf</td>
<td align="right">inf</td>
</tr>
<tr>
<td align="right">OXTR</td>
<td align="right">4105.199915</td>
<td align="right">-2.959533</td>
<td align="right">0.100562</td>
<td align="right">-29.429803</td>
<td align="right">2.283138e-190</td>
<td align="right">2.468072e-187</td>
<td align="right">down</td>
<td align="right">186.607642</td>
<td align="right">1.0</td>
<td align="right">189.641468</td>
<td align="right">189.641468</td>
</tr>
<tr>
<td align="right">SAMD11</td>
<td align="right">1574.582498</td>
<td align="right">-6.502040</td>
<td align="right">0.235542</td>
<td align="right">-27.604533</td>
<td align="right">9.816869e-168</td>
<td align="right">8.682574e-165</td>
<td align="right">down</td>
<td align="right">164.061352</td>
<td align="right">1.0</td>
<td align="right">167.008027</td>
<td align="right">167.008027</td>
</tr>
</tbody>
</table>
<h3 id="14-rnk">1.4 rnk矩阵提取</h3>
<p>我们发现数据中有inf，这对后续的分析有一定的影响，所以我们对inf进行赋值。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">rnk</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">rnk</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">gseada</span><span class="o">.</span><span class="n">index</span>
<span class="linenos" data-linenos=" 3 "></span><span class="n">rnk</span><span class="p">[</span><span class="s1">&#39;rnk&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">gseada</span><span class="p">[</span><span class="s1">&#39;metric&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="linenos" data-linenos=" 4 "></span><span class="n">k</span><span class="o">=</span><span class="mi">1</span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">total</span><span class="o">=</span><span class="mi">0</span>
<span class="linenos" data-linenos=" 6 "></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="p">)):</span>
<span class="linenos" data-linenos=" 7 "></span>    <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;rnk&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos=" 8 "></span>        <span class="n">total</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos=" 9 "></span><span class="c1">#200跟274根据你的数据进行更改，保证inf比你数据最大的大，-inf比数据最小的小就好</span>
<span class="linenos" data-linenos="10 "></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="p">)):</span>
<span class="linenos" data-linenos="11 "></span>    <span class="k">if</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;rnk&#39;</span><span class="p">]</span><span class="o">==</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos="12 "></span>        <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;rnk&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">200</span><span class="o">+</span><span class="p">(</span><span class="n">total</span><span class="o">-</span><span class="n">k</span><span class="p">)</span>
<span class="linenos" data-linenos="13 "></span>        <span class="n">k</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos="14 "></span>    <span class="k">elif</span> <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;rnk&#39;</span><span class="p">]</span><span class="o">==-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">:</span> 
<span class="linenos" data-linenos="15 "></span>        <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;rnk&#39;</span><span class="p">]</span><span class="o">=-</span><span class="p">(</span><span class="mi">274</span><span class="o">+</span><span class="n">k</span><span class="p">)</span>
<span class="linenos" data-linenos="16 "></span>        <span class="n">k</span><span class="o">+=</span><span class="mi">1</span>
<span class="linenos" data-linenos="17 "></span><span class="c1">#rnk=rnk.replace(np.inf,300)</span>
<span class="linenos" data-linenos="18 "></span><span class="c1">#rnk=rnk.replace(-np.inf,-300)</span>
<span class="linenos" data-linenos="19 "></span><span class="n">rnk</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<table>
<thead>
<tr>
<th align="right"></th>
<th align="right">gene_name</th>
<th align="right">rnk</th>
</tr>
</thead>
<tbody>
<tr>
<td align="right">0</td>
<td align="right">ID3</td>
<td align="right">201.000000</td>
</tr>
<tr>
<td align="right">1</td>
<td align="right">ID1</td>
<td align="right">200.000000</td>
</tr>
<tr>
<td align="right">2</td>
<td align="right">OXTR</td>
<td align="right">189.641468</td>
</tr>
<tr>
<td align="right">3</td>
<td align="right">SAMD11</td>
<td align="right">167.008027</td>
</tr>
<tr>
<td align="right">4</td>
<td align="right">ATOH8</td>
<td align="right">146.165895</td>
</tr>
</tbody>
</table>
<h2 id="2-gsea">2. GSEA通路富集</h2>
<h3 id="21-index">2.1 基因index转大写</h3>
<p>由于gseapy的包只能识别大写的基因名，所以我们需要先把所有的基因名转换成大写的。</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rnk</span><span class="p">)):</span>
<span class="linenos" data-linenos="2 "></span>    <span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">rnk</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
<span class="linenos" data-linenos="3 "></span><span class="n">rnk</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</code></pre></div>
<h3 id="22-gsea">2.2 GSEA通路富集</h3>
<p>这里简单介绍以下函数的参数</p>
<p>rnk：输入排序矩阵</p>
<p>gene_sets：需要富集到的通路数据集</p>
<p>processes：并行使用的线程数</p>
<p>permutation_num：检验的速度</p>
<p>outdir：结果输出目录</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1">#我们可以通过以下函数观察都有哪些数据集可被用来富集</span>
<span class="linenos" data-linenos="2 "></span><span class="n">names</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">get_library_name</span><span class="p">()</span> <span class="c1"># default: Human</span>
<span class="linenos" data-linenos="3 "></span><span class="n">names</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
</code></pre></div>
<p>在这里，我们选择的是KEGG_2019_Human数据集，因为NEGF是人源的细胞系</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">pre_res</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">prerank</span><span class="p">(</span><span class="n">rnk</span><span class="o">=</span><span class="n">rnk</span><span class="p">,</span> <span class="n">gene_sets</span><span class="o">=</span><span class="s1">&#39;KEGG_2019_Mouse&#39;</span><span class="p">,</span>
<span class="linenos" data-linenos="2 "></span>                     <span class="n">processes</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
<span class="linenos" data-linenos="3 "></span>                     <span class="n">permutation_num</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="c1"># reduce number to speed up testing</span>
<span class="linenos" data-linenos="4 "></span>                     <span class="n">outdir</span><span class="o">=</span><span class="s1">&#39;NEGF_gsea/prerank_report_kegg&#39;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;png&#39;</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
</code></pre></div>
<p>输出结果被保存到了NEGF_gsea目录下</p>
<p>我们可以通过index查看</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
</code></pre></div>
<table>
<thead>
<tr>
<th align="right"></th>
<th align="right">es</th>
<th align="right">nes</th>
<th align="right">pval</th>
<th align="right">fdr</th>
<th align="right">geneset_size</th>
<th align="right">matched_size</th>
<th align="right">genes</th>
<th align="right">ledge_genes</th>
</tr>
</thead>
<tbody>
<tr>
<td align="right">Term</td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
<td align="right"></td>
</tr>
<tr>
<td align="right">Chemokine signaling pathway</td>
<td align="right">-0.767073</td>
<td align="right">-1.330912</td>
<td align="right">0.045977</td>
<td align="right">0.091794</td>
<td align="right">190</td>
<td align="right">16</td>
<td align="right">CXCL11;CCL11;CX3CL1;HCK;CXCL10;CCL20;CCL8;VAV3...</td>
<td align="right">CXCL3;CXCL5;CXCL2;CXCL8;CXCL1;CXCL6</td>
</tr>
<tr>
<td align="right">Cytokine-cytokine receptor interaction</td>
<td align="right">-0.702018</td>
<td align="right">-1.351795</td>
<td align="right">0.055556</td>
<td align="right">0.094755</td>
<td align="right">294</td>
<td align="right">29</td>
<td align="right">IL18;IL1RN;CXCL11;CCL11;CX3CL1;CXCL10;IL18R1;C...</td>
<td align="right">LIF;CXCL3;CXCL5;CXCL2;ACKR4;IL6;IL1B;IL32;CXCL...</td>
</tr>
<tr>
<td align="right">IL-17 signaling pathway</td>
<td align="right">-0.801394</td>
<td align="right">-1.427337</td>
<td align="right">0.000000</td>
<td align="right">0.029611</td>
<td align="right">93</td>
<td align="right">18</td>
<td align="right">CCL11;CXCL10;CCL20;MMP9;CSF2;CSF3;CCL7;CXCL3;C...</td>
<td align="right">CXCL3;CXCL5;CXCL2;IL6;IL1B;PTGS2;TNFAIP3;CXCL8...</td>
</tr>
<tr>
<td align="right">NOD-like receptor signaling pathway</td>
<td align="right">-0.840998</td>
<td align="right">-1.518649</td>
<td align="right">0.000000</td>
<td align="right">0.000000</td>
<td align="right">178</td>
<td align="right">15</td>
<td align="right">IL18;IRF7;NOD2;GBP4;CXCL3;CXCL2;IL6;IL1B;BIRC3...</td>
<td align="right">CXCL3;CXCL2;IL6;IL1B;BIRC3;OAS1;TNFAIP3;OAS2;C...</td>
</tr>
</tbody>
</table>
<p>Term就代表了富集到的通路</p>
<h3 id="23">2.3 单通路可视化</h3>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="kn">from</span> <span class="nn">gseapy.plot</span> <span class="kn">import</span> <span class="n">gseaplot</span>
<span class="linenos" data-linenos="2 "></span>
<span class="linenos" data-linenos="3 "></span><span class="c1"># to save your figure, make sure that ofname is not None</span>
<span class="linenos" data-linenos="4 "></span><span class="n">gseaplot</span><span class="p">(</span><span class="n">rank_metric</span><span class="o">=</span><span class="n">pre_res</span><span class="o">.</span><span class="n">ranking</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="n">terms</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="o">**</span><span class="n">pre_res</span><span class="o">.</span><span class="n">results</span><span class="p">[</span><span class="n">terms</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
</code></pre></div>
<p>我们想观察每一个通路具体的富集情况，可以通过gseaplot绘制相关函数</p>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/NOD-like%20receptor%20signaling%20pathway.prerank.png" alt="NOD-like receptor signaling pathway.prerank" style="zoom:33%;" /></p>
<h3 id="24">2.4 多通路可视化</h3>
<p>我们还想观察富集到的所有通路结果，这里提供了一种气泡图的方式进行观察</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">sc_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;NEGF_gsea\prerank_report_kegg\gseapy.prerank.gene_sets.report.csv&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">sc_data</span><span class="o">=</span><span class="n">sc_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;pval&#39;</span><span class="p">]</span><span class="o">&lt;</span><span class="mf">0.1</span><span class="p">]</span>
<span class="linenos" data-linenos=" 3 "></span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;logp&#39;</span><span class="p">]</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;pval&#39;</span><span class="p">]</span><span class="o">+</span><span class="mf">0.001</span><span class="p">)</span>
<span class="linenos" data-linenos=" 4 "></span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;com&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;matched_size&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;geneset_size&#39;</span><span class="p">]</span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">sc_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="linenos" data-linenos=" 6 "></span>
<span class="linenos" data-linenos=" 7 "></span><span class="n">pp</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
<span class="linenos" data-linenos=" 8 "></span><span class="n">a</span><span class="o">=</span><span class="n">pp</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
<span class="linenos" data-linenos=" 9 "></span>
<span class="linenos" data-linenos="10 "></span><span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sc_data</span><span class="p">))</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;Term&#39;</span><span class="p">],</span><span class="n">s</span><span class="o">=</span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;com&#39;</span><span class="p">]</span><span class="o">*</span><span class="mi">500</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="linenos" data-linenos="11 "></span>            <span class="n">c</span><span class="o">=</span><span class="n">sc_data</span><span class="p">[</span><span class="s1">&#39;logp&#39;</span><span class="p">],</span>
<span class="linenos" data-linenos="12 "></span>            <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Greens&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="13 "></span><span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([</span><span class="mi">1</span><span class="p">],[</span><span class="s1">&#39;NHDF&#39;</span><span class="p">])</span>
<span class="linenos" data-linenos="14 "></span><span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="linenos" data-linenos="15 "></span><span class="c1">#plt.colorbar(a)</span>
<span class="linenos" data-linenos="16 "></span><span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos="17 "></span><span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="linenos" data-linenos="18 "></span><span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="mf">1.5</span><span class="p">)</span>
<span class="linenos" data-linenos="19 "></span><span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;NHDF_pathway.png&#39;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span><span class="n">bbox_inches</span> <span class="o">=</span> <span class="s1">&#39;tight&#39;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="untitled1" src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/untitled1.png" /></p>





                

  

              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fc8c2696.min.js"></script>
      
        <script src="../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>