
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>4-7. MOFA下游分析-Python环节 · GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    
    <link rel="prev" href="mofa_6.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Bioinformatics Tutorial
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../">
            
                <a href="../">
            
                    
                    目录
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../CONFIG/">
            
                <a href="../CONFIG/">
            
                    
                    配置：基础环境配置
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../CONFIG/config_1.html">
            
                <a href="../CONFIG/config_1.html">
            
                    
                    配置1：Python环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../CONFIG/config_2.html">
            
                <a href="../CONFIG/config_2.html">
            
                    
                    配置2：Linux环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../CONFIG/config_3.html">
            
                <a href="../CONFIG/config_3.html">
            
                    
                    配置3：R语言环境
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../RNASEQUP/">
            
                <a href="../RNASEQUP/">
            
                    
                    分析1：RNA-seq上游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../RNASEQUP/rnasequp_1.html">
            
                <a href="../RNASEQUP/rnasequp_1.html">
            
                    
                    1-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../RNASEQUP/rnasequp_2.html">
            
                <a href="../RNASEQUP/rnasequp_2.html">
            
                    
                    1-2. 上游分析全教程
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../RNASEQ/">
            
                <a href="../RNASEQ/">
            
                    
                    分析2：RNA-seq下游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="../RNASEQ/rnaseq_1.html">
            
                <a href="../RNASEQ/rnaseq_1.html">
            
                    
                    2-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="../RNASEQ/rnaseq_2.html">
            
                <a href="../RNASEQ/rnaseq_2.html">
            
                    
                    2-2. 差异表达基因(DEG)分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="../RNASEQ/rnaseq_3.html">
            
                <a href="../RNASEQ/rnaseq_3.html">
            
                    
                    2-3. 通路富集分析(GSEA)
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../SCGLUE/">
            
                <a href="../SCGLUE/">
            
                    
                    分析3：单细胞样本对齐
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" data-path="../SCGLUE/scglue_1.html">
            
                <a href="../SCGLUE/scglue_1.html">
            
                    
                    3-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.2" data-path="../SCGLUE/scglue_2.html">
            
                <a href="../SCGLUE/scglue_2.html">
            
                    
                    3-2. 数据预处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.3" data-path="../SCGLUE/scglue_3.html">
            
                <a href="../SCGLUE/scglue_3.html">
            
                    
                    3-3. 模型准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.4" data-path="../SCGLUE/scglue_4.html">
            
                <a href="../SCGLUE/scglue_4.html">
            
                    
                    3-4. 模型训练
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.5" data-path="../SCGLUE/scglue_5.html">
            
                <a href="../SCGLUE/scglue_5.html">
            
                    
                    3-5. 对齐细胞
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="./">
            
                <a href="./">
            
                    
                    分析4：MOFA单细胞多组学因子分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.1" data-path="mofa_1.html">
            
                <a href="mofa_1.html">
            
                    
                    4-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2" data-path="mofa_2.html">
            
                <a href="mofa_2.html">
            
                    
                    4-2. 单细胞样本对齐
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.3" data-path="mofa_3.html">
            
                <a href="mofa_3.html">
            
                    
                    4-3. scRNA-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.4" data-path="mofa_4.html">
            
                <a href="mofa_4.html">
            
                    
                    4-4. scATAC-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.5" data-path="mofa_5.html">
            
                <a href="mofa_5.html">
            
                    
                    4-5. MOFA模型构建
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.6" data-path="mofa_6.html">
            
                <a href="mofa_6.html">
            
                    
                    4-6. MOFA下游分析-R语言环节
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.7.7" data-path="mofa_7.html">
            
                <a href="mofa_7.html">
            
                    
                    4-7. MOFA下游分析-Python环节
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >4-7. MOFA下游分析-Python环节</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="mofa&#x4E0B;&#x6E38;&#x5206;&#x6790;-python&#x73AF;&#x8282;">MOFA&#x4E0B;&#x6E38;&#x5206;&#x6790;: Python&#x73AF;&#x8282;</h1>
<h2 id="1-&#x6570;&#x636E;&#x51C6;&#x5907;">1. &#x6570;&#x636E;&#x51C6;&#x5907;</h2>
<h3 id="11-&#x5BFC;&#x5165;&#x4F9D;&#x8D56;">1.1 &#x5BFC;&#x5165;&#x4F9D;&#x8D56;</h3>
<pre><code>import h5py
import pandas as pd
import numpy as np
</code></pre><h3 id="12-&#x5BFC;&#x5165;&#x6570;&#x636E;">1.2 &#x5BFC;&#x5165;&#x6570;&#x636E;</h3>
<pre><code>f_ad = h5py.File(&apos;mofa_factor_ad.hdf5&apos;,&apos;r&apos;)   #&#x6253;&#x5F00;h5&#x6587;&#x4EF6;
# &#x53EF;&#x4EE5;&#x67E5;&#x770B;&#x6240;&#x6709;&#x7684;&#x4E3B;&#x952E;
for key in f_ad.keys():
    print(f_ad[key].name)
</code></pre><blockquote>
<p>/data /expectations /features /groups /intercepts /model_options /samples /training_opts /training_stats /variance_explained /views</p>
</blockquote>
<h3 id="13-&#x5BFC;&#x51FA;&#x6743;&#x91CD;&#x6570;&#x636E;">1.3 &#x5BFC;&#x51FA;&#x6743;&#x91CD;&#x6570;&#x636E;</h3>
<pre><code>h5ls_out_ad=list(f_ad.keys())
if(&apos;views&apos; in h5ls_out_ad):
  view_names=f_ad[&apos;views&apos;][&apos;views&apos;][:]
  group_names=f_ad[&apos;groups&apos;][&apos;groups&apos;][:]
  feature_names=np.array([f_ad[&apos;features&apos;][i][:] for i in view_names])
  sample_names=np.array([f_ad[&apos;samples&apos;][i][:] for i in group_names])
</code></pre><h3 id="14-&#x6807;&#x51C6;&#x5316;&#x51FD;&#x6570;">1.4 &#x6807;&#x51C6;&#x5316;&#x51FD;&#x6570;</h3>
<pre><code>def normalization(data):
    _range = np.max(abs(data))
    return data / _range
</code></pre><h3 id="15-&#x83B7;&#x53D6;&#x6743;&#x91CD;&#x51FD;&#x6570;">1.5 &#x83B7;&#x53D6;&#x6743;&#x91CD;&#x51FD;&#x6570;</h3>
<pre><code>def get_weights(f,view,factor,scale=True):
  view_names=f[&apos;views&apos;][&apos;views&apos;][:]
  f_name=feature_names[np.where(view_names==str.encode(view))[0][0]]
  f_w=f[&apos;expectations&apos;][&apos;W&apos;][view][factor-1]
  if scale==True:
    f_w=normalization(f_w)
  res=pd.DataFrame()
  res[&apos;feature&apos;]=f_name
  res[&apos;weights&apos;]=f_w
  res[&apos;abs_weights&apos;]=abs(f_w)
  res[&apos;sig&apos;]=&apos;+&apos;
  res.loc[(res.weights&lt;0),&apos;sig&apos;] = &apos;-&apos;

  return res
</code></pre><h3 id="16-&#x83B7;&#x53D6;&#x7279;&#x5B9A;&#x56E0;&#x5B50;&#x7684;&#x57FA;&#x56E0;&#x6743;&#x91CD;">1.6 &#x83B7;&#x53D6;&#x7279;&#x5B9A;&#x56E0;&#x5B50;&#x7684;&#x57FA;&#x56E0;&#x6743;&#x91CD;</h3>
<pre><code>res1=get_weights(f_ad,&apos;rna&apos;,2)
xl=[]
for i in res1[&apos;feature&apos;].values:
  xl.append(bytes.decode(i).replace(&apos;Mutation&apos;,&apos;&apos;))

res1[&apos;feature&apos;]=xl
res1.head()
</code></pre><table>
<thead>
<tr>
<th style="text-align:left"></th>
<th style="text-align:right">feature</th>
<th style="text-align:right">weights</th>
<th style="text-align:right">abs_weights</th>
<th style="text-align:right">sig</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">0</td>
<td style="text-align:right">HES4</td>
<td style="text-align:right">0.158418</td>
<td style="text-align:right">0.158418</td>
<td style="text-align:right">+</td>
</tr>
<tr>
<td style="text-align:left">1</td>
<td style="text-align:right">ATAD3C</td>
<td style="text-align:right">-0.025514</td>
<td style="text-align:right">0.025514</td>
<td style="text-align:right">-</td>
</tr>
<tr>
<td style="text-align:left">2</td>
<td style="text-align:right">HES5</td>
<td style="text-align:right">0.244909</td>
<td style="text-align:right">0.244909</td>
<td style="text-align:right">+</td>
</tr>
<tr>
<td style="text-align:left">3</td>
<td style="text-align:right">LINC00982</td>
<td style="text-align:right">0.620071</td>
<td style="text-align:right">0.620071</td>
<td style="text-align:right">+</td>
</tr>
<tr>
<td style="text-align:left">4</td>
<td style="text-align:right">PRDM16</td>
<td style="text-align:right">0.815560</td>
<td style="text-align:right">0.815560</td>
<td style="text-align:right">+</td>
</tr>
</tbody>
</table>
<h3 id="17-&#x53EF;&#x89C6;&#x5316;&#x7279;&#x5B9A;&#x56E0;&#x5B50;">1.7 &#x53EF;&#x89C6;&#x5316;&#x7279;&#x5B9A;&#x56E0;&#x5B50;</h3>
<pre><code>import matplotlib.pyplot as plt


def plot_high_weight(res,n_feature=10):
  #&#x5B9A;&#x4E49;&#x5B57;&#x4F53;
  font1={
    &apos;size&apos;:15,
  }
  #&#x6392;&#x5E8F;
  res=res.sort_values(&apos;abs_weights&apos;,ascending=False)
  res_p=res.iloc[:n_feature].sort_values(&apos;weights&apos;,ascending=True)
  #&#x83B7;&#x53D6;xy
  x=res_p[&apos;weights&apos;].values
  y=range(len(res_p[&apos;weights&apos;]))

  #&#x5B9A;&#x4E49;&#x56FE;&#x50CF;&#x5927;&#x5C0F;
  pp=plt.figure(figsize=(4,6))
  ax=pp.add_subplot(1,1,1)
  #&#x5B9A;&#x4E49;+&#x7684;&#x8D77;&#x70B9;
  max_start=len(res_p)-len(res_p.loc[res_p[&apos;sig&apos;]==&apos;+&apos;])
  #&#x7ED8;&#x5236;&#x6563;&#x70B9;&#x56FE;
  ax.scatter(x[:max_start],y[:max_start],c=&quot;#00B0E0&quot;,alpha=0.5,linewidths=(np.zeros(len(res_p)-max_start)+2))
  ax.scatter(x[max_start:],y[max_start:],c=&quot;#E82089&quot;,alpha=0.5,linewidths=(np.zeros(len(res_p)-max_start)+2))

  #&#x5B9A;&#x4E49;&#x6587;&#x5B57;&#x4F4D;&#x7F6E;
  ti=len(res_p)-len(res_p.loc[res_p[&apos;sig&apos;]==&apos;+&apos;]) #+
  fi=0 #-
  ti_max=7 #&#x6362;&#x884C;&#x6807;&#x8BB0;
  fi_max=7
  for i in range(10):
    #&#x5982;&#x679C;&#x4E3A;-
    if res_p[&apos;sig&apos;].iloc[i]==&apos;-&apos;:
      #&#x8BA1;&#x7B97;&#x6587;&#x5B57;&#x957F;&#x5EA6;
      zl=len(res_p[&apos;feature&apos;].iloc[i])
      #&#x8BA1;&#x7B97;&#x8BE5;&#x884C;&#x5269;&#x4F59;&#x957F;&#x5EA6;
      fi_max-=zl

      ax.annotate(res_p[&apos;feature&apos;].iloc[i], xy=(x[i],y[i]), xytext=(-0.7+fi_max/7,fi),weight=&apos;heavy&apos;,
               #bbox=dict(boxstyle=&apos;round,pad=0.3&apos;, fc=&apos;white&apos;, alpha=0.5),
               arrowprops=dict(arrowstyle=&apos;-&gt;&apos;,color=&apos;grey&apos;),size=12)
      #&#x5982;&#x679C;&#x5269;&#x4F59;&#x957F;&#x5EA6;&#x4E0D;&#x8DB3;0&#xFF0C;&#x90A3;&#x4E48;&#x6362;&#x884C;
      if fi_max&lt;0:
        fi_max=10
        fi+=1
    else:
      zl=len(res_p[&apos;feature&apos;].iloc[i])
      ax.annotate(res_p[&apos;feature&apos;].iloc[i], xy=(x[i],y[i]), xytext=(1-ti_max/7,ti),weight=&apos;heavy&apos;,
               #bbox=dict(boxstyle=&apos;round,pad=0.3&apos;, fc=&apos;white&apos;, alpha=0.5),
               arrowprops=dict(arrowstyle=&apos;-&gt;&apos;,color=&apos;grey&apos;),size=12)
      ti_max-=zl
      if ti_max&lt;0:
        ti_max=10
        ti+=1

  plt.xlim(-1.5,1.5)
  plt.yticks(fontsize=15)
  plt.xticks(fontsize=15)
  #&#x8BBE;&#x7F6E;&#x56FE;&#x6CE8;
  plt.legend([&quot;-&quot;,&quot;+&quot;],loc=&apos;best&apos;,fontsize=12)
  #&#x8BBE;&#x7F6E;&#x6A2A;&#x7EB5;&#x6807;&#x9898;
  ax.set_ylabel(&apos;Rank&apos;,font1)                   
  ax.set_xlabel(&apos;Weights&apos;,font1)

plot_high_weight(res1,10)
</code></pre><p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_7.assets\&#x4E0B;&#x8F7D; (3" alt="&#x4E0B;&#x8F7D; (3)">.png)</p>
<h3 id="18-var&#x8DDF;cor&#x6570;&#x636E;&#x5BFC;&#x5165;">1.8 Var&#x8DDF;Cor&#x6570;&#x636E;&#x5BFC;&#x5165;</h3>
<pre><code>variance=pd.read_csv(&apos;/content/drive/MyDrive/GSE174367/ad/variance_explained.csv&apos;)
corr=pd.read_csv(&apos;/content/drive/MyDrive/GSE174367/ad/correlate_factors.csv&apos;)
variance.index=corr.index
</code></pre><h3 id="19-var&#x53EF;&#x89C6;&#x5316;">1.9 Var&#x53EF;&#x89C6;&#x5316;</h3>
<pre><code>

fig, ax = plt.subplots(figsize=(2,6))         # Sample figsize in inches
g=sns.heatmap(variance,
      cmap=palettable.colorbrewer.sequential.Blues_9.mpl_colors, 
      linewidths=.5,annot_kws={&quot;size&quot;: 15},ax=ax
      )
g.set_xticklabels(g.get_xmajorticklabels(), fontsize = 15,rotation=45, horizontalalignment=&apos;right&apos;,)
g.set_yticklabels(g.get_ymajorticklabels(), fontsize = 15)
cbar = g.collections[0].colorbar
# here set the labelsize by 20
cbar.ax.tick_params(labelsize=15)
</code></pre><p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_7.assets\&#x4E0B;&#x8F7D; (4" alt="&#x4E0B;&#x8F7D; (4)">.png)</p>
<h3 id="110-cor&#x53EF;&#x89C6;&#x5316;">1.10 Cor&#x53EF;&#x89C6;&#x5316;</h3>
<pre><code class="lang-python">fig, ax = plt.subplots(figsize=(<span class="hljs-number">3</span>,<span class="hljs-number">6</span>))         <span class="hljs-comment"># Sample figsize in inches</span>
g=sns.heatmap(corr,
      cmap=palettable.colorbrewer.sequential.Reds_9.mpl_colors, 
      linewidths=<span class="hljs-number">.5</span>,annot_kws={<span class="hljs-string">&quot;size&quot;</span>: <span class="hljs-number">15</span>},ax=ax,
      )
g.set_xticklabels(g.get_xmajorticklabels(), fontsize = <span class="hljs-number">15</span>,rotation=<span class="hljs-number">45</span>, horizontalalignment=<span class="hljs-string">&apos;right&apos;</span>,)
g.set_yticklabels(g.get_ymajorticklabels(), fontsize = <span class="hljs-number">15</span>)
cbar = g.collections[<span class="hljs-number">0</span>].colorbar
<span class="hljs-comment"># here set the labelsize by 20</span>
cbar.ax.tick_params(labelsize=<span class="hljs-number">15</span>)
</code></pre>
<p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_7.assets\&#x4E0B;&#x8F7D; (5" alt="&#x4E0B;&#x8F7D; (5)">.png)</p>
<h2 id="2-&#x5355;&#x7EC6;&#x80DE;&#x73AF;&#x8282;">2. &#x5355;&#x7EC6;&#x80DE;&#x73AF;&#x8282;</h2>
<h3 id="21-&#x5BFC;&#x5165;&#x6570;&#x636E;">2.1 &#x5BFC;&#x5165;&#x6570;&#x636E;</h3>
<pre><code>rna=anndata.read_h5ad(&quot;rna_pair_mofa.h5ad&quot;)
atac=anndata.read_h5ad(&quot;atac_pair_mofa.h5ad&quot;)
ret3= list(set(rna.obs.index).intersection(atac.obs.index))
</code></pre><h4 id="211-rna&#x6570;&#x636E;">2.1.1 rna&#x6570;&#x636E;</h4>
<pre><code>rna_sex=rna[ret3]
for i in list(set(rna_sex.obs[&apos;cell_type&apos;])):
  print(i,len(rna_sex.obs.loc[rna_sex.obs[&apos;cell_type&apos;]==i])/len(rna_sex.obs))
</code></pre><blockquote>
<p>PER.END 0.0039946737683089215 </p>
<p>EX 0.07789613848202397 </p>
<p>ODC 0.7200399467376831 </p>
<p>MG 0.046937416777629824 </p>
<p>INH 0.05858854860186418 </p>
<p>OPC 0.027796271637816245 </p>
<p>ASC 0.06474700399467377</p>
</blockquote>
<h4 id="212-atac&#x6570;&#x636E;">2.1.2 atac&#x6570;&#x636E;</h4>
<pre><code>atac_sex=atac[ret3]
for i in list(set(atac_sex.obs[&apos;cell_type&apos;])):
  print(i,len(atac_sex.obs.loc[atac_sex.obs[&apos;cell_type&apos;]==i])/len(atac_sex.obs))
</code></pre><blockquote>
<p>PER.END 0.003828229027962716 </p>
<p>EX 0.07440079893475367 </p>
<p>ODC 0.7187083888149135 </p>
<p>MG 0.051098535286284955 </p>
<p>INH 0.05858854860186418 </p>
<p>OPC 0.02796271637816245 </p>
<p>ASC 0.06541278295605858</p>
</blockquote>
<pre><code>ret_sex= list(set(rna_sex.obs.index).intersection(atac_sex.obs.index))
len(ret_sex)
</code></pre><h3 id="22-scrna-seq&#x4E0E;scatac-seq&#x7EC6;&#x80DE;&#x7C7B;&#x578B;&#x6BD4;&#x8F83;">2.2 scRNA-seq&#x4E0E;scATAC-seq&#x7EC6;&#x80DE;&#x7C7B;&#x578B;&#x6BD4;&#x8F83;</h3>
<pre><code>cell_type=list(set(atac_sex[ret_sex].obs[&apos;cell_type&apos;]))
test_df=pd.DataFrame(columns=cell_type,index=cell_type)
for i in cell_type:
  for j in cell_type:
    test_index=atac_sex[ret_sex].obs[atac_sex[ret_sex].obs[&apos;cell_type&apos;]==i].index
    test_df.loc[i,j]=len(rna_sex[test_index].obs[rna_sex[test_index].obs[&apos;cell_type&apos;]==j])/len(test_index)
test_df
</code></pre><table>
<thead>
<tr>
<th style="text-align:left"></th>
<th style="text-align:right">ODC</th>
<th style="text-align:right">PER.END</th>
<th style="text-align:right">EX</th>
<th style="text-align:right">INH</th>
<th style="text-align:right">ASC</th>
<th style="text-align:right">OPC</th>
<th style="text-align:right">MG</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">ODC</td>
<td style="text-align:right">0.994908</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0.00203666</td>
<td style="text-align:right">0.00152749</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0.00152749</td>
<td style="text-align:right">0</td>
</tr>
<tr>
<td style="text-align:left">PER.END</td>
<td style="text-align:right">0.0833333</td>
<td style="text-align:right">0.916667</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
</tr>
<tr>
<td style="text-align:left">EX</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0.00222222</td>
<td style="text-align:right">0.991111</td>
<td style="text-align:right">0.00666667</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
</tr>
<tr>
<td style="text-align:left">INH</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0.00510204</td>
<td style="text-align:right">0.994898</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
</tr>
<tr>
<td style="text-align:left">ASC</td>
<td style="text-align:right">0.0135922</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0.0291262</td>
<td style="text-align:right">0.0038835</td>
<td style="text-align:right">0.953398</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
</tr>
<tr>
<td style="text-align:left">OPC</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0.00813008</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0.99187</td>
<td style="text-align:right">0</td>
</tr>
<tr>
<td style="text-align:left">MG</td>
<td style="text-align:right">0.113514</td>
<td style="text-align:right">0</td>
<td style="text-align:right">0.0648649</td>
<td style="text-align:right">0.00540541</td>
<td style="text-align:right">0.027027</td>
<td style="text-align:right">0.0108108</td>
<td style="text-align:right">0.778378</td>
</tr>
</tbody>
</table>
<pre><code>import seaborn as sns
import matplotlib.pyplot as plt
fig, ax = plt.subplots(figsize=(4,4))         # Sample figsize in inches
new_blues=sns.color_palette(&quot;Reds&quot;, 1000)[0:700]
g=sns.heatmap(test_df.astype(float),square=True,cmap=new_blues,
       linewidths=.5,annot_kws={&quot;size&quot;: 15},ax=ax)
g.set_xticklabels(g.get_xmajorticklabels(), fontsize = 15,rotation=45, horizontalalignment=&apos;right&apos;,)
g.set_yticklabels(g.get_ymajorticklabels(), fontsize = 15,rotation=360)
cbar = g.collections[0].colorbar
# here set the labelsize by 20
cbar.ax.tick_params(labelsize=15)
plt.savefig(&quot;fig1_b_corr.png&quot;,dpi=300,bbox_inches = &apos;tight&apos;)
</code></pre><p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_7.assets\&#x4E0B;&#x8F7D; (6).png" alt="&#x4E0B;&#x8F7D; (6)" style="zoom:50%;"></p>
<h3 id="23-scrna-seq&#x7684;ad&#x4E0E;ctrl&#x7EC4;&#x805A;&#x7C7B;">2.3 scRNA-seq&#x7684;AD&#x4E0E;Ctrl&#x7EC4;&#x805A;&#x7C7B;</h3>
<h4 id="231-&#x5BFC;&#x5165;&#x6570;&#x636E;">2.3.1 &#x5BFC;&#x5165;&#x6570;&#x636E;</h4>
<pre><code>ad_meta=pd.read_csv(&apos;/content/drive/MyDrive/GSE174367/mofa_sex_ad_meta.csv&apos;)
ctrl_meta=pd.read_csv(&apos;/content/drive/MyDrive/GSE174367/mofa_sex_ctrl_meta.csv&apos;)

res_ad=rna_sex[ad_meta[&apos;sample&apos;]]
res_ctrl=rna_sex[ctrl_meta[&apos;sample&apos;]]
</code></pre><h4 id="232-scrna-seq&#x7684;factor&#x8BBE;&#x7F6E;">2.3.2 scRNA-seq&#x7684;factor&#x8BBE;&#x7F6E;</h4>
<pre><code>f_ad = h5py.File(&apos;mofa_factor_ad.hdf5&apos;,&apos;r&apos;)   #&#x6253;&#x5F00;h5&#x6587;&#x4EF6;
f_ctrl = h5py.File(&apos;mofa_factor_ctrl.hdf5&apos;,&apos;r&apos;)   #&#x6253;&#x5F00;h5&#x6587;&#x4EF6;

for i in range(f_ad[&apos;expectations&apos;][&apos;Z&apos;][&apos;group0&apos;].shape[0]):
  res_ad.obs[&apos;factor{0}&apos;.format(i+1)]=f_ad[&apos;expectations&apos;][&apos;Z&apos;][&apos;group0&apos;][i]
for i in range(f_ctrl[&apos;expectations&apos;][&apos;Z&apos;][&apos;group0&apos;].shape[0]):
  res_ctrl.obs[&apos;factor{0}&apos;.format(i+1)]=f_ctrl[&apos;expectations&apos;][&apos;Z&apos;][&apos;group0&apos;][i]
</code></pre><h4 id="233-&#x7ED8;&#x56FE;&#x8BBE;&#x7F6E;">2.3.3 &#x7ED8;&#x56FE;&#x8BBE;&#x7F6E;</h4>
<pre><code>sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)
sc.logging.print_header()
sc.settings.set_figure_params(dpi=80, facecolor=&apos;white&apos;,figsize=14)
scnmt=[&apos;#f2535d&apos;,&apos;#a2e8c4&apos;,&apos;#e5e24f&apos;,&apos;#6aafda&apos;,&apos;#c76fc7&apos;,&apos;#c4425d&apos;,&apos;#f48a3d&apos;]
</code></pre><h4 id="234-ad&#x7EC4;&#x7ED8;&#x56FE;">2.3.4 AD&#x7EC4;&#x7ED8;&#x56FE;</h4>
<pre><code>sc.pp.neighbors(res_ad, metric=&quot;cosine&quot;)
sc.tl.umap(res_ad,random_state=41822099)
sc.settings.set_figure_params(dpi=80, facecolor=&apos;white&apos;,figsize=[4,4])
sc.pl.umap(res_ad,color=[&quot;cell_type&quot;,&quot;factor1&quot;,&quot;factor2&quot;,&quot;factor4&quot;], wspace=0.2, palette=scnmt,
      cmap=&apos;coolwarm&apos;,vmin=-5,vmax=5,frameon=False,add_outline=True,outline_color=(&apos;white&apos;,&apos;white&apos;))
</code></pre><p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_7.assets\&#x4E0B;&#x8F7D; (7" alt="&#x4E0B;&#x8F7D; (7)">.png)</p>
<h4 id="235-ctrl&#x7EC4;&#x7ED8;&#x56FE;">2.3.5 Ctrl&#x7EC4;&#x7ED8;&#x56FE;</h4>
<pre><code>sc.pp.neighbors(res_ctrl, metric=&quot;cosine&quot;)
sc.tl.umap(res_ctrl,random_state=41822099)
sc.settings.set_figure_params(dpi=80, facecolor=&apos;white&apos;,figsize=[4,4])
sc.pl.umap(res_ctrl,color=[&quot;cell_type&quot;,&quot;factor1&quot;,&quot;factor2&quot;,&quot;factor4&quot;], wspace=0.2, palette=scnmt,
      cmap=&apos;coolwarm&apos;,vmin=-5,vmax=5,frameon=False,add_outline=True,outline_color=(&apos;white&apos;,&apos;white&apos;))
</code></pre><p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_7.assets\&#x4E0B;&#x8F7D; (8" alt="&#x4E0B;&#x8F7D; (8)">.png)</p>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="mofa_6.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page: 4-6. MOFA下游分析-R语言环节">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"4-7. MOFA下游分析-Python环节","level":"1.7.7","depth":2,"previous":{"title":"4-6. MOFA下游分析-R语言环节","level":"1.7.6","depth":2,"path":"MOFA/mofa_6.md","ref":"MOFA/mofa_6.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{},"plugins":["expandable-chapters"],"pluginsConfig":{"expandable-chapters":{},"highlight":{},"search":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"fontsettings":{"theme":"white","family":"sans","size":2},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"MOFA/mofa_7.md","mtime":"2021-12-07T08:48:33.000Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2021-12-07T08:49:01.522Z"},"basePath":"..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

