
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../TMT/tmt_3/">
      
      
        <link rel="next" href="../book-2/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.1.2">
    
    
      
        <title>6-1. 数据准备 - Bioinformatic Tutorial (Daxing Research)</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.7bf56d0a.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.a0c5b2b5.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Bioinformatic Tutorial (Daxing Research)" class="md-header__button md-logo" aria-label="Bioinformatic Tutorial (Daxing Research)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Bioinformatic Tutorial (Daxing Research)
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              6-1. 数据准备
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
            </label>
          
        
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
            </label>
          
        
      </form>
    
    
    
    
      <div class="md-header__source">
        <a href="https://github.com/Starlitnightly/bioinformatic_tutorial" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Bioinformatic Tutorial (Daxing Research)" class="md-nav__button md-logo" aria-label="Bioinformatic Tutorial (Daxing Research)" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Bioinformatic Tutorial (Daxing Research)
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Starlitnightly/bioinformatic_tutorial" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        写在前面
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          基础环境配置
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          基础环境配置
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_1/" class="md-nav__link">
        配置1. Python环境
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_2/" class="md-nav__link">
        配置2. Linux环境
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../CONFIG/config_3/" class="md-nav__link">
        配置3. R语言环境
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          RNA-seq上游分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          RNA-seq上游分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQUP/rnasequp_1/" class="md-nav__link">
        1-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQUP/rnasequp_2/" class="md-nav__link">
        1-2. 上游分析全教程
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          RNA-seq下游分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          RNA-seq下游分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_1/" class="md-nav__link">
        2-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_2/" class="md-nav__link">
        2-2. 差异表达基因(DEG)分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../RNASEQ/rnaseq_3/" class="md-nav__link">
        2-3. 通路富集分析(GSEA)
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          单细胞样本对齐
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          单细胞样本对齐
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_1/" class="md-nav__link">
        3-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_2/" class="md-nav__link">
        3-2. 数据预处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_3/" class="md-nav__link">
        3-3. 模型准备
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_4/" class="md-nav__link">
        3-4. 模型训练
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../SCGLUE/scglue_5/" class="md-nav__link">
        3-5. 对齐细胞
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          MOFA单细胞多组学因子分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          MOFA单细胞多组学因子分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_1/" class="md-nav__link">
        4-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_2/" class="md-nav__link">
        4-2. 单细胞样本对齐
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_3/" class="md-nav__link">
        4-3. scRNA-seq数据处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_4/" class="md-nav__link">
        4-4. scATAC-seq数据处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_5/" class="md-nav__link">
        4-5. MOFA模型构建
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_6/" class="md-nav__link">
        4-6. MOFA下游分析-R语言环节
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../MOFA/mofa_7/" class="md-nav__link">
        4-7. MOFA下游分析-Python环节
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
          TMT蛋白组学分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_7">
          <span class="md-nav__icon md-icon"></span>
          TMT蛋白组学分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_1/" class="md-nav__link">
        5-1. 环境配置
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_2/" class="md-nav__link">
        5-2. TMT蛋白组学上游分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../TMT/tmt_3/" class="md-nav__link">
        5-3. TMT蛋白组学下游分析
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" checked>
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
          多组学分析
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_8">
          <span class="md-nav__icon md-icon"></span>
          多组学分析
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          6-1. 数据准备
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        6-1. 数据准备
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1" class="md-nav__link">
    1. 数据预处理
  </a>
  
    <nav class="md-nav" aria-label="1. 数据预处理">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-rna" class="md-nav__link">
    1.1 RNA数据预处理
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-atac" class="md-nav__link">
    1.2 ATAC数据预处理
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-atacgene-activity" class="md-nav__link">
    1.3 ATAC数据生成Gene-Activity矩阵
  </a>
  
    <nav class="md-nav" aria-label="1.3 ATAC数据生成Gene-Activity矩阵">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-filtered_peak_bc_matrixh5" class="md-nav__link">
    方法1: 使用filtered_peak_bc_matrix.h5文件
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2atac_rawh5ad" class="md-nav__link">
    方法2：使用前面生成的atac_raw.h5ad文件
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-gene-activityh5ad" class="md-nav__link">
    1.4 Gene-Activity矩阵保存为h5ad
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-2/" class="md-nav__link">
        6-2. GLUE多组学整合
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-3/" class="md-nav__link">
        6-3. 细胞亚群分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-4/" class="md-nav__link">
        6-4. 差异表达分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-5/" class="md-nav__link">
        6-5. 细胞通讯分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-6/" class="md-nav__link">
        6-6. 动态调控分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../book-7/" class="md-nav__link">
        6-7. 转录因子网络
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_9" >
      
      
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
          Python可视化
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_9">
          <span class="md-nav__icon md-icon"></span>
          Python可视化
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_1/" class="md-nav__link">
        1-1. 火山图
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_2/" class="md-nav__link">
        1-2. 箱线图
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../PLOT/plot_3/" class="md-nav__link">
        1-3. 热图(聚类)
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_10" >
      
      
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
      
      
        <label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
          视网膜母细胞瘤分析教程
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_10_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_10">
          <span class="md-nav__icon md-icon"></span>
          视网膜母细胞瘤分析教程
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/rb_tutorial/" class="md-nav__link">
        Base
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_1_preprocess/" class="md-nav__link">
        教程1: 单细胞上游结果预处理
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_2_filtered_qc/" class="md-nav__link">
        教程2: 质量控制与过滤
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_3_glue/" class="md-nav__link">
        教程3: scRNA-seq与scATAC-seq整合
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_4_scbasset/" class="md-nav__link">
        教程4: Motif活性分数计算
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_5_analysis1/" class="md-nav__link">
        教程5: 综合分析1
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_6_analysis2/" class="md-nav__link">
        教程6: 综合分析1-Chromatin activity和Motif score评估
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_7_STER_cell/" class="md-nav__link">
        教程7: STER cell下游分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_8_kineic_photo/" class="md-nav__link">
        教程8: 转录动力学分析
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../rb_tutorial/Tutorial_9_cellphonedb/" class="md-nav__link">
        教程9: 细胞通讯分析
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1" class="md-nav__link">
    1. 数据预处理
  </a>
  
    <nav class="md-nav" aria-label="1. 数据预处理">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-rna" class="md-nav__link">
    1.1 RNA数据预处理
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-atac" class="md-nav__link">
    1.2 ATAC数据预处理
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-atacgene-activity" class="md-nav__link">
    1.3 ATAC数据生成Gene-Activity矩阵
  </a>
  
    <nav class="md-nav" aria-label="1.3 ATAC数据生成Gene-Activity矩阵">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-filtered_peak_bc_matrixh5" class="md-nav__link">
    方法1: 使用filtered_peak_bc_matrix.h5文件
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2atac_rawh5ad" class="md-nav__link">
    方法2：使用前面生成的atac_raw.h5ad文件
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#14-gene-activityh5ad" class="md-nav__link">
    1.4 Gene-Activity矩阵保存为h5ad
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
  
                  

  
  


  <h1>6-1. 数据准备</h1>

<h2 id="1">1. 数据预处理</h2>
<p>在本章中，我们将介绍如何对数据进行初步的处理，数据来源分别为<code>Cellranger</code>,<code>Cellranger-atac</code>,<code>velocyto</code>三个上游分析结果的数据，格式分别为：</p>
<ul>
<li>cellranger：<code>filtered_feature_bc_matrix.h5</code>或<code>filtered_feature_bc_matrix</code>文件夹</li>
<li>cellranger-atac: <code>filtered_peak_bc_matrix</code>文件夹下的<code>matrix.mtx</code>,<code>barcodes.tsv</code>,<code>peaks.bed</code>三个文件以及<code>filtered_peak_bc_matrix.h5</code>文件（如果有的话最好）</li>
<li>velocyto: <code>possorted_genome_bam.loom</code>文件</li>
</ul>
<h3 id="11-rna">1.1 RNA数据预处理</h3>
<p>我们将<code>filtered_feature_bc_matrix.h5</code>与<code>possorted_genome_bam.loom</code>文件放在同一个目录下，然后使用scvelo的教程读取这两个文件，然后保存成h5ad格式</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="linenos" data-linenos=" 2 "></span><span class="kn">import</span> <span class="nn">scvelo</span> <span class="k">as</span> <span class="nn">scv</span>
<span class="linenos" data-linenos=" 3 "></span>
<span class="linenos" data-linenos=" 4 "></span><span class="c1">#读取矩阵文件</span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">adata</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span><span class="s1">&#39;filtered_feature_bc_matrix.h5&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 6 "></span><span class="c1">#使得obs跟var名唯一</span>
<span class="linenos" data-linenos=" 7 "></span><span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
<span class="linenos" data-linenos=" 8 "></span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names_make_unique</span><span class="p">()</span>
<span class="linenos" data-linenos=" 9 "></span><span class="c1">#读取velocyto文件</span>
<span class="linenos" data-linenos="10 "></span><span class="n">ldata</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;possorted_genome_bam.loom&#39;</span><span class="p">,</span> <span class="n">cache</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="linenos" data-linenos="11 "></span><span class="c1">#合并两文件</span>
<span class="linenos" data-linenos="12 "></span><span class="n">adata</span> <span class="o">=</span> <span class="n">scv</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">ldata</span><span class="p">)</span>
<span class="linenos" data-linenos="13 "></span><span class="c1">#保存文件</span>
<span class="linenos" data-linenos="14 "></span><span class="n">adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s1">&#39;rna_raw.h5ad&#39;</span><span class="p">,</span><span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
</code></pre></div>
<h3 id="12-atac">1.2 ATAC数据预处理</h3>
<p>我们找到<code>filtered_peak_bc_matrix</code>文件夹下的<code>matrix.mtx</code>,<code>barcodes.tsv</code>,<code>peaks.bed</code>三个文件，使用episcanpy读取后保存</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="kn">import</span> <span class="nn">episcanpy</span>
<span class="linenos" data-linenos="2 "></span><span class="n">adata</span><span class="o">=</span><span class="n">episcanpy</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">read_ATAC_10x</span><span class="p">(</span><span class="s1">&#39;matrix.mtx&#39;</span><span class="p">,</span> \
<span class="linenos" data-linenos="3 "></span>                                <span class="n">cell_names</span><span class="o">=</span><span class="s1">&#39;barcodes.tsv&#39;</span><span class="p">,</span> \
<span class="linenos" data-linenos="4 "></span>                                <span class="n">var_names</span><span class="o">=</span><span class="s1">&#39;peaks.bed&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="5 "></span><span class="n">adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s1">&#39;atac_raw.h5ad&#39;</span><span class="p">,</span><span class="n">compression</span><span class="o">=</span><span class="s2">&quot;gzip&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="13-atacgene-activity">1.3 ATAC数据生成Gene-Activity矩阵</h3>
<p>这里提供两种不同的方法达到此目的，一个较简单，一个较繁琐。区别在于，简单的方法你不一定有文件，但繁琐的方法你一定能实现</p>
<h4 id="1-filtered_peak_bc_matrixh5">方法1: 使用<code>filtered_peak_bc_matrix.h5</code>文件</h4>
<p>然后再运行下面的代码生成一个<code>gene_activity_gene_score.h5</code>文件
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>MAESTRO<span class="w"> </span>scatac-genescore<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="2 "></span>--format<span class="w"> </span>h5<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="3 "></span>--peakcount<span class="w"> </span>filtered_peak_bc_matrix.h5<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="4 "></span>--genedistance<span class="w"> </span><span class="m">10000</span><span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="5 "></span>--species<span class="w"> </span>GRCh38<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="6 "></span>--model<span class="w"> </span>Enhanced<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="7 "></span>-d<span class="w"> </span>/data/result<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="8 "></span>--outprefix<span class="w"> </span>gene_activity
</code></pre></div></p>
<h4 id="2atac_rawh5ad">方法2：使用前面生成的<code>atac_raw.h5ad</code>文件</h4>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="linenos" data-linenos="2 "></span><span class="n">atac</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;../data/raw_data/atac.h5ad&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="n">atac</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;../data/raw_data/brca_atac.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</code></pre></div>
<p>然后再运行下面的代码生成一个<code>gene_activity_gene_score.h5</code>文件
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>MAESTRO<span class="w"> </span>scatac-genescore<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="2 "></span>--format<span class="w"> </span>plain<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="3 "></span>--peakcount<span class="w"> </span>brca_atac.tsv<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="4 "></span>--genedistance<span class="w"> </span><span class="m">10000</span><span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="5 "></span>--species<span class="w"> </span>GRCh38<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="6 "></span>--model<span class="w"> </span>Enhanced<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="7 "></span>-d<span class="w"> </span>/data/result<span class="w"> </span><span class="se">\</span>
<span class="linenos" data-linenos="8 "></span>--outprefix<span class="w"> </span>gene_activity
</code></pre></div></p>
<h3 id="14-gene-activityh5ad">1.4 Gene-Activity矩阵保存为h5ad</h3>
<p>这一步骤稍微有一些繁琐，不过代码照着运行就好了</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="kn">import</span> <span class="nn">os</span>
<span class="linenos" data-linenos=" 2 "></span><span class="kn">import</span> <span class="nn">collections</span>
<span class="linenos" data-linenos=" 3 "></span><span class="kn">import</span> <span class="nn">tables</span>
<span class="linenos" data-linenos=" 4 "></span><span class="kn">import</span> <span class="nn">h5py</span>
<span class="linenos" data-linenos=" 5 "></span><span class="kn">import</span> <span class="nn">scipy.io</span>
<span class="linenos" data-linenos=" 6 "></span><span class="kn">import</span> <span class="nn">csv</span>
<span class="linenos" data-linenos=" 7 "></span><span class="kn">import</span> <span class="nn">gzip</span>
<span class="linenos" data-linenos=" 8 "></span><span class="kn">import</span> <span class="nn">scipy.sparse</span> <span class="k">as</span> <span class="nn">sp_sparse</span>
<span class="linenos" data-linenos=" 9 "></span><span class="kn">import</span> <span class="nn">argparse</span> <span class="k">as</span> <span class="nn">ap</span>
<span class="linenos" data-linenos="10 "></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="linenos" data-linenos="11 "></span><span class="n">FeatureBCMatrix</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span><span class="s1">&#39;FeatureBCMatrix&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;ids&#39;</span><span class="p">,</span> <span class="s1">&#39;names&#39;</span><span class="p">,</span> <span class="s1">&#39;barcodes&#39;</span><span class="p">,</span> <span class="s1">&#39;matrix&#39;</span><span class="p">])</span>
<span class="linenos" data-linenos="12 "></span><span class="k">def</span> <span class="nf">read_10X_h5</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
<span class="linenos" data-linenos="13 "></span><span class="w">    </span><span class="sd">&quot;&quot;&quot;Read 10X HDF5 files, support both gene expression and peaks.&quot;&quot;&quot;</span>
<span class="linenos" data-linenos="14 "></span>    <span class="k">with</span> <span class="n">tables</span><span class="o">.</span><span class="n">open_file</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="linenos" data-linenos="15 "></span>        <span class="k">try</span><span class="p">:</span>
<span class="linenos" data-linenos="16 "></span>            <span class="n">group</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">get_node</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">root</span><span class="p">,</span> <span class="s1">&#39;matrix&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="17 "></span>        <span class="k">except</span> <span class="n">tables</span><span class="o">.</span><span class="n">NoSuchNodeError</span><span class="p">:</span>
<span class="linenos" data-linenos="18 "></span>            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Matrix group does not exist in this file.&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="19 "></span>            <span class="k">return</span> <span class="kc">None</span>
<span class="linenos" data-linenos="20 "></span>        <span class="n">feature_group</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="s1">&#39;features&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos="21 "></span>        <span class="n">ids</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">feature_group</span><span class="p">,</span> <span class="s1">&#39;id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="linenos" data-linenos="22 "></span>        <span class="n">names</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">feature_group</span><span class="p">,</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="linenos" data-linenos="23 "></span>        <span class="n">barcodes</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="s1">&#39;barcodes&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="linenos" data-linenos="24 "></span>        <span class="n">data</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="s1">&#39;data&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="linenos" data-linenos="25 "></span>        <span class="n">indices</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="s1">&#39;indices&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="linenos" data-linenos="26 "></span>        <span class="n">indptr</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="s1">&#39;indptr&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="linenos" data-linenos="27 "></span>        <span class="n">shape</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="s1">&#39;shape&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="linenos" data-linenos="28 "></span>        <span class="n">matrix</span> <span class="o">=</span> <span class="n">sp_sparse</span><span class="o">.</span><span class="n">csc_matrix</span><span class="p">((</span><span class="n">data</span><span class="p">,</span> <span class="n">indices</span><span class="p">,</span> <span class="n">indptr</span><span class="p">),</span> <span class="n">shape</span><span class="o">=</span><span class="n">shape</span><span class="p">)</span>
<span class="linenos" data-linenos="29 "></span>        <span class="k">return</span> <span class="n">FeatureBCMatrix</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span> <span class="n">names</span><span class="p">,</span> <span class="n">barcodes</span><span class="p">,</span> <span class="n">matrix</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="kn">import</span> <span class="nn">re</span>
<span class="linenos" data-linenos=" 2 "></span><span class="n">scatac_count</span><span class="o">=</span><span class="n">read_10X_h5</span><span class="p">(</span><span class="s1">&#39;gene_activity_gene_score.h5&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 3 "></span><span class="n">peakmatrix</span> <span class="o">=</span> <span class="n">scatac_count</span><span class="o">.</span><span class="n">matrix</span>
<span class="linenos" data-linenos=" 4 "></span><span class="n">features</span> <span class="o">=</span> <span class="n">scatac_count</span><span class="o">.</span><span class="n">names</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;\W&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span> <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">features</span><span class="p">]</span>
<span class="linenos" data-linenos=" 6 "></span><span class="n">features</span> <span class="o">=</span> <span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">encode</span><span class="p">()</span> <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">features</span><span class="p">]</span>
<span class="linenos" data-linenos=" 7 "></span><span class="n">barcodes</span> <span class="o">=</span> <span class="n">scatac_count</span><span class="o">.</span><span class="n">barcodes</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="linenos" data-linenos=" 8 "></span><span class="n">adata</span><span class="o">=</span><span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">peakmatrix</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">obs</span><span class="o">=</span><span class="n">barcodes</span><span class="p">,</span><span class="n">var</span><span class="o">=</span><span class="n">features</span><span class="p">)</span>
<span class="linenos" data-linenos=" 9 "></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<span class="linenos" data-linenos="10 "></span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s1">&#39;utf-8&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
<span class="linenos" data-linenos="11 "></span><span class="k">del</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="12 "></span><span class="k">del</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="linenos" data-linenos="13 "></span><span class="n">adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s1">&#39;gene_activity_gene_score.h5ad&#39;</span><span class="p">,</span><span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
</code></pre></div>
<p>到此，我们数据分析所需要的所有基本文件就准备完成了</p>





                

  

              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.indexes"], "search": "../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fc8c2696.min.js"></script>
      
        <script src="../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>