
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>6-4. 差异表达分析 · GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-back-to-top-button/plugin.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-cuav-chapters/cuav-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-theme-ivoryhza/ivory.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-theme-ivoryhza/footer.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="book-5.html" />
    
    
    <link rel="prev" href="book-3.html" />
    

    </head>
    <body>
        
<div class="book">
	<div class="header-inner">
		<!-- LOGO -->
		<div class="header-inner-left">
			<div class="logo"></div>
			<span class="title"></span>
		</div>

		<!-- Nav -->
		<div class="header-inner-right">
			<ul class="header-nav">
				<li>
					<a href="http://..." target="_blank">HOME</a>
				</li>
			</ul>
			<!-- Search -->
			
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

		</div>


	</div>

	<div class="book-summary">
		<div class="book-summary-title">目录</div>
		
		
		<nav role="navigation">
			


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Bioinformatics Tutorial
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../">
            
                <a href="../">
            
                    
                    目录
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../CONFIG/">
            
                <a href="../CONFIG/">
            
                    
                    配置：基础环境配置
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../CONFIG/config_1.html">
            
                <a href="../CONFIG/config_1.html">
            
                    
                    配置1：Python环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../CONFIG/config_2.html">
            
                <a href="../CONFIG/config_2.html">
            
                    
                    配置2：Linux环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../CONFIG/config_3.html">
            
                <a href="../CONFIG/config_3.html">
            
                    
                    配置3：R语言环境
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../RNASEQUP/">
            
                <a href="../RNASEQUP/">
            
                    
                    分析1：RNA-seq上游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../RNASEQUP/rnasequp_1.html">
            
                <a href="../RNASEQUP/rnasequp_1.html">
            
                    
                    1-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../RNASEQUP/rnasequp_2.html">
            
                <a href="../RNASEQUP/rnasequp_2.html">
            
                    
                    1-2. 上游分析全教程
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../RNASEQ/">
            
                <a href="../RNASEQ/">
            
                    
                    分析2：RNA-seq下游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="../RNASEQ/rnaseq_1.html">
            
                <a href="../RNASEQ/rnaseq_1.html">
            
                    
                    2-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="../RNASEQ/rnaseq_2.html">
            
                <a href="../RNASEQ/rnaseq_2.html">
            
                    
                    2-2. 差异表达基因(DEG)分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="../RNASEQ/rnaseq_3.html">
            
                <a href="../RNASEQ/rnaseq_3.html">
            
                    
                    2-3. 通路富集分析(GSEA)
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../SCGLUE/">
            
                <a href="../SCGLUE/">
            
                    
                    分析3：单细胞样本对齐
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" data-path="../SCGLUE/scglue_1.html">
            
                <a href="../SCGLUE/scglue_1.html">
            
                    
                    3-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.2" data-path="../SCGLUE/scglue_2.html">
            
                <a href="../SCGLUE/scglue_2.html">
            
                    
                    3-2. 数据预处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.3" data-path="../SCGLUE/scglue_3.html">
            
                <a href="../SCGLUE/scglue_3.html">
            
                    
                    3-3. 模型准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.4" data-path="../SCGLUE/scglue_4.html">
            
                <a href="../SCGLUE/scglue_4.html">
            
                    
                    3-4. 模型训练
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.5" data-path="../SCGLUE/scglue_5.html">
            
                <a href="../SCGLUE/scglue_5.html">
            
                    
                    3-5. 对齐细胞
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="../MOFA/">
            
                <a href="../MOFA/">
            
                    
                    分析4：MOFA单细胞多组学因子分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.1" data-path="../MOFA/mofa_1.html">
            
                <a href="../MOFA/mofa_1.html">
            
                    
                    4-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2" data-path="../MOFA/mofa_2.html">
            
                <a href="../MOFA/mofa_2.html">
            
                    
                    4-2. 单细胞样本对齐
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.3" data-path="../MOFA/mofa_3.html">
            
                <a href="../MOFA/mofa_3.html">
            
                    
                    4-3. scRNA-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.4" data-path="../MOFA/mofa_4.html">
            
                <a href="../MOFA/mofa_4.html">
            
                    
                    4-4. scATAC-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.5" data-path="../MOFA/mofa_5.html">
            
                <a href="../MOFA/mofa_5.html">
            
                    
                    4-5. MOFA模型构建
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.6" data-path="../MOFA/mofa_6.html">
            
                <a href="../MOFA/mofa_6.html">
            
                    
                    4-6. MOFA下游分析-R语言环节
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.7" data-path="../MOFA/mofa_7.html">
            
                <a href="../MOFA/mofa_7.html">
            
                    
                    4-7. MOFA下游分析-Python环节
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.8" data-path="../TMT/">
            
                <a href="../TMT/">
            
                    
                    分析5：TMT蛋白组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.8.1" data-path="../TMT/tmt_1.html">
            
                <a href="../TMT/tmt_1.html">
            
                    
                    5-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8.2" data-path="../TMT/tmt_2.html">
            
                <a href="../TMT/tmt_2.html">
            
                    
                    5-2. TMT蛋白组学上游分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8.3" data-path="../TMT/tmt_3.html">
            
                <a href="../TMT/tmt_3.html">
            
                    
                    5-3. TMT蛋白组学下游分析
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.9" data-path="./">
            
                <a href="./">
            
                    
                    分析6：多组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.9.1" data-path="book-1.html">
            
                <a href="book-1.html">
            
                    
                    6-1. 数据准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.2" data-path="book-2.html">
            
                <a href="book-2.html">
            
                    
                    6-2. GLUE多组学整合
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.3" data-path="book-3.html">
            
                <a href="book-3.html">
            
                    
                    6-3. 细胞亚群分析
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.9.4" data-path="book-4.html">
            
                <a href="book-4.html">
            
                    
                    6-4. 差异表达分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.5" data-path="book-5.html">
            
                <a href="book-5.html">
            
                    
                    6-5. 细胞通讯分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.6" data-path="book-6.html">
            
                <a href="book-6.html">
            
                    
                    6-6. 动态调控分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.7" data-path="book-7.html">
            
                <a href="book-7.html">
            
                    
                    6-7. 转录因子网络
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


		</nav>
		
		
	</div>

	<!-- Content nav -->
	<div class="book-anchor">
		<div class="book-anchor-title">大纲</div>
		<div class="book-anchor-body">

		</div>
	</div>

	<div class="book-body">
		
		<div class="body-inner">
			
			

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >6-4. 差异表达分析</a>
    </h1>
</div>




			<div class="page-wrapper" tabindex="-1" role="main">
				<div class="page-inner">
					
<div id="book-search-results">
    <div class="search-noresults">
    
					<section class="normal markdown-section">
						
						<h3 id="42-&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x5206;&#x6790;">4.2 &#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x5206;&#x6790;</h3>
<p>&#x6211;&#x4EEC;&#x9700;&#x8981;&#x5206;&#x6790;&#x7684;&#x662F;&#x9634;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x4E0E;&#x9633;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#xFF0C;&#x4E00;&#x4E2A;&#x5F88;&#x76F4;&#x89C2;&#x7684;&#x601D;&#x60F3;&#x5C31;&#x662F;&#xFF0C;&#x6211;&#x4EEC;&#x662F;&#x4E0D;&#x662F;&#x5E0C;&#x671B;&#x77E5;&#x9053;&#x9633;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x8DDF;&#x9634;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x6709;&#x4EC0;&#x4E48;&#x533A;&#x522B;&#xFF1F;&#x90A3;&#x4E48;&#x6211;&#x4EEC;&#x73B0;&#x5728;&#x5F97;&#x5230;&#x4E86;&#x6DCB;&#x5DF4;&#x7ED3;&#x7684;scRNA-seq&#x4E0E;scATAC-seq&#x6570;&#x636E;&#xFF0C;&#x6211;&#x4EEC;&#x662F;&#x4E0D;&#x662F;&#x53EF;&#x4EE5;&#x627E;&#x51FA;&#x4E24;&#x4E2A;&#x7EC4;&#x5B66;&#x5C42;&#x7684;&#x5171;&#x540C;&#x70B9;&#xFF1F;</p>
<p>&#x6211;&#x4EEC;&#x9996;&#x5148;&#x5BFC;&#x5165;&#x5305;</p>
<p>&#x6211;&#x4EEC;&#x9996;&#x5148;&#x9700;&#x8981;&#x5BFC;&#x5165;&#x5305;&#xFF0C;&#x4EE5;&#x53CA;&#x4E00;&#x4E9B;&#x76F8;&#x5173;&#x51FD;&#x6570;&#x7684;&#x51C6;&#x5907;</p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x5BFC;&#x5165;&#x5305;</span>
<span class="hljs-keyword">import</span> anndata
print(<span class="hljs-string">&apos;anndata(Ver): &apos;</span>,anndata.__version__)
<span class="hljs-keyword">import</span> scanpy <span class="hljs-keyword">as</span> sc
print(<span class="hljs-string">&apos;scanpy(Ver): &apos;</span>,sc.__version__)
<span class="hljs-keyword">import</span> scltnn <span class="hljs-comment">#&#x975E;&#x5FC5;&#x9700;</span>
print(<span class="hljs-string">&apos;scltnn(Ver): &apos;</span>,scltnn.__version__)
<span class="hljs-keyword">import</span> matplotlib.pyplot <span class="hljs-keyword">as</span> plt
<span class="hljs-keyword">import</span> matplotlib
print(<span class="hljs-string">&apos;matplotlib(Ver): &apos;</span>,matplotlib.__version__)
<span class="hljs-keyword">import</span> seaborn <span class="hljs-keyword">as</span> sns
print(<span class="hljs-string">&apos;seaborn(Ver): &apos;</span>,sns.__version__)
<span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np
print(<span class="hljs-string">&apos;numpy(Ver): &apos;</span>,np.__version__)
<span class="hljs-keyword">import</span> pandas <span class="hljs-keyword">as</span> pd
print(<span class="hljs-string">&apos;pandas(Ver): &apos;</span>,pd.__version__)
<span class="hljs-keyword">import</span> scvelo <span class="hljs-keyword">as</span> scv
print(<span class="hljs-string">&apos;scvelo(Ver): &apos;</span>,scv.__version__)
<span class="hljs-keyword">import</span> Pyomic
print(<span class="hljs-string">&apos;Pyomic(Ver): &apos;</span>,Pyomic.__version__)

<span class="hljs-comment">#&#x7ED8;&#x56FE;&#x53C2;&#x6570;&#x8BBE;&#x7F6E;</span>
sc.settings.verbosity = <span class="hljs-number">3</span>             <span class="hljs-comment"># verbosity: errors (0), warnings (1), info (2), hints (3)</span>
sc.settings.set_figure_params(dpi=<span class="hljs-number">80</span>, facecolor=<span class="hljs-string">&apos;white&apos;</span>)

sc_color=[<span class="hljs-string">&apos;#7CBB5F&apos;</span>,<span class="hljs-string">&apos;#368650&apos;</span>,<span class="hljs-string">&apos;#A499CC&apos;</span>,<span class="hljs-string">&apos;#5E4D9A&apos;</span>,<span class="hljs-string">&apos;#78C2ED&apos;</span>,<span class="hljs-string">&apos;#866017&apos;</span>,<span class="hljs-string">&apos;#9F987F&apos;</span>, <span class="hljs-string">&apos;#E0DFED&apos;</span>, <span class="hljs-string">&apos;#EF7B77&apos;</span>, <span class="hljs-string">&apos;#279AD7&apos;</span>,
 <span class="hljs-string">&apos;#F0EEF0&apos;</span>, <span class="hljs-string">&apos;#1F577B&apos;</span>, <span class="hljs-string">&apos;#A56BA7&apos;</span>, <span class="hljs-string">&apos;#E0A7C8&apos;</span>, <span class="hljs-string">&apos;#E069A6&apos;</span>, <span class="hljs-string">&apos;#941456&apos;</span>, <span class="hljs-string">&apos;#FCBC10&apos;</span>, <span class="hljs-string">&apos;#EAEFC5&apos;</span>, <span class="hljs-string">&apos;#01A0A7&apos;</span>, <span class="hljs-string">&apos;#75C8CC&apos;</span>, 
<span class="hljs-string">&apos;#F0D7BC&apos;</span>, <span class="hljs-string">&apos;#D5B26C&apos;</span>, <span class="hljs-string">&apos;#D5DA48&apos;</span>, <span class="hljs-string">&apos;#B6B812&apos;</span>,<span class="hljs-string">&apos;#9DC3C3&apos;</span>, <span class="hljs-string">&apos;#A89C92&apos;</span>, <span class="hljs-string">&apos;#FEE00C&apos;</span>,<span class="hljs-string">&apos;#FEF2A1&apos;</span>]
</code></pre>
<p>&#x6211;&#x4EEC;&#x5728;&#x524D;&#x9762;&#x5DF2;&#x7ECF;&#x5B8C;&#x6210;&#x4E86;GLUE&#x7684;&#x5206;&#x6790;&#x4E86;&#xFF0C;&#x610F;&#x5473;&#x7740;&#x6211;&#x4EEC;&#x62FF;&#x5230;&#x4E86;&#x914D;&#x5BF9;&#x597D;&#x7684;&#x7EC6;&#x80DE;&#x7ED3;&#x679C;&#xFF0C;&#x6240;&#x4EE5;&#x6211;&#x4EEC;&#x8FD9;&#x91CC;&#x5B9A;&#x4E49;&#x4E00;&#x4E2A;&#x65B0;&#x7684;&#x51FD;&#x6570;&#x7528;&#x4E8E;&#x914D;&#x5BF9;&#x7EC6;&#x80DE;</p>
<pre><code class="lang-python"><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">adata_rename</span><span class="hljs-params">(adata,pair,omic_name=<span class="hljs-string">&apos;omic_1&apos;</span>)</span>:</span>
    ret=list(set(pair[omic_name].values) &amp; set(adata.obs.index))
    new_pair=pair.loc[pair[omic_name].isin(ret)]
    adata=adata[new_pair[omic_name].values]
    adata.obs.index=new_pair.index
    <span class="hljs-keyword">return</span> adata
</code></pre>
<p>&#x8BE5;&#x51FD;&#x6570;&#x9700;&#x8981;&#x7684;pair&#x6587;&#x4EF6;&#x5982;&#x4E0B;</p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x5176;&#x4E2D;omic_1&#x4EE3;&#x8868;rna&#x5C42;&#xFF0C;omic_2&#x4EE3;&#x8868;atac&#x5C42;</span>
res_pair=pd.read_csv(<span class="hljs-string">&apos;../multi_omics/res_pair.csv&apos;</span>)
res_pair.set_index(res_pair.columns[<span class="hljs-number">0</span>],inplace=<span class="hljs-keyword">True</span>)
res_pair.head()
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right"></th>
<th style="text-align:right">omic_1</th>
<th style="text-align:right">omic_2</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">Unnamed: 0</td>
<td style="text-align:right"></td>
<td style="text-align:right"></td>
</tr>
<tr>
<td style="text-align:right">cell_0</td>
<td style="text-align:right">AACTCAGCATGATCCA</td>
<td style="text-align:right">neg-GCACGGTGTGCAAGAC-1</td>
</tr>
<tr>
<td style="text-align:right">cell_1</td>
<td style="text-align:right">AAACGGGTCGGCGCAT</td>
<td style="text-align:right">neg-TCTATTGAGAGGCAGG-1</td>
</tr>
<tr>
<td style="text-align:right">cell_2</td>
<td style="text-align:right">GACGCGTAGACAAGCC</td>
<td style="text-align:right">neg-GTGTCCTGTCATTGCA-1</td>
</tr>
<tr>
<td style="text-align:right">cell_3</td>
<td style="text-align:right">AAACCTGTCCCAACGG-1</td>
<td style="text-align:right">neg-AGTGTACCATGTGGGA-1</td>
</tr>
<tr>
<td style="text-align:right">cell_4</td>
<td style="text-align:right">CATATTCCAAACCCAT</td>
<td style="text-align:right">neg-TCAGCTCCAACGGGTA-1</td>
</tr>
</tbody>
</table>
<p>&#x6211;&#x4EEC;&#x8BFB;&#x53D6;&#x524D;&#x9762;&#x6CE8;&#x91CA;&#x597D;&#x7684;B&#x7EC6;&#x80DE;&#x7684;rna&#x6587;&#x4EF6;&#xFF0C;&#x5E76;&#x4F7F;&#x7528;&#x6211;&#x4EEC;&#x5B9A;&#x4E49;&#x7684;&#x8FD9;&#x4E2A;&#x51FD;&#x6570;&#x7B5B;&#x9009;&#x51FA;&#x914D;&#x5BF9;&#x597D;&#x7684;&#x7EC6;&#x80DE;</p>
<pre><code class="lang-python">rna=sc.read(<span class="hljs-string">&apos;B_cell_anno_new.h5ad&apos;</span>)
rna=adata_rename(rna,res_pair,<span class="hljs-string">&apos;omic_1&apos;</span>)
rna
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">View of AnnData object with n_obs &#xD7; n_vars = 4315 &#xD7; 1945
    obs: &apos;Type&apos;, &apos;initial_size_spliced&apos;, &apos;initial_size_unspliced&apos;, &apos;initial_size&apos;, &apos;domain&apos;, &apos;n_genes&apos;, &apos;leiden&apos;, &apos;n_counts&apos;, &apos;major_celltype&apos;, &apos;velocity_self_transition&apos;, &apos;root_cells&apos;, &apos;end_points&apos;, &apos;velocity_pseudotime&apos;, &apos;latent_time&apos;, &apos;B_celltype&apos;
    var: &apos;gene_ids&apos;, &apos;feature_types&apos;, &apos;genome&apos;, &apos;Accession&apos;, &apos;Chromosome&apos;, &apos;End&apos;, &apos;Start&apos;, &apos;Strand&apos;, &apos;n_cells&apos;, &apos;highly_variable&apos;, &apos;highly_variable_rank&apos;, &apos;means&apos;, &apos;variances&apos;, &apos;variances_norm&apos;, &apos;mean&apos;, &apos;std&apos;, &apos;chrom&apos;, &apos;chromStart&apos;, &apos;chromEnd&apos;, &apos;name&apos;, &apos;score&apos;, &apos;strand&apos;, &apos;thickStart&apos;, &apos;thickEnd&apos;, &apos;itemRgb&apos;, &apos;blockCount&apos;, &apos;blockSizes&apos;, &apos;blockStarts&apos;, &apos;gene_id&apos;, &apos;gene_type&apos;, &apos;hgnc_id&apos;, &apos;havana_gene&apos;, &apos;tag&apos;, &apos;dell&apos;, &apos;fit_r2&apos;, &apos;fit_alpha&apos;, &apos;fit_beta&apos;, &apos;fit_gamma&apos;, &apos;fit_t_&apos;, &apos;fit_scaling&apos;, &apos;fit_std_u&apos;, &apos;fit_std_s&apos;, &apos;fit_likelihood&apos;, &apos;fit_u0&apos;, &apos;fit_s0&apos;, &apos;fit_pval_steady&apos;, &apos;fit_steady_u&apos;, &apos;fit_steady_s&apos;, &apos;fit_variance&apos;, &apos;fit_alignment_scaling&apos;, &apos;velocity_genes&apos;
    uns: &apos;B_celltype_colors&apos;, &apos;B_celltype_sizes&apos;, &apos;Type_colors&apos;, &apos;cosg&apos;, &apos;dendrogram_B_celltype&apos;, &apos;dendrogram_leiden&apos;, &apos;hvg&apos;, &apos;leiden&apos;, &apos;leiden_colors&apos;, &apos;leiden_sizes&apos;, &apos;log1p&apos;, &apos;major_celltype_colors&apos;, &apos;neighbors&apos;, &apos;paga&apos;, &apos;pca&apos;, &apos;rank_genes_groups&apos;, &apos;recover_dynamics&apos;, &apos;umap&apos;, &apos;velocity_graph&apos;, &apos;velocity_graph_neg&apos;, &apos;velocity_params&apos;
    obsm: &apos;X_glue&apos;, &apos;X_pca&apos;, &apos;X_umap&apos;, &apos;velocity_umap&apos;
    varm: &apos;PCs&apos;, &apos;X_glue&apos;, &apos;loss&apos;
    layers: &apos;Ms&apos;, &apos;Mu&apos;, &apos;ambiguous&apos;, &apos;counts&apos;, &apos;fit_t&apos;, &apos;fit_tau&apos;, &apos;fit_tau_&apos;, &apos;matrix&apos;, &apos;spliced&apos;, &apos;unspliced&apos;, &apos;velocity&apos;, &apos;velocity_u&apos;
    obsp: &apos;connectivities&apos;, &apos;distances&apos;
</code></pre>
<p>&#x6211;&#x4EEC;&#x8BFB;&#x53D6;&#x6CA1;&#x6709;&#x6CE8;&#x91CA;&#x7684;atac_gs_anno&#x6587;&#x4EF6;(&#x8FD9;&#x662F;&#x57FA;&#x56E0;&#x6D3B;&#x6027;&#x77E9;&#x9635;)&#xFF0C;&#x5E76;&#x4F7F;&#x7528;&#x524D;&#x9762;&#x5B9A;&#x4E49;&#x7684;<code>adata_rename</code>&#x51FD;&#x6570;&#x4EE5;&#x53CA;rna&#x7684;obs&#x7279;&#x5F81;&#x7B5B;&#x9009;&#x51FA;&#x914D;&#x5BF9;&#x597D;&#x7684;&#x7EC6;&#x80DE;</p>
<pre><code class="lang-python">atac_act=sc.read(<span class="hljs-string">&apos;../cellanno/atac_gs_anno.h5ad&apos;</span>)
atac_act=adata_rename(atac_act,res_pair,<span class="hljs-string">&apos;omic_2&apos;</span>)
atac_act=atac_act[rna.obs.index]
atac_act=atac_act.raw.to_adata()
atac_act.obs=rna.obs
atac_act
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">AnnData object with n_obs &#xD7; n_vars = 4315 &#xD7; 28307
    obs: &apos;Type&apos;, &apos;initial_size_spliced&apos;, &apos;initial_size_unspliced&apos;, &apos;initial_size&apos;, &apos;domain&apos;, &apos;n_genes&apos;, &apos;leiden&apos;, &apos;n_counts&apos;, &apos;major_celltype&apos;, &apos;velocity_self_transition&apos;, &apos;root_cells&apos;, &apos;end_points&apos;, &apos;velocity_pseudotime&apos;, &apos;latent_time&apos;, &apos;B_celltype&apos;
    var: &apos;mt&apos;, &apos;n_cells_by_counts&apos;, &apos;mean_counts&apos;, &apos;pct_dropout_by_counts&apos;, &apos;total_counts&apos;
    uns: &apos;Type_colors&apos;, &apos;dendrogram_leiden&apos;, &apos;leiden&apos;, &apos;leiden_colors&apos;, &apos;leiden_sizes&apos;, &apos;log1p&apos;, &apos;major_celltype_colors&apos;, &apos;neighbors&apos;, &apos;paga&apos;, &apos;pca&apos;, &apos;rank_genes_groups&apos;, &apos;umap&apos;
    obsm: &apos;X_glue&apos;, &apos;X_lsi&apos;, &apos;X_pca&apos;, &apos;X_umap&apos;
    obsp: &apos;connectivities&apos;, &apos;distances&apos;
</code></pre>
<p>&#x968F;&#x540E;&#xFF0C;&#x6211;&#x4EEC;&#x5B9A;&#x4E49;&#x4E86;&#x4E00;&#x4E2A;&#x65B0;&#x7684;&#x51FD;&#x6570;&#xFF0C;&#x7528;&#x6765;&#x8BA1;&#x7B97;&#x9633;&#x6027;&#x8DDF;&#x9634;&#x6027;&#x7684;&#x6DCB;&#x5DF4;&#x7ED3;&#x4E4B;&#x95F4;&#x7684;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x60C5;&#x51B5;</p>
<pre><code class="lang-python"><span class="hljs-keyword">from</span> scipy  <span class="hljs-keyword">import</span> stats,sparse

<span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">scdeg</span><span class="hljs-params">(adata1,adata2)</span>:</span>
    mean1=adata1.X.mean(axis=<span class="hljs-number">0</span>)
    mean2=adata2.X.mean(axis=<span class="hljs-number">0</span>)

    <span class="hljs-keyword">if</span> sparse.issparse(adata1.X):
        fold=np.array(adata1.X.mean(axis=<span class="hljs-number">0</span>)-adata2.X.mean(axis=<span class="hljs-number">0</span>))[<span class="hljs-number">0</span>]
        <span class="hljs-comment">#fold=fold[0]</span>
        <span class="hljs-comment">#log2fold=np.log2(fold)</span>
        ttest = stats.ttest_ind(adata1[:,:].X.toarray(),adata2[:,:].X.toarray())
    <span class="hljs-keyword">else</span>:
        fold=adata1.X.mean(axis=<span class="hljs-number">0</span>)-adata2.X.mean(axis=<span class="hljs-number">0</span>)
        <span class="hljs-comment">#log2fold=np.log2(fold)</span>
        ttest = stats.ttest_ind(adata1[:,:].X,adata2[:,:].X)
    pvalue=ttest.pvalue

    res=pd.DataFrame(index=adata1.var.index)
    res[<span class="hljs-string">&apos;log2FC&apos;</span>]=fold
    res[<span class="hljs-string">&apos;pvalue&apos;</span>]=pvalue
    <span class="hljs-keyword">return</span> res

<span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">scdeg_atac</span><span class="hljs-params">(adata1,adata2)</span>:</span>
    mean1=adata1.X.mean(axis=<span class="hljs-number">0</span>)
    mean2=adata2.X.mean(axis=<span class="hljs-number">0</span>)

    <span class="hljs-keyword">if</span> sparse.issparse(adata1.X):
        fold=np.array(adata1.X.mean(axis=<span class="hljs-number">0</span>)/adata2.X.mean(axis=<span class="hljs-number">0</span>))[<span class="hljs-number">0</span>]
        <span class="hljs-comment">#fold=fold[0]</span>
        log2fold=np.log2(fold+<span class="hljs-number">0.0000001</span>)
        ttest = stats.ttest_ind(adata1[:,:].X.toarray(),adata2[:,:].X.toarray())
    <span class="hljs-keyword">else</span>:
        fold=adata1.X.mean(axis=<span class="hljs-number">0</span>)/adata2.X.mean(axis=<span class="hljs-number">0</span>)
        log2fold=np.log2(fold+<span class="hljs-number">0.0000001</span>)
        ttest = stats.ttest_ind(adata1[:,:].X,adata2[:,:].X)
    pvalue=ttest.pvalue

    res=pd.DataFrame(index=adata1.var.index)
    res[<span class="hljs-string">&apos;log2FC&apos;</span>]=log2fold
    res[<span class="hljs-string">&apos;pvalue&apos;</span>]=pvalue
    <span class="hljs-keyword">return</span> res
</code></pre>
<p>&#x6211;&#x4EEC;&#x5BF9;rna&#x6570;&#x636E;&#x8FDB;&#x884C;&#x9884;&#x5904;&#x7406;&#xFF0C;&#x6CE8;&#x610F;&#xFF0C;&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x4E0D;&#x5BF9;atac&#x7684;&#x6D3B;&#x6027;&#x77E9;&#x9635;&#x6570;&#x636E;&#x8FDB;&#x884C;&#x9884;&#x5904;&#x7406;</p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x5F97;&#x5230;rna&#x8DDF;atac&#x5171;&#x6709;&#x7684;&#x57FA;&#x56E0;&#x96C6;</span>
ret_gene=list(set(rna.raw.var.index.tolist()) &amp; set(atac_act.var.index.tolist()))
len(ret_gene)
<span class="hljs-comment">#&#x7B5B;&#x9009;&#x5171;&#x6709;&#x57FA;&#x56E0;</span>
deg_rna=rna.raw.to_adata()[:,ret_gene]
deg_atac_act=atac_act[:,ret_gene]
deg_rna,deg_atac_act
<span class="hljs-comment">#rna&#x9884;&#x5904;&#x7406;</span>
sc.pp.normalize_total(deg_rna)
sc.pp.log1p(deg_rna)
sc.pp.scale(deg_rna)
</code></pre>
<p>&#x6211;&#x4EEC;&#x9996;&#x5148;&#x7814;&#x7A76;B&#x7EC6;&#x80DE;&#x603B;&#x4F53;&#x7684;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x60C5;&#x51B5;</p>
<pre><code class="lang-python">b_cell_type=<span class="hljs-string">&apos;&apos;</span>

deg_r=scdeg(deg_rna[(deg_rna.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>)&amp;(deg_rna.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>]!=b_cell_type)],
            deg_rna[(deg_rna.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Neg&apos;</span>)&amp;(deg_rna.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>]!=b_cell_type)])
deg_r.columns=[i+<span class="hljs-string">&apos;-rna&apos;</span> <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> deg_r.columns]

deg_a=scdeg_atac(deg_atac_act[(deg_atac_act.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>)&amp;(deg_atac_act.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>]!=b_cell_type)],
            deg_atac_act[(deg_atac_act.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Neg&apos;</span>)&amp;(deg_atac_act.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>]!=b_cell_type)])
deg_a.columns=[i+<span class="hljs-string">&apos;-atac&apos;</span> <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> deg_a.columns]
</code></pre>
<p>&#x6211;&#x4EEC;&#x5F97;&#x5230;&#x4E86;<code>deg_r</code>&#x4E0E;<code>deg_a</code>&#x4E24;&#x4E2A;DataFrame&#xFF0C;&#x5206;&#x522B;&#x5B58;&#x50A8;&#x4E86;&#x76F8;&#x5173;&#x7684;&#x5DEE;&#x5F02;&#x7EC6;&#x80DE;&#x7C7B;&#x578B;&#xFF0C;&#x6211;&#x4EEC;&#x63A5;&#x7740;&#x5408;&#x5E76;&#x8FD9;&#x4E24;&#x4E2A;DataFrame&#xFF0C;&#x7528;&#x4E8E;&#x540E;&#x7EED;&#x7684;&#x7ED8;&#x56FE;</p>
<pre><code class="lang-python">deg_b=pd.concat([deg_r,deg_a],join=<span class="hljs-string">&apos;outer&apos;</span>,axis=<span class="hljs-number">1</span>)
deg_b[<span class="hljs-string">&apos;sig&apos;</span>]=<span class="hljs-string">&apos;normal&apos;</span>
deg_b.loc[(deg_b[<span class="hljs-string">&apos;log2FC-rna&apos;</span>]&lt;<span class="hljs-number">-0.3</span>) &amp; (deg_b[<span class="hljs-string">&apos;log2FC-atac&apos;</span>]&lt;<span class="hljs-number">-0.01</span>) &amp; (deg_b[<span class="hljs-string">&apos;pvalue-rna&apos;</span>]&lt;<span class="hljs-number">0.05</span>) &amp; (deg_b[<span class="hljs-string">&apos;pvalue-atac&apos;</span>]&lt;<span class="hljs-number">0.05</span>),<span class="hljs-string">&apos;sig&apos;</span>]=<span class="hljs-string">&apos;down-closed&apos;</span>
deg_b.loc[(deg_b[<span class="hljs-string">&apos;log2FC-rna&apos;</span>]&gt;<span class="hljs-number">0.3</span>) &amp; (deg_b[<span class="hljs-string">&apos;log2FC-atac&apos;</span>]&gt;<span class="hljs-number">0.01</span>) &amp; (deg_b[<span class="hljs-string">&apos;pvalue-rna&apos;</span>]&lt;<span class="hljs-number">0.05</span>) &amp; (deg_b[<span class="hljs-string">&apos;pvalue-atac&apos;</span>]&lt;<span class="hljs-number">0.05</span>),<span class="hljs-string">&apos;sig&apos;</span>]=<span class="hljs-string">&apos;up-open&apos;</span>

deg_b.loc[(deg_b[<span class="hljs-string">&apos;log2FC-rna&apos;</span>]&lt;<span class="hljs-number">-0.3</span>) &amp; (deg_b[<span class="hljs-string">&apos;log2FC-atac&apos;</span>]&gt;<span class="hljs-number">0.01</span>) &amp; (deg_b[<span class="hljs-string">&apos;pvalue-rna&apos;</span>]&lt;<span class="hljs-number">0.05</span>) &amp; (deg_b[<span class="hljs-string">&apos;pvalue-atac&apos;</span>]&lt;<span class="hljs-number">0.05</span>),<span class="hljs-string">&apos;sig&apos;</span>]=<span class="hljs-string">&apos;down-open&apos;</span>
deg_b.loc[(deg_b[<span class="hljs-string">&apos;log2FC-rna&apos;</span>]&gt;<span class="hljs-number">0.3</span>) &amp; (deg_b[<span class="hljs-string">&apos;log2FC-atac&apos;</span>]&lt;<span class="hljs-number">-0.01</span>) &amp; (deg_b[<span class="hljs-string">&apos;pvalue-rna&apos;</span>]&lt;<span class="hljs-number">0.05</span>) &amp; (deg_b[<span class="hljs-string">&apos;pvalue-atac&apos;</span>]&lt;<span class="hljs-number">0.05</span>),<span class="hljs-string">&apos;sig&apos;</span>]=<span class="hljs-string">&apos;up-closed&apos;</span>

deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]!=<span class="hljs-string">&apos;normal&apos;</span>]
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right"></th>
<th style="text-align:right">log2FC-rna</th>
<th style="text-align:right">pvalue-rna</th>
<th style="text-align:right">log2FC-atac</th>
<th style="text-align:right">pvalue-atac</th>
<th style="text-align:right">sig</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">CD69</td>
<td style="text-align:right">-0.481565</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">-0.414049</td>
<td style="text-align:right">7.953599e-15</td>
<td style="text-align:right">down-closed</td>
</tr>
<tr>
<td style="text-align:right">DUSP1</td>
<td style="text-align:right">-0.351880</td>
<td style="text-align:right">8.813636e-31</td>
<td style="text-align:right">-0.334110</td>
<td style="text-align:right">4.529727e-15</td>
<td style="text-align:right">down-closed</td>
</tr>
<tr>
<td style="text-align:right">KLF6</td>
<td style="text-align:right">-0.412127</td>
<td style="text-align:right">8.021032e-42</td>
<td style="text-align:right">-0.394257</td>
<td style="text-align:right">1.385903e-16</td>
<td style="text-align:right">down-closed</td>
</tr>
<tr>
<td style="text-align:right">RGS1</td>
<td style="text-align:right">-0.342126</td>
<td style="text-align:right">3.633539e-29</td>
<td style="text-align:right">-0.183141</td>
<td style="text-align:right">3.073230e-02</td>
<td style="text-align:right">down-closed</td>
</tr>
<tr>
<td style="text-align:right">HSP90AA1</td>
<td style="text-align:right">-0.530347</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">-0.333989</td>
<td style="text-align:right">1.227807e-11</td>
<td style="text-align:right">down-closed</td>
</tr>
<tr>
<td style="text-align:right">NFKBIA</td>
<td style="text-align:right">-0.409938</td>
<td style="text-align:right">2.176217e-41</td>
<td style="text-align:right">-0.218099</td>
<td style="text-align:right">7.971724e-05</td>
<td style="text-align:right">down-closed</td>
</tr>
<tr>
<td style="text-align:right">JUNB</td>
<td style="text-align:right">-0.335438</td>
<td style="text-align:right">4.362629e-28</td>
<td style="text-align:right">-0.257169</td>
<td style="text-align:right">1.043540e-08</td>
<td style="text-align:right">down-closed</td>
</tr>
</tbody>
</table>
<p>&#x6211;&#x4EEC;&#x53EF;&#x4EE5;&#x53EF;&#x89C6;&#x5316;&#x4E00;&#x4E0B;&#x8FD9;&#x4E9B;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x57FA;&#x56E0;&#x5728;&#x4E0D;&#x540C;&#x6DCB;&#x5DF4;&#x7ED3;&#x4E2D;&#x7684;&#x5206;&#x5E03;&#x60C5;&#x51B5;</p>
<pre><code class="lang-python">type_color={
    <span class="hljs-string">&apos;Pos&apos;</span>:<span class="hljs-string">&apos;#9B7170&apos;</span>,
    <span class="hljs-string">&apos;Neg&apos;</span>:<span class="hljs-string">&apos;#C65A50&apos;</span>
}
<span class="hljs-comment">#&#x5728;rna&#x4E2D;</span>
sc.pl.violin(deg_rna[deg_rna.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>]!=b_cell_type], 
             deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]!=<span class="hljs-string">&apos;normal&apos;</span>].index.tolist(), groupby=<span class="hljs-string">&apos;Type&apos;</span>,size=<span class="hljs-number">3</span>,palette=[<span class="hljs-string">&apos;#C65A50&apos;</span>,<span class="hljs-string">&apos;#9B7170&apos;</span>])
</code></pre>
<p><img src="book-4.assets/image-20221019012328268-6120000.png" alt="image-20221019012328268" style="zoom:50%;"></p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x5728;gene-activity&#x4E2D;</span>
sc.pl.violin(deg_atac_act[deg_atac_act.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>]!=b_cell_type], 
             deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]!=<span class="hljs-string">&apos;normal&apos;</span>].index.tolist(), 
             groupby=<span class="hljs-string">&apos;Type&apos;</span>,size=<span class="hljs-number">3</span>,palette=[<span class="hljs-string">&apos;#C65A50&apos;</span>,<span class="hljs-string">&apos;#9B7170&apos;</span>],
            title=deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]!=<span class="hljs-string">&apos;normal&apos;</span>].index.tolist(),stripplot=<span class="hljs-keyword">True</span>)
</code></pre>
<p><img src="book-4.assets/image-20221019012404062-6120003.png" alt="image-20221019012404062" style="zoom:50%;"></p>
<p>&#x6211;&#x4EEC;&#x63A5;&#x4E0B;&#x6765;&#x7ED8;&#x5236;&#x672C;&#x5C0F;&#x8282;&#x6700;&#x5173;&#x952E;&#x7684;&#x56FE;&#xFF0C;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x56FE;</p>
<pre><code class="lang-python"><span class="hljs-keyword">from</span> adjustText <span class="hljs-keyword">import</span> adjust_text
plt.figure(figsize=(<span class="hljs-number">3</span>, <span class="hljs-number">3</span>))
plt.scatter(deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;normal&apos;</span>][<span class="hljs-string">&apos;log2FC-rna&apos;</span>],
            deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;normal&apos;</span>][<span class="hljs-string">&apos;log2FC-atac&apos;</span>],
            s=-np.log(deg_r.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;normal&apos;</span>][<span class="hljs-string">&apos;pvalue-rna&apos;</span>]),color=<span class="hljs-string">&apos;#4e6d89&apos;</span>,alpha=<span class="hljs-number">.2</span>,)

plt.scatter(deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;up-open&apos;</span>][<span class="hljs-string">&apos;log2FC-rna&apos;</span>],
            deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;up-open&apos;</span>][<span class="hljs-string">&apos;log2FC-atac&apos;</span>],
            s=-np.log(deg_r.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;up-open&apos;</span>][<span class="hljs-string">&apos;pvalue-rna&apos;</span>]),color=sc_color[<span class="hljs-number">15</span>],
           linewidths=<span class="hljs-number">1</span>,edgecolors=sc_color[<span class="hljs-number">14</span>])

plt.scatter(deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;up-closed&apos;</span>][<span class="hljs-string">&apos;log2FC-rna&apos;</span>],
            deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;up-closed&apos;</span>][<span class="hljs-string">&apos;log2FC-atac&apos;</span>],
            s=-np.log(deg_r.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;up-closed&apos;</span>][<span class="hljs-string">&apos;pvalue-rna&apos;</span>]),color=sc_color[<span class="hljs-number">14</span>],
           linewidths=<span class="hljs-number">1</span>,edgecolors=sc_color[<span class="hljs-number">15</span>])

plt.scatter(deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;down-open&apos;</span>][<span class="hljs-string">&apos;log2FC-rna&apos;</span>],
            deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;down-open&apos;</span>][<span class="hljs-string">&apos;log2FC-atac&apos;</span>],
            s=-np.log(deg_r.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;down-open&apos;</span>][<span class="hljs-string">&apos;pvalue-rna&apos;</span>]),color=<span class="hljs-string">&apos;#5ca8dc&apos;</span>,
           linewidths=<span class="hljs-number">1</span>,edgecolors=<span class="hljs-string">&apos;#4f61c7&apos;</span>)

plt.scatter(deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;down-closed&apos;</span>][<span class="hljs-string">&apos;log2FC-rna&apos;</span>],
            deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;down-closed&apos;</span>][<span class="hljs-string">&apos;log2FC-atac&apos;</span>],
            s=-np.log(deg_r.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]==<span class="hljs-string">&apos;down-closed&apos;</span>][<span class="hljs-string">&apos;pvalue-rna&apos;</span>]),color=sc_color[<span class="hljs-number">9</span>],
           linewidths=<span class="hljs-number">1</span>,edgecolors=sc_color[<span class="hljs-number">11</span>])

plot_index=deg_b.loc[deg_b[<span class="hljs-string">&apos;sig&apos;</span>]!=<span class="hljs-string">&apos;normal&apos;</span>].index
texts=[plt.text(deg_b.loc[i,<span class="hljs-string">&apos;log2FC-rna&apos;</span>],deg_b.loc[i,<span class="hljs-string">&apos;log2FC-atac&apos;</span>],i
        ,fontdict={<span class="hljs-string">&apos;size&apos;</span>:<span class="hljs-number">10</span>,<span class="hljs-string">&apos;weight&apos;</span>:<span class="hljs-string">&apos;bold&apos;</span>}) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> plot_index[:<span class="hljs-number">20</span>]]

adjust_text(texts,only_move={<span class="hljs-string">&apos;text&apos;</span>: <span class="hljs-string">&apos;xy&apos;</span>},arrowprops=dict(arrowstyle=<span class="hljs-string">&apos;-&gt;&apos;</span>, color=<span class="hljs-string">&apos;red&apos;</span>),)

ax = plt.gca()
<span class="hljs-comment"># &#x53BB;&#x6389;&#x4E0A;&#x3001;&#x53F3;&#x4E8C;&#x4FA7;&#x7684;&#x8FB9;&#x6846;&#x7EBF;</span>
ax.spines[<span class="hljs-string">&apos;top&apos;</span>].set_color(<span class="hljs-string">&apos;none&apos;</span>)
ax.spines[<span class="hljs-string">&apos;right&apos;</span>].set_color(<span class="hljs-string">&apos;none&apos;</span>)

<span class="hljs-comment"># &#x5C06;&#x5DE6;&#x4FA7;&#x7684;y&#x8F74;&#xFF0C;&#x79FB;&#x5230;x=0&#x7684;&#x4F4D;&#x7F6E;</span>
ax.spines[<span class="hljs-string">&apos;left&apos;</span>].set_position((<span class="hljs-string">&apos;data&apos;</span>, <span class="hljs-number">0</span>))
ax.spines[<span class="hljs-string">&apos;bottom&apos;</span>].set_position((<span class="hljs-string">&apos;data&apos;</span>, <span class="hljs-number">0</span>))
plt.tick_params(labelsize=<span class="hljs-number">10</span>)
plt.grid(<span class="hljs-keyword">None</span>)



print(<span class="hljs-string">&apos;log2FC-rna&apos;</span>,deg_b[<span class="hljs-string">&apos;log2FC-rna&apos;</span>].min(),deg_b[<span class="hljs-string">&apos;log2FC-rna&apos;</span>].max())
print(<span class="hljs-string">&apos;log2FC-atac&apos;</span>,deg_b[<span class="hljs-string">&apos;log2FC-atac&apos;</span>].min(),deg_b[<span class="hljs-string">&apos;log2FC-atac&apos;</span>].max())

plot_xmin=round(deg_b[<span class="hljs-string">&apos;log2FC-rna&apos;</span>].min(),<span class="hljs-number">2</span>)<span class="hljs-number">-0.1</span>
plot_xmax=round(deg_b[<span class="hljs-string">&apos;log2FC-rna&apos;</span>].max(),<span class="hljs-number">2</span>)+<span class="hljs-number">0.1</span>
plot_ymin=round(deg_b[<span class="hljs-string">&apos;log2FC-atac&apos;</span>].min(),<span class="hljs-number">2</span>)<span class="hljs-number">-0.1</span>
plot_ymax=round(deg_b[<span class="hljs-string">&apos;log2FC-atac&apos;</span>].max(),<span class="hljs-number">2</span>)+<span class="hljs-number">0.1</span>

plt.xlim(plot_xmin,plot_xmax)
plt.ylim(plot_ymin,plot_ymax)

plt.arrow(<span class="hljs-number">0</span>, plot_ymax, <span class="hljs-number">0</span>, <span class="hljs-number">0</span>, width=<span class="hljs-number">0.01</span>, color=<span class="hljs-string">&quot;k&quot;</span>, clip_on=<span class="hljs-keyword">False</span>, head_width=<span class="hljs-number">0.02</span>, head_length=<span class="hljs-number">0.05</span>)
plt.arrow(plot_xmax, <span class="hljs-number">0</span>, <span class="hljs-number">0.001</span>, <span class="hljs-number">0</span>, width=<span class="hljs-number">0.01</span>, color=<span class="hljs-string">&quot;k&quot;</span>, clip_on=<span class="hljs-keyword">False</span>, head_width=<span class="hljs-number">0.05</span>, head_length=<span class="hljs-number">0.02</span>)



plt.gca().add_patch(plt.Rectangle(xy=(<span class="hljs-number">0.25</span>,plot_ymin),
width=<span class="hljs-number">15</span>, color=<span class="hljs-string">&apos;red&apos;</span>,
height=<span class="hljs-number">10</span>,alpha=<span class="hljs-number">.2</span>,
fill=<span class="hljs-keyword">True</span>, linewidth=<span class="hljs-number">2</span>))

plt.gca().add_patch(plt.Rectangle(xy=(<span class="hljs-number">-0.25</span>,plot_ymin),
width=<span class="hljs-number">-15</span>, color=<span class="hljs-string">&apos;#177cb0&apos;</span>,
height=<span class="hljs-number">10</span>,alpha=<span class="hljs-number">.2</span>,
fill=<span class="hljs-keyword">True</span>, linewidth=<span class="hljs-number">2</span>))

ax.set_ylabel(<span class="hljs-string">&apos;log2FC-ATAC&apos;</span>, ha=<span class="hljs-string">&apos;left&apos;</span>, y=<span class="hljs-number">2</span>, rotation=<span class="hljs-number">90</span>, labelpad=<span class="hljs-number">75</span>)
ax.set_xlabel(<span class="hljs-string">&apos;log2FC-RNA&apos;</span>, ha=<span class="hljs-string">&apos;left&apos;</span>, y=<span class="hljs-number">0</span>, rotation=<span class="hljs-number">0</span>, labelpad=<span class="hljs-number">130</span>)
plt.text(<span class="hljs-number">0.25</span>,plot_ymin,<span class="hljs-string">&apos;&gt;0.25&apos;</span>,fontsize=<span class="hljs-number">12</span>,color=<span class="hljs-string">&apos;red&apos;</span>,fontweight=<span class="hljs-string">&apos;bold&apos;</span>)
plt.text(<span class="hljs-number">-0.25</span>,plot_ymin,<span class="hljs-string">&apos;&lt;-0.25&apos;</span>,fontsize=<span class="hljs-number">12</span>,color=<span class="hljs-string">&apos;#177cb0&apos;</span>,ha=<span class="hljs-string">&apos;right&apos;</span>,fontweight=<span class="hljs-string">&apos;bold&apos;</span>)

plt.xlabel(<span class="hljs-string">&apos;log2FC-RNA&apos;</span>,fontsize=<span class="hljs-number">12</span>,)
plt.ylabel(<span class="hljs-string">&apos;log2FC-ATAC&apos;</span>,fontsize=<span class="hljs-number">12</span>,)
plt.title(<span class="hljs-string">&apos;B cell&apos;</span>,fontsize=<span class="hljs-number">13</span>,fontweight=<span class="hljs-string">&apos;bold&apos;</span>)

plt.yticks(np.linspace(<span class="hljs-number">-1</span>, <span class="hljs-number">0.5</span>, <span class="hljs-number">1</span>))
plt.xticks(np.linspace(<span class="hljs-number">-1</span>, <span class="hljs-number">0.5</span>, <span class="hljs-number">1</span>))

plt.savefig(<span class="hljs-string">&quot;deg/deg_{}.png&quot;</span>.format(b_cell_type),dpi=<span class="hljs-number">300</span>,bbox_inches = <span class="hljs-string">&apos;tight&apos;</span>)
</code></pre>
<p><img src="book-4.assets/image-20221019012503861-6120007.png" alt="image-20221019012503861" style="zoom:50%;"></p>
<p>&#x6211;&#x4EEC;&#x53EF;&#x4EE5;&#x4F7F;&#x7528;&#x540C;&#x6837;&#x7684;&#x65B9;&#x6CD5;&#xFF0C;&#x7ED8;&#x5236;&#x51FA;&#x5176;&#x4ED6;5&#x79CD;&#x4E0D;&#x540C;&#x7684;B&#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;&#x7684;&#x57FA;&#x56E0;&#x5206;&#x5E03;&#x56FE; </p>
<footer class="page-footer"><span class="copyright">&#x7248;&#x6743;&#x6240;&#x6709; &#xA9; Starlitnightly</span><span class="footer-modification">&#x6700;&#x540E;&#x4FEE;&#x8BA2;&#x65F6;&#x95F4;&#xFF1A;
2022-10-17
</span></footer>
						
					</section>
					
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

				</div>
			</div>
			
		</div>

		
		
		<a href="book-3.html" title="上一页" class="navigation navigation-prev " aria-label="Previous page: 6-3. 细胞亚群分析">
			<i class="fa fa-angle-left"></i>
		</a>
		
		
		<a href="book-5.html" title="下一页" class="navigation navigation-next " aria-label="Next page: 6-5. 细胞通讯分析">
			<i class="fa fa-angle-right"></i>
		</a>
		
		
		
	</div>
	<script>
		var gitbook = gitbook || [];
		gitbook.push(function() {
			gitbook.page.hasChanged({"page":{"title":"6-4. 差异表达分析","level":"1.9.4","depth":2,"next":{"title":"6-5. 细胞通讯分析","level":"1.9.5","depth":2,"path":"MultiOMIC/book-5.md","ref":"MultiOMIC/book-5.md","articles":[]},"previous":{"title":"6-3. 细胞亚群分析","level":"1.9.3","depth":2,"path":"MultiOMIC/book-3.md","ref":"MultiOMIC/book-3.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{"themeIvoryhza":{"nav":[{"url":"http://...","target":"_blank","name":"HOME"}]}},"plugins":["expandable-chapters","-sharing","-fontsettings","back-to-top-button","copy-code-button","cuav-chapters","heading-anchors","theme-ivoryhza"],"pluginsConfig":{"search":{},"theme-ivoryhza":{"search-placeholder":"搜索","logo":"./logo.png","appleTouchIconPrecomposed152":"./logo.png","copyright":"版权所有 © Starlitnightly","modify_label":"最后修订时间：","modify_format":"2022-10-17"},"cuav-chapters":{"summaryMaxSize":20,"useLimitExpanded":false},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"heading-anchors":{},"highlight":{},"back-to-top-button":{},"copy-code-button":{},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false},"expandable-chapters":{}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"MultiOMIC/book-4.md","mtime":"2022-10-18T19:06:48.293Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2022-10-18T19:40:25.819Z"},"basePath":"..","book":{"language":""}});
		});
	</script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-back-to-top-button/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-copy-code-button/toggle.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-cuav-chapters/cuav-chapters.js"></script>
        
    
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/1.2.1/anchor.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-heading-anchors/anchor-style.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-theme-ivoryhza/ivory.js"></script>
        
    

    </body>
</html>

