
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>4-4. scATAC-seq数据处理 · GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="mofa_5.html" />
    
    
    <link rel="prev" href="mofa_3.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Bioinformatics Tutorial
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../">
            
                <a href="../">
            
                    
                    目录
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../CONFIG/">
            
                <a href="../CONFIG/">
            
                    
                    配置：基础环境配置
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../CONFIG/config_1.html">
            
                <a href="../CONFIG/config_1.html">
            
                    
                    配置1：Python环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../CONFIG/config_2.html">
            
                <a href="../CONFIG/config_2.html">
            
                    
                    配置2：Linux环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../CONFIG/config_3.html">
            
                <a href="../CONFIG/config_3.html">
            
                    
                    配置3：R语言环境
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../RNASEQUP/">
            
                <a href="../RNASEQUP/">
            
                    
                    分析1：RNA-seq上游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../RNASEQUP/rnasequp_1.html">
            
                <a href="../RNASEQUP/rnasequp_1.html">
            
                    
                    1-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../RNASEQUP/rnasequp_2.html">
            
                <a href="../RNASEQUP/rnasequp_2.html">
            
                    
                    1-2. 上游分析全教程
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../RNASEQ/">
            
                <a href="../RNASEQ/">
            
                    
                    分析2：RNA-seq下游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="../RNASEQ/rnaseq_1.html">
            
                <a href="../RNASEQ/rnaseq_1.html">
            
                    
                    2-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="../RNASEQ/rnaseq_2.html">
            
                <a href="../RNASEQ/rnaseq_2.html">
            
                    
                    2-2. 差异表达基因(DEG)分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="../RNASEQ/rnaseq_3.html">
            
                <a href="../RNASEQ/rnaseq_3.html">
            
                    
                    2-3. 通路富集分析(GSEA)
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../SCGLUE/">
            
                <a href="../SCGLUE/">
            
                    
                    分析3：单细胞样本对齐
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" data-path="../SCGLUE/scglue_1.html">
            
                <a href="../SCGLUE/scglue_1.html">
            
                    
                    3-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.2" data-path="../SCGLUE/scglue_2.html">
            
                <a href="../SCGLUE/scglue_2.html">
            
                    
                    3-2. 数据预处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.3" data-path="../SCGLUE/scglue_3.html">
            
                <a href="../SCGLUE/scglue_3.html">
            
                    
                    3-3. 模型准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.4" data-path="../SCGLUE/scglue_4.html">
            
                <a href="../SCGLUE/scglue_4.html">
            
                    
                    3-4. 模型训练
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.5" data-path="../SCGLUE/scglue_5.html">
            
                <a href="../SCGLUE/scglue_5.html">
            
                    
                    3-5. 对齐细胞
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="./">
            
                <a href="./">
            
                    
                    分析4：MOFA单细胞多组学因子分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.1" data-path="mofa_1.html">
            
                <a href="mofa_1.html">
            
                    
                    4-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2" data-path="mofa_2.html">
            
                <a href="mofa_2.html">
            
                    
                    4-2. 单细胞样本对齐
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.3" data-path="mofa_3.html">
            
                <a href="mofa_3.html">
            
                    
                    4-3. scRNA-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.7.4" data-path="mofa_4.html">
            
                <a href="mofa_4.html">
            
                    
                    4-4. scATAC-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.5" data-path="mofa_5.html">
            
                <a href="mofa_5.html">
            
                    
                    4-5. MOFA模型构建
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.6" data-path="mofa_6.html">
            
                <a href="mofa_6.html">
            
                    
                    4-6. MOFA下游分析-R语言环节
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.7" data-path="mofa_7.html">
            
                <a href="mofa_7.html">
            
                    
                    4-7. MOFA下游分析-Python环节
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >4-4. scATAC-seq数据处理</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="mofa&#x5206;&#x6790;&#xFF1A;scatac-seq&#x6570;&#x636E;&#x5904;&#x7406;">MOFA&#x5206;&#x6790;&#xFF1A;scATAC-seq&#x6570;&#x636E;&#x5904;&#x7406;</h1>
<p>&#x7531;&#x4E8E;&#x5355;&#x7EC6;&#x80DE;ATAC-seq&#x6570;&#x636E;&#x4F9D;&#x7136;&#x8F83;&#x5927;&#xFF0C;&#x6211;&#x4EEC;&#x9700;&#x8981;&#x5BF9;&#x5176;&#x8FDB;&#x884C;&#x8FC7;&#x6EE4;</p>
<h2 id="1-&#x6570;&#x636E;&#x51C6;&#x5907;">1. &#x6570;&#x636E;&#x51C6;&#x5907;</h2>
<h3 id="11-&#x5BFC;&#x5165;&#x5305;">1.1 &#x5BFC;&#x5165;&#x5305;</h3>
<pre><code class="lang-python"><span class="hljs-keyword">import</span> anndata
<span class="hljs-keyword">import</span> networkx <span class="hljs-keyword">as</span> nx
<span class="hljs-keyword">import</span> scanpy <span class="hljs-keyword">as</span> sc
<span class="hljs-keyword">from</span> matplotlib <span class="hljs-keyword">import</span> rcParams
<span class="hljs-keyword">import</span> pandas <span class="hljs-keyword">as</span> pd
<span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np
<span class="hljs-keyword">import</span> episcanpy.api <span class="hljs-keyword">as</span> epi
</code></pre>
<h3 id="12-&#x5BFC;&#x5165;&#x6570;&#x636E;">1.2 &#x5BFC;&#x5165;&#x6570;&#x636E;</h3>
<pre><code class="lang-python">atac_pair = anndata.read_h5ad(<span class="hljs-string">&quot;atac_pair.h5ad&quot;</span>)
new_pair=pd.read_csv(<span class="hljs-string">&apos;mofa_pre_pair.csv&apos;</span>)
</code></pre>
<h3 id="13-&#x914D;&#x5BF9;&#x7EC6;&#x80DE;&#x91CD;&#x547D;&#x540D;">1.3 &#x914D;&#x5BF9;&#x7EC6;&#x80DE;&#x91CD;&#x547D;&#x540D;</h3>
<pre><code class="lang-python"><span class="hljs-comment">#new_cell=[]</span>
<span class="hljs-comment">#for i in atac_pair.obs.index:</span>
<span class="hljs-comment">#  new_cell.append(new_pair[new_pair[&apos;scATAC&apos;]==i][&apos;sample&apos;].iloc[0])</span>
<span class="hljs-comment">#atac_pair.obs.index=new_cell</span>
<span class="hljs-comment">#new_cell[:5]</span>

r1=atac_pair[new_pair[<span class="hljs-string">&apos;scATAC&apos;</span>]]
r1.obs.index=new_pair.index.values
r1.write_h5ad(<span class="hljs-string">&apos;atac_mofa_pre1.h5ad&apos;</span>,compression=<span class="hljs-string">&quot;gzip&quot;</span>)<span class="hljs-comment">#&#x5FC5;&#x987B;&#x5148;&#x4FDD;&#x5B58;</span>
atac_pair=anndata.read_h5ad(<span class="hljs-string">&quot;atac_mofa_pre1.h5ad&quot;</span>)
</code></pre>
<h3 id="14-&#x67D3;&#x8272;&#x4F53;&#x4F4D;&#x7F6E;&#x91CD;&#x547D;&#x540D;">1.4 &#x67D3;&#x8272;&#x4F53;&#x4F4D;&#x7F6E;&#x91CD;&#x547D;&#x540D;</h3>
<pre><code class="lang-python">lo=[]
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> atac_pair.var_names:
  lo.append(i.replace(<span class="hljs-string">&apos;:&apos;</span>,<span class="hljs-string">&apos;_&apos;</span>).replace(<span class="hljs-string">&apos;-&apos;</span>,<span class="hljs-string">&apos;_&apos;</span>))
atac_pair.var.index=lo
</code></pre>
<h3 id="15-&#x67D3;&#x8272;&#x4F53;&#x4F4D;&#x7F6E;&#x76F8;&#x5173;&#x57FA;&#x56E0;&#x6807;&#x6CE8;">1.5 &#x67D3;&#x8272;&#x4F53;&#x4F4D;&#x7F6E;&#x76F8;&#x5173;&#x57FA;&#x56E0;&#x6807;&#x6CE8;</h3>
<h4 id="151-annotation&#x4E0B;&#x8F7D;">1.5.1 Annotation&#x4E0B;&#x8F7D;</h4>
<pre><code class="lang-shell">wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz -O gencode.v19.annotation.gtf.gz
gunzip gencode.v19.annotation.gtf
</code></pre>
<h4 id="152-annotation&#x6807;&#x6CE8;">1.5.2 Annotation&#x6807;&#x6CE8;</h4>
<pre><code class="lang-python">epi.tl.find_genes(atac_pair,
           gtf_file=<span class="hljs-string">&apos;gencode.v19.annotation.gtf&apos;</span>,
           key_added=<span class="hljs-string">&apos;transcript_annotation&apos;</span>,
           upstream=<span class="hljs-number">2000</span>,
           feature_type=<span class="hljs-string">&apos;transcript&apos;</span>,
           annotation=<span class="hljs-string">&apos;HAVANA&apos;</span>,
           raw=<span class="hljs-keyword">False</span>)
</code></pre>
<h2 id="2-&#x6570;&#x636E;&#x5904;&#x7406;">2. &#x6570;&#x636E;&#x5904;&#x7406;</h2>
<h3 id="21-&#x8FC7;&#x6EE4;&#x7A7A;feature&#x8DDF;barcode">2.1 &#x8FC7;&#x6EE4;&#x7A7A;feature&#x8DDF;barcode</h3>
<pre><code class="lang-python"><span class="hljs-comment"># remove any potential empty features or barcodes</span>
epi.pp.filter_cells(atac_pair, min_features=<span class="hljs-number">1</span>)
epi.pp.filter_features(atac_pair, min_cells=<span class="hljs-number">1</span>)
atac_pair
</code></pre>
<blockquote>
<p>AnnData object with n_obs &#xD7; n_vars = 23770 &#xD7; 214125</p>
</blockquote>
<h3 id="22-&#x5BF9;&#x6570;&#x5316;">2.2 &#x5BF9;&#x6570;&#x5316;</h3>
<pre><code class="lang-python">atac_pair.obs[<span class="hljs-string">&apos;log_nb_features&apos;</span>] = [np.log10(x) <span class="hljs-keyword">for</span> x <span class="hljs-keyword">in</span> atac_pair.obs[<span class="hljs-string">&apos;nb_features&apos;</span>]]
epi.pl.violin(atac_pair, [<span class="hljs-string">&apos;nb_features&apos;</span>])
epi.pl.violin(atac_pair, [<span class="hljs-string">&apos;log_nb_features&apos;</span>])
</code></pre>
<p><img src="mofa_4.assets/do8.png" alt="do8"><img src="mofa_4.assets/do7.png" alt="do7"></p>
<h3 id="23-&#x8BBE;&#x7F6E;&#x6BCF;&#x4E2A;&#x7EC6;&#x80DE;&#x6700;&#x5C0F;&#x5177;&#x6709;&#x7684;feature&#x6570;">2.3 &#x8BBE;&#x7F6E;&#x6BCF;&#x4E2A;&#x7EC6;&#x80DE;&#x6700;&#x5C0F;&#x5177;&#x6709;&#x7684;feature&#x6570;</h3>
<pre><code class="lang-python"><span class="hljs-comment"># set a minimum number of cells to keep </span>
min_features = <span class="hljs-number">400</span>

epi.pp.coverage_cells(atac_pair, binary=<span class="hljs-keyword">True</span>, log=<span class="hljs-keyword">False</span>, bins=<span class="hljs-number">50</span>,
               threshold=min_features, save=<span class="hljs-string">&apos;Buenrostro_bulk_peaks_coverage_cells.png&apos;</span>)
epi.pp.coverage_cells(atac_pair, binary=<span class="hljs-keyword">True</span>, log=<span class="hljs-number">10</span>, bins=<span class="hljs-number">50</span>,
               threshold=min_features, save=<span class="hljs-string">&apos;Buenrostro_bulk_peaks_coverage_cells_log10.png&apos;</span>)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/do9.png" alt="do9"></p>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/do10.png" alt="do10"></p>
<h3 id="24-&#x8BBE;&#x7F6E;&#x6BCF;&#x4E2A;feature&#x5B58;&#x5728;&#x4E8E;&#x6700;&#x5C0F;&#x7684;&#x7EC6;&#x80DE;&#x6570;">2.4 &#x8BBE;&#x7F6E;&#x6BCF;&#x4E2A;feature&#x5B58;&#x5728;&#x4E8E;&#x6700;&#x5C0F;&#x7684;&#x7EC6;&#x80DE;&#x6570;</h3>
<pre><code class="lang-python"><span class="hljs-comment"># minimum number of cells sharing a feature</span>
min_cells = <span class="hljs-number">80</span>
epi.pp.coverage_features(atac_pair, binary=<span class="hljs-keyword">True</span>, log=<span class="hljs-keyword">False</span>, 
                        threshold=min_cells, save=<span class="hljs-string">&apos;Buenrostro_bulk_peaks_coverage_peaks.png&apos;</span>)
epi.pp.coverage_features(atac_pair, binary=<span class="hljs-keyword">True</span>, log=<span class="hljs-keyword">True</span>, 
                        threshold=min_cells, save=<span class="hljs-string">&apos;Buenrostro_bulk_peaks_coverage_peaks_log10.png&apos;</span>)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/do11.png" alt="do11"></p>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/do12.png" alt="do12"></p>
<h3 id="25-&#x8FC7;&#x6EE4;&#x7EC6;&#x80DE;&#x4E0E;feature">2.5 &#x8FC7;&#x6EE4;&#x7EC6;&#x80DE;&#x4E0E;feature</h3>
<pre><code class="lang-python">min_features = <span class="hljs-number">400</span>
epi.pp.filter_cells(atac_pair, min_features=min_features)
min_cells = <span class="hljs-number">80</span>
epi.pp.filter_features(atac_pair, min_cells=min_cells)
</code></pre>
<h3 id="26-&#x8BA1;&#x7B97;&#x9AD8;&#x53D8;feature">2.6 &#x8BA1;&#x7B97;&#x9AD8;&#x53D8;feature</h3>
<pre><code class="lang-python">min_score_value = <span class="hljs-number">0.515</span>
nb_feature_selected = <span class="hljs-number">20000</span>
epi.pl.variability_features(atac_pair,log=<span class="hljs-keyword">None</span>,
                     min_score=min_score_value, nb_features=nb_feature_selected,
                     save=<span class="hljs-string">&apos;variability_features_plot_bonemarrow_peakmatrix.png&apos;</span>)

epi.pl.variability_features(atac_pair,log=<span class="hljs-string">&apos;log10&apos;</span>,
                     min_score=min_score_value, nb_features=nb_feature_selected,
                     save=<span class="hljs-string">&apos;variability_features_plot_bonemarrow_peakmatrix_log10.png&apos;</span>)
</code></pre>
<p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_4.assets\do13.png" alt="do13"></p>
<p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_4.assets\do14.png" alt="do14"></p>
<h2 id="3-&#x8FC7;&#x6EE4;&#x9AD8;&#x53D8;feature&#x7684;cell">3. &#x8FC7;&#x6EE4;&#x9AD8;&#x53D8;feature&#x7684;cell</h2>
<pre><code class="lang-python">atac_pair.raw=atac_pair
<span class="hljs-comment"># create a new AnnData containing only the most variable features</span>
atac_pair = epi.pp.select_var_feature(atac_pair,nb_features=nb_feature_selected,show=<span class="hljs-keyword">False</span>,copy=<span class="hljs-keyword">True</span>)
epi.pl.violin(atac_pair, [<span class="hljs-string">&apos;nb_features&apos;</span>])
epi.pl.violin(atac_pair, [<span class="hljs-string">&apos;log_nb_features&apos;</span>])
</code></pre>
<p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_4.assets\do15.png" alt="do15"></p>
<p><img src="C:\Users\FernandoZeng\Desktop\biobook\MOFA\mofa_4.assets\do16-1636004475065.png" alt="do16"></p>
<pre><code class="lang-python">epi.pp.filter_cells(atac_pair, min_features=<span class="hljs-number">1000</span>)
epi.pp.filter_cells(atac_pair, max_features=<span class="hljs-number">4000</span>)
atac_pair.write_h5ad(<span class="hljs-string">&apos;atac_mofa.h5ad&apos;</span>,compression=<span class="hljs-string">&quot;gzip&quot;</span>)
</code></pre>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="mofa_3.html" class="navigation navigation-prev " aria-label="Previous page: 4-3. scRNA-seq数据处理">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="mofa_5.html" class="navigation navigation-next " aria-label="Next page: 4-5. MOFA模型构建">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"4-4. scATAC-seq数据处理","level":"1.7.4","depth":2,"next":{"title":"4-5. MOFA模型构建","level":"1.7.5","depth":2,"path":"MOFA/mofa_5.md","ref":"MOFA/mofa_5.md","articles":[]},"previous":{"title":"4-3. scRNA-seq数据处理","level":"1.7.3","depth":2,"path":"MOFA/mofa_3.md","ref":"MOFA/mofa_3.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{},"plugins":["expandable-chapters"],"pluginsConfig":{"expandable-chapters":{},"highlight":{},"search":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"fontsettings":{"theme":"white","family":"sans","size":2},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"MOFA/mofa_4.md","mtime":"2021-12-07T03:05:29.779Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2021-12-07T08:49:01.522Z"},"basePath":"..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

