
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>6-3. 细胞亚群分析 · GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-back-to-top-button/plugin.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-cuav-chapters/cuav-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-theme-ivoryhza/ivory.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-theme-ivoryhza/footer.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="book-4.html" />
    
    
    <link rel="prev" href="book-2.html" />
    

    </head>
    <body>
        
<div class="book">
	<div class="header-inner">
		<!-- LOGO -->
		<div class="header-inner-left">
			<div class="logo"></div>
			<span class="title"></span>
		</div>

		<!-- Nav -->
		<div class="header-inner-right">
			<ul class="header-nav">
				<li>
					<a href="http://..." target="_blank">HOME</a>
				</li>
			</ul>
			<!-- Search -->
			
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

		</div>


	</div>

	<div class="book-summary">
		<div class="book-summary-title">目录</div>
		
		
		<nav role="navigation">
			


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Bioinformatics Tutorial
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../">
            
                <a href="../">
            
                    
                    目录
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../CONFIG/">
            
                <a href="../CONFIG/">
            
                    
                    配置：基础环境配置
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../CONFIG/config_1.html">
            
                <a href="../CONFIG/config_1.html">
            
                    
                    配置1：Python环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../CONFIG/config_2.html">
            
                <a href="../CONFIG/config_2.html">
            
                    
                    配置2：Linux环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../CONFIG/config_3.html">
            
                <a href="../CONFIG/config_3.html">
            
                    
                    配置3：R语言环境
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../RNASEQUP/">
            
                <a href="../RNASEQUP/">
            
                    
                    分析1：RNA-seq上游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../RNASEQUP/rnasequp_1.html">
            
                <a href="../RNASEQUP/rnasequp_1.html">
            
                    
                    1-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../RNASEQUP/rnasequp_2.html">
            
                <a href="../RNASEQUP/rnasequp_2.html">
            
                    
                    1-2. 上游分析全教程
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../RNASEQ/">
            
                <a href="../RNASEQ/">
            
                    
                    分析2：RNA-seq下游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="../RNASEQ/rnaseq_1.html">
            
                <a href="../RNASEQ/rnaseq_1.html">
            
                    
                    2-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="../RNASEQ/rnaseq_2.html">
            
                <a href="../RNASEQ/rnaseq_2.html">
            
                    
                    2-2. 差异表达基因(DEG)分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="../RNASEQ/rnaseq_3.html">
            
                <a href="../RNASEQ/rnaseq_3.html">
            
                    
                    2-3. 通路富集分析(GSEA)
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../SCGLUE/">
            
                <a href="../SCGLUE/">
            
                    
                    分析3：单细胞样本对齐
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" data-path="../SCGLUE/scglue_1.html">
            
                <a href="../SCGLUE/scglue_1.html">
            
                    
                    3-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.2" data-path="../SCGLUE/scglue_2.html">
            
                <a href="../SCGLUE/scglue_2.html">
            
                    
                    3-2. 数据预处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.3" data-path="../SCGLUE/scglue_3.html">
            
                <a href="../SCGLUE/scglue_3.html">
            
                    
                    3-3. 模型准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.4" data-path="../SCGLUE/scglue_4.html">
            
                <a href="../SCGLUE/scglue_4.html">
            
                    
                    3-4. 模型训练
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.5" data-path="../SCGLUE/scglue_5.html">
            
                <a href="../SCGLUE/scglue_5.html">
            
                    
                    3-5. 对齐细胞
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="../MOFA/">
            
                <a href="../MOFA/">
            
                    
                    分析4：MOFA单细胞多组学因子分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.1" data-path="../MOFA/mofa_1.html">
            
                <a href="../MOFA/mofa_1.html">
            
                    
                    4-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2" data-path="../MOFA/mofa_2.html">
            
                <a href="../MOFA/mofa_2.html">
            
                    
                    4-2. 单细胞样本对齐
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.3" data-path="../MOFA/mofa_3.html">
            
                <a href="../MOFA/mofa_3.html">
            
                    
                    4-3. scRNA-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.4" data-path="../MOFA/mofa_4.html">
            
                <a href="../MOFA/mofa_4.html">
            
                    
                    4-4. scATAC-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.5" data-path="../MOFA/mofa_5.html">
            
                <a href="../MOFA/mofa_5.html">
            
                    
                    4-5. MOFA模型构建
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.6" data-path="../MOFA/mofa_6.html">
            
                <a href="../MOFA/mofa_6.html">
            
                    
                    4-6. MOFA下游分析-R语言环节
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.7" data-path="../MOFA/mofa_7.html">
            
                <a href="../MOFA/mofa_7.html">
            
                    
                    4-7. MOFA下游分析-Python环节
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.8" data-path="../TMT/">
            
                <a href="../TMT/">
            
                    
                    分析5：TMT蛋白组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.8.1" data-path="../TMT/tmt_1.html">
            
                <a href="../TMT/tmt_1.html">
            
                    
                    5-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8.2" data-path="../TMT/tmt_2.html">
            
                <a href="../TMT/tmt_2.html">
            
                    
                    5-2. TMT蛋白组学上游分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8.3" data-path="../TMT/tmt_3.html">
            
                <a href="../TMT/tmt_3.html">
            
                    
                    5-3. TMT蛋白组学下游分析
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.9" data-path="./">
            
                <a href="./">
            
                    
                    分析6：多组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.9.1" data-path="book-1.html">
            
                <a href="book-1.html">
            
                    
                    6-1. 数据准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.2" data-path="book-2.html">
            
                <a href="book-2.html">
            
                    
                    6-2. GLUE多组学整合
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.9.3" data-path="book-3.html">
            
                <a href="book-3.html">
            
                    
                    6-3. 细胞亚群分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.4" data-path="book-4.html">
            
                <a href="book-4.html">
            
                    
                    6-4. 差异表达分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.5" data-path="book-5.html">
            
                <a href="book-5.html">
            
                    
                    6-5. 细胞通讯分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.6" data-path="book-6.html">
            
                <a href="book-6.html">
            
                    
                    6-6. 动态调控分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.7" data-path="book-7.html">
            
                <a href="book-7.html">
            
                    
                    6-7. 转录因子网络
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.10" data-path="../PLOT/">
            
                <a href="../PLOT/">
            
                    
                    绘图：Python可视化
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.10.1" data-path="../PLOT/plot_1.html">
            
                <a href="../PLOT/plot_1.html">
            
                    
                    1-1. 火山图
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.10.2" data-path="../PLOT/plot_2.html">
            
                <a href="../PLOT/plot_2.html">
            
                    
                    1-2. 箱线图
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


		</nav>
		
		
	</div>

	<!-- Content nav -->
	<div class="book-anchor">
		<div class="book-anchor-title">大纲</div>
		<div class="book-anchor-body">

		</div>
	</div>

	<div class="book-body">
		
		<div class="body-inner">
			
			

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >6-3. 细胞亚群分析</a>
    </h1>
</div>




			<div class="page-wrapper" tabindex="-1" role="main">
				<div class="page-inner">
					
<div id="book-search-results">
    <div class="search-noresults">
    
					<section class="normal markdown-section">
						
						<h2 id="4-&#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;&#x5206;&#x6790;-&#x771F;&#x5B9E;&#x6848;&#x4F8B;">4. &#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;&#x5206;&#x6790;-&#x771F;&#x5B9E;&#x6848;&#x4F8B;</h2>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x5BF9;B&#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;&#x8FDB;&#x884C;&#x4E0B;&#x6E38;&#x5206;&#x6790;&#xFF08;&#x57FA;&#x4E8E;&#x4E73;&#x817A;&#x764C;&#x6DCB;&#x5DF4;&#x7ED3;&#x8F6C;&#x79FB;&#x7684;B&#x7EC6;&#x80DE;&#x6570;&#x636E;&#xFF09;</p>
<h3 id="41-&#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;&#x5206;&#x6790;">4.1 &#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;&#x5206;&#x6790;</h3>
<p>&#x6211;&#x4EEC;&#x9996;&#x5148;&#x9700;&#x8981;&#x5BFC;&#x5165;&#x5305;&#xFF0C;&#x4EE5;&#x53CA;&#x4E00;&#x4E9B;&#x76F8;&#x5173;&#x51FD;&#x6570;&#x7684;&#x51C6;&#x5907;</p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x5BFC;&#x5165;&#x5305;</span>
<span class="hljs-keyword">import</span> anndata
print(<span class="hljs-string">&apos;anndata(Ver): &apos;</span>,anndata.__version__)
<span class="hljs-keyword">import</span> scanpy <span class="hljs-keyword">as</span> sc
print(<span class="hljs-string">&apos;scanpy(Ver): &apos;</span>,sc.__version__)
<span class="hljs-keyword">import</span> scltnn <span class="hljs-comment">#&#x975E;&#x5FC5;&#x9700;</span>
print(<span class="hljs-string">&apos;scltnn(Ver): &apos;</span>,scltnn.__version__)
<span class="hljs-keyword">import</span> matplotlib.pyplot <span class="hljs-keyword">as</span> plt
<span class="hljs-keyword">import</span> matplotlib
print(<span class="hljs-string">&apos;matplotlib(Ver): &apos;</span>,matplotlib.__version__)
<span class="hljs-keyword">import</span> seaborn <span class="hljs-keyword">as</span> sns
print(<span class="hljs-string">&apos;seaborn(Ver): &apos;</span>,sns.__version__)
<span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np
print(<span class="hljs-string">&apos;numpy(Ver): &apos;</span>,np.__version__)
<span class="hljs-keyword">import</span> pandas <span class="hljs-keyword">as</span> pd
print(<span class="hljs-string">&apos;pandas(Ver): &apos;</span>,pd.__version__)
<span class="hljs-keyword">import</span> scvelo <span class="hljs-keyword">as</span> scv
print(<span class="hljs-string">&apos;scvelo(Ver): &apos;</span>,scv.__version__)
<span class="hljs-keyword">import</span> Pyomic
print(<span class="hljs-string">&apos;Pyomic(Ver): &apos;</span>,Pyomic.__version__)

<span class="hljs-comment">#&#x7ED8;&#x56FE;&#x53C2;&#x6570;&#x8BBE;&#x7F6E;</span>
sc.settings.verbosity = <span class="hljs-number">3</span>             <span class="hljs-comment"># verbosity: errors (0), warnings (1), info (2), hints (3)</span>
sc.settings.set_figure_params(dpi=<span class="hljs-number">80</span>, facecolor=<span class="hljs-string">&apos;white&apos;</span>)

sc_color=[<span class="hljs-string">&apos;#7CBB5F&apos;</span>,<span class="hljs-string">&apos;#368650&apos;</span>,<span class="hljs-string">&apos;#A499CC&apos;</span>,<span class="hljs-string">&apos;#5E4D9A&apos;</span>,<span class="hljs-string">&apos;#78C2ED&apos;</span>,<span class="hljs-string">&apos;#866017&apos;</span>,<span class="hljs-string">&apos;#9F987F&apos;</span>, <span class="hljs-string">&apos;#E0DFED&apos;</span>, <span class="hljs-string">&apos;#EF7B77&apos;</span>, <span class="hljs-string">&apos;#279AD7&apos;</span>,
 <span class="hljs-string">&apos;#F0EEF0&apos;</span>, <span class="hljs-string">&apos;#1F577B&apos;</span>, <span class="hljs-string">&apos;#A56BA7&apos;</span>, <span class="hljs-string">&apos;#E0A7C8&apos;</span>, <span class="hljs-string">&apos;#E069A6&apos;</span>, <span class="hljs-string">&apos;#941456&apos;</span>, <span class="hljs-string">&apos;#FCBC10&apos;</span>, <span class="hljs-string">&apos;#EAEFC5&apos;</span>, <span class="hljs-string">&apos;#01A0A7&apos;</span>, <span class="hljs-string">&apos;#75C8CC&apos;</span>, 
<span class="hljs-string">&apos;#F0D7BC&apos;</span>, <span class="hljs-string">&apos;#D5B26C&apos;</span>, <span class="hljs-string">&apos;#D5DA48&apos;</span>, <span class="hljs-string">&apos;#B6B812&apos;</span>,<span class="hljs-string">&apos;#9DC3C3&apos;</span>, <span class="hljs-string">&apos;#A89C92&apos;</span>, <span class="hljs-string">&apos;#FEE00C&apos;</span>,<span class="hljs-string">&apos;#FEF2A1&apos;</span>]
</code></pre>
<p>&#x7D27;&#x63A5;&#x7740;&#xFF0C;&#x6211;&#x4EEC;&#x5BFC;&#x5165;&#x4E4B;&#x524D;&#x6807;&#x6CE8;&#x597D;&#x7EC6;&#x80DE;&#x7C7B;&#x578B;&#x7684;rna.h5ad&#x6587;&#x4EF6;&#xFF0C;&#x5E76;&#x9009;&#x53D6;&#x5176;&#x4E2D;&#x7684;B&#x7EC6;&#x80DE;</p>
<pre><code class="lang-python">adata=sc.read(<span class="hljs-string">&apos;../cellanno/rna_anno.h5ad&apos;</span>)
<span class="hljs-comment">#&#x7B5B;&#x9009;</span>
adata=adata[adata.obs[<span class="hljs-string">&apos;major_celltype&apos;</span>]==<span class="hljs-string">&apos;B-cell&apos;</span>]
</code></pre>
<pre><code class="lang-shell">View of AnnData object with n_obs &#xD7; n_vars = 5899 &#xD7; 1945
    obs: &apos;Type&apos;, &apos;initial_size_spliced&apos;, &apos;initial_size_unspliced&apos;, &apos;initial_size&apos;, &apos;domain&apos;, &apos;n_genes&apos;, &apos;leiden&apos;, &apos;n_counts&apos;, &apos;major_celltype&apos;
    var: &apos;gene_ids&apos;, &apos;feature_types&apos;, &apos;genome&apos;, &apos;Accession&apos;, &apos;Chromosome&apos;, &apos;End&apos;, &apos;Start&apos;, &apos;Strand&apos;, &apos;n_cells&apos;, &apos;highly_variable&apos;, &apos;highly_variable_rank&apos;, &apos;means&apos;, &apos;variances&apos;, &apos;variances_norm&apos;, &apos;mean&apos;, &apos;std&apos;, &apos;chrom&apos;, &apos;chromStart&apos;, &apos;chromEnd&apos;, &apos;name&apos;, &apos;score&apos;, &apos;strand&apos;, &apos;thickStart&apos;, &apos;thickEnd&apos;, &apos;itemRgb&apos;, &apos;blockCount&apos;, &apos;blockSizes&apos;, &apos;blockStarts&apos;, &apos;gene_id&apos;, &apos;gene_type&apos;, &apos;hgnc_id&apos;, &apos;havana_gene&apos;, &apos;tag&apos;, &apos;dell&apos;
    uns: &apos;Type_colors&apos;, &apos;dendrogram_leiden&apos;, &apos;hvg&apos;, &apos;leiden&apos;, &apos;leiden_colors&apos;, &apos;leiden_sizes&apos;, &apos;log1p&apos;, &apos;major_celltype_colors&apos;, &apos;neighbors&apos;, &apos;paga&apos;, &apos;pca&apos;, &apos;rank_genes_groups&apos;, &apos;umap&apos;
    obsm: &apos;X_glue&apos;, &apos;X_pca&apos;, &apos;X_umap&apos;
    varm: &apos;PCs&apos;, &apos;X_glue&apos;
    layers: &apos;Ms&apos;, &apos;Mu&apos;, &apos;ambiguous&apos;, &apos;counts&apos;, &apos;matrix&apos;, &apos;spliced&apos;, &apos;unspliced&apos;
    obsp: &apos;connectivities&apos;, &apos;distances&apos;
</code></pre>
<p>&#x6211;&#x4EEC;&#x5BF9;&#x6570;&#x636E;&#x91CD;&#x65B0;&#x5206;&#x7C7B;&#x522B;&#xFF0C;&#x63D0;&#x53D6;&#x5176;&#x4E2D;&#x7684;B&#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;</p>
<pre><code class="lang-python">sc.tl.pca(adata, n_comps=<span class="hljs-number">100</span>, svd_solver=<span class="hljs-string">&quot;arpack&quot;</span>)
sc.pp.neighbors(adata, use_rep=<span class="hljs-string">&quot;X_glue&quot;</span>, metric=<span class="hljs-string">&quot;cosine&quot;</span>,n_neighbors=<span class="hljs-number">15</span>, random_state = <span class="hljs-number">112</span>)
sc.tl.leiden(adata)
sc.tl.paga(adata)
sc.pl.paga(adata, plot=<span class="hljs-keyword">False</span>)  <span class="hljs-comment"># remove `plot=False` if you want to see the coarse-grained graph</span>
sc.tl.umap(adata,init_pos=<span class="hljs-string">&apos;paga&apos;</span>)
sc.pl.umap(adata,color=<span class="hljs-string">&apos;leiden&apos;</span>)
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">computing PCA
    on highly variable genes
    with n_comps=100
    finished (0:00:01)
computing neighbors
    finished: added to `.uns[&apos;neighbors&apos;]`
    `.obsp[&apos;distances&apos;]`, distances for each pair of neighbors
    `.obsp[&apos;connectivities&apos;]`, weighted adjacency matrix (0:00:00)
running Leiden clustering
    finished: found 13 clusters and added
    &apos;leiden&apos;, the cluster labels (adata.obs, categorical) (0:00:00)
running PAGA
    finished: added
    &apos;paga/connectivities&apos;, connectivities adjacency (adata.uns)
    &apos;paga/connectivities_tree&apos;, connectivities subtree (adata.uns) (0:00:00)
--&gt; added &apos;pos&apos;, the PAGA positions (adata.uns[&apos;paga&apos;])
computing UMAP
    finished: added
    &apos;X_umap&apos;, UMAP coordinates (adata.obsm) (0:00:05)
</code></pre>
<p><img src="book-3.assets/image-20221019004245874-6119960.png" alt="image-20221019004245874" style="zoom:50%;"></p>
<p>&#x6211;&#x4EEC;&#x63A5;&#x7740;&#x6784;&#x5EFA;velocity&#x7684;&#x52A8;&#x6001;&#x8C03;&#x63A7;&#x8FC7;&#x7A0B;&#x7F51;</p>
<pre><code class="lang-python">scv.tl.recover_dynamics(adata,n_jobs=<span class="hljs-number">6</span>)
scv.tl.velocity(adata, mode=<span class="hljs-string">&apos;dynamical&apos;</span>)
scv.tl.velocity_graph(adata)
scv.pl.velocity_embedding_stream(adata, basis=<span class="hljs-string">&apos;umap&apos;</span>,color=[<span class="hljs-string">&apos;major_celltype&apos;</span>,<span class="hljs-string">&apos;leiden&apos;</span>,<span class="hljs-string">&apos;Type&apos;</span>])
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">recovering dynamics (using 6/8 cores)
Error displaying widget: model not found
    finished (0:00:13) --&gt; added 
    &apos;fit_pars&apos;, fitted parameters for splicing dynamics (adata.var)
computing velocities
    finished (0:00:00) --&gt; added 
    &apos;velocity&apos;, velocity vectors for each individual cell (adata.layers)
computing velocity graph (using 1/8 cores)
Error displaying widget: model not found
    finished (0:00:01) --&gt; added 
    &apos;velocity_graph&apos;, sparse matrix with cosine correlations (adata.uns)
computing velocity embedding
    finished (0:00:00) --&gt; added
    &apos;velocity_umap&apos;, embedded velocity vectors (adata.obsm)
</code></pre>
<p><img src="book-3.assets/image-20221019004442469-6119965.png" alt="image-20221019004442469" style="zoom:50%;"></p>
<p>&#x6211;&#x4EEC;&#x63A5;&#x7740;&#x8BA1;&#x7B97;&#x62DF;&#x65F6;&#x5E8F;</p>
<pre><code class="lang-python">scv.tl.latent_time(adata)
scv.pl.scatter(adata, color=<span class="hljs-string">&apos;latent_time&apos;</span>, color_map=<span class="hljs-string">&apos;Reds&apos;</span>, size=<span class="hljs-number">80</span>)
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">computing terminal states
    identified 4 regions of root cells and 1 region of end points .
    finished (0:00:00) --&gt; added
    &apos;root_cells&apos;, root cells of Markov diffusion process (adata.obs)
    &apos;end_points&apos;, end points of Markov diffusion process (adata.obs)
computing latent time using root_cells as prior
    finished (0:00:00) --&gt; added 
    &apos;latent_time&apos;, shared time (adata.obs)
</code></pre>
<p><img src="book-3.assets/image-20221019004556158-6119968.png" alt="image-20221019004556158" style="zoom:50%;"></p>
<p>&#x7136;&#x540E;&#xFF0C;&#x6211;&#x4EEC;&#x4F7F;&#x7528;B&#x7EC6;&#x80DE;&#x7684;&#x5B57;&#x5178;&#xFF0C;&#x6765;&#x5BF9;B&#x7EC6;&#x80DE;&#x7684;&#x7EC6;&#x80DE;&#x7C7B;&#x578B;&#x8FDB;&#x884C;&#x6807;&#x6CE8;&#xFF0C;&#x5E76;&#x4E14;&#x4F7F;&#x7528;t-test&#x627E;&#x51FA;&#x6BCF;&#x4E00;&#x7C7B;leiden&#x7684;marker&#x57FA;&#x56E0;&#x7528;&#x4E8E;&#x8F85;&#x52A9;&#x6CE8;&#x91CA;</p>
<pre><code class="lang-python"><span class="hljs-comment">#B&#x7EC6;&#x80DE;&#x5B57;&#x5178;</span>
res_marker_dict={
    <span class="hljs-string">&apos;na&#xEF;ve B cell&apos;</span>:[<span class="hljs-string">&apos;IGHD&apos;</span>, <span class="hljs-string">&apos;FCER2&apos;</span>, <span class="hljs-string">&apos;TCL1A&apos;</span>, <span class="hljs-string">&apos;IL4R&apos;</span>],
    <span class="hljs-string">&apos;memory B cell&apos;</span>:[<span class="hljs-string">&apos;CD27&apos;</span>,<span class="hljs-string">&apos;AIM2&apos;</span>,<span class="hljs-string">&apos;TNFRSF13B&apos;</span>],
    <span class="hljs-string">&apos;Pre-B cell&apos;</span>:[<span class="hljs-string">&apos;TMSB4X&apos;</span>,<span class="hljs-string">&apos;TMSB10&apos;</span>],
    <span class="hljs-string">&apos;Breg&apos;</span>: [<span class="hljs-string">&apos;CD24&apos;</span>,<span class="hljs-string">&apos;CD27&apos;</span>],
    <span class="hljs-string">&apos;(GC) B cell&apos;</span>:[ <span class="hljs-string">&apos;SUGCT&apos;</span>, <span class="hljs-string">&apos;MME&apos;</span>, <span class="hljs-string">&apos;MKI67&apos;</span>, <span class="hljs-string">&apos;AICDA&apos;</span>],
    <span class="hljs-string">&apos;IgG plasma cell&apos;</span>:[<span class="hljs-string">&apos;JCHAIN&apos;</span>],
    <span class="hljs-string">&apos;IgA plasma cell&apos;</span>:[<span class="hljs-string">&apos;CD38&apos;</span>],
    <span class="hljs-string">&apos;Follicular B Cell&apos;</span>:[<span class="hljs-string">&apos;CD69&apos;</span>,<span class="hljs-string">&apos;CD83&apos;</span>],
    <span class="hljs-string">&apos;activated B cells&apos;</span>:[<span class="hljs-string">&apos;IGHA1&apos;</span>],
    <span class="hljs-string">&apos;Test&apos;</span>:[<span class="hljs-string">&apos;SPN&apos;</span>,<span class="hljs-string">&apos;MALAT1&apos;</span>],
}
sc.tl.dendrogram(adata,<span class="hljs-string">&apos;leiden&apos;</span>)
sc.pl.dotplot(adata, res_marker_dict, <span class="hljs-string">&apos;leiden&apos;</span>, dendrogram=<span class="hljs-keyword">True</span>,standard_scale=<span class="hljs-string">&apos;var&apos;</span>)
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">    using &apos;X_pca&apos; with n_pcs = 100
Storing dendrogram info using `.uns[&apos;dendrogram_leiden&apos;]`
WARNING: Groups are not reordered because the `groupby` categories and the `var_group_labels` are different.
categories: 0, 1, 2, etc.
var_group_labels: na&#xEF;ve B cell, memory B cell, Pre-B cell, etc.
</code></pre>
<p><img src="book-3.assets/image-20221019004807328-6119971.png" alt="image-20221019004807328" style="zoom:50%;"></p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x6CE8;&#x610F;&#xFF0C;&#x8FD9;&#x91CC;&#x4E5F;&#x53EF;&#x4EE5;&#x4F7F;&#x7528;cosg&#x6765;&#x5BFB;&#x627E;marker&#x57FA;&#x56E0;</span>
adata.uns[<span class="hljs-string">&apos;log1p&apos;</span>][<span class="hljs-string">&apos;base&apos;</span>]=<span class="hljs-keyword">None</span>
sc.tl.rank_genes_groups(adata, <span class="hljs-string">&apos;leiden&apos;</span>, method=<span class="hljs-string">&apos;t-test&apos;</span>)
sc.pl.rank_genes_groups_dotplot(adata,groupby=<span class="hljs-string">&apos;leiden&apos;</span>,
                                cmap=<span class="hljs-string">&apos;Spectral_r&apos;</span>,
                                standard_scale=<span class="hljs-string">&apos;var&apos;</span>,n_genes=<span class="hljs-number">3</span>)
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">ranking genes
WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.
    finished (0:00:00)
</code></pre>
<p><img src="book-3.assets/image-20221019004930306-6119974.png" alt="image-20221019004930306" style="zoom:50%;"></p>
<p>&#x6700;&#x540E;&#xFF0C;&#x6211;&#x4EEC;&#x6807;&#x6CE8;leiden&#x5BF9;&#x5E94;&#x7684;&#x7EC6;&#x80DE;&#x7C7B;&#x578B;&#xFF0C;&#x5E76;&#x4E0A;&#x8272;</p>
<pre><code class="lang-python"><span class="hljs-comment">#leiden&#x5BF9;&#x5E94;&#x7684;&#x7EC6;&#x80DE;&#x7C7B;&#x578B;&#x6807;&#x6CE8;</span>
cluster2annotation = {
     <span class="hljs-string">&apos;0&apos;</span>: <span class="hljs-string">&apos;activated B cells&apos;</span>,
     <span class="hljs-string">&apos;1&apos;</span>: <span class="hljs-string">&apos;na&#xEF;ve B cells&apos;</span>,
     <span class="hljs-string">&apos;2&apos;</span>: <span class="hljs-string">&apos;Pre-B cells&apos;</span>,
     <span class="hljs-string">&apos;3&apos;</span>: <span class="hljs-string">&apos;B-cell (MALAT1+)&apos;</span>,
     <span class="hljs-string">&apos;4&apos;</span>: <span class="hljs-string">&apos;activated B cells&apos;</span>,
     <span class="hljs-string">&apos;5&apos;</span>: <span class="hljs-string">&apos;Pre-B cells&apos;</span>,
     <span class="hljs-string">&apos;6&apos;</span>: <span class="hljs-string">&apos;na&#xEF;ve B cells&apos;</span>,
     <span class="hljs-string">&apos;7&apos;</span>: <span class="hljs-string">&apos;activated B cells&apos;</span>,
     <span class="hljs-string">&apos;8&apos;</span>: <span class="hljs-string">&apos;na&#xEF;ve B cells&apos;</span>,
     <span class="hljs-string">&apos;9&apos;</span>: <span class="hljs-string">&apos;Pre-B cells&apos;</span>,
     <span class="hljs-string">&apos;10&apos;</span>: <span class="hljs-string">&apos;Pre-B cells&apos;</span>,
     <span class="hljs-string">&apos;11&apos;</span>: <span class="hljs-string">&apos;na&#xEF;ve B cells&apos;</span>,
     <span class="hljs-string">&apos;12&apos;</span>: <span class="hljs-string">&apos;Plasma B cells&apos;</span>,
}
adata.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>] = adata.obs[<span class="hljs-string">&apos;leiden&apos;</span>].map(cluster2annotation).astype(<span class="hljs-string">&apos;category&apos;</span>)

<span class="hljs-comment">#&#x4E0A;&#x8272;</span>
type_color_B={
    <span class="hljs-string">&apos;na&#xEF;ve B cells&apos;</span>:sc_color[<span class="hljs-number">13</span>],
    <span class="hljs-string">&apos;Follicular B cells&apos;</span>:sc_color[<span class="hljs-number">20</span>],
    <span class="hljs-string">&apos;Pre-B cells&apos;</span>:sc_color[<span class="hljs-number">14</span>],
    <span class="hljs-string">&apos;activated B cells&apos;</span>:sc_color[<span class="hljs-number">15</span>],
    <span class="hljs-string">&apos;IgA+ Plasma B cells&apos;</span>:sc_color[<span class="hljs-number">8</span>],
    <span class="hljs-string">&apos;IgG+ Plasma B cells&apos;</span>:sc_color[<span class="hljs-number">12</span>],
    <span class="hljs-string">&apos;Plasma B cells&apos;</span>:sc_color[<span class="hljs-number">12</span>],
    <span class="hljs-string">&apos;B-cell&apos;</span>:sc_color[<span class="hljs-number">13</span>],
    <span class="hljs-string">&apos;B-cell (MALAT1+)&apos;</span>:<span class="hljs-string">&apos;#a3007d&apos;</span>,

}
adata.uns[<span class="hljs-string">&apos;B_celltype_colors&apos;</span>]=adata.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>].cat.categories.map(type_color_B).values.tolist()

<span class="hljs-comment">#&#x7ED8;&#x5236;&#x6548;&#x679C;</span>
ax=scv.pl.velocity_embedding_stream(adata, basis=<span class="hljs-string">&apos;umap&apos;</span>,color=[<span class="hljs-string">&apos;major_celltype&apos;</span>,<span class="hljs-string">&apos;B_celltype&apos;</span>,<span class="hljs-string">&apos;Type&apos;</span>],
                                legend_align_text=<span class="hljs-string">&apos;y&apos;</span>,show=<span class="hljs-keyword">False</span>,)
plt.savefig(<span class="hljs-string">&quot;prop/cell_type.png&quot;</span>,dpi=<span class="hljs-number">300</span>,bbox_inches = <span class="hljs-string">&apos;tight&apos;</span>)
</code></pre>
<p><img src="book-3.assets/image-20221019005115998-6119977.png" alt="image-20221019005115998" style="zoom:50%;"></p>
<p>&#x7136;&#x540E;&#xFF0C;&#x4E3A;&#x4E86;&#x5206;&#x6790;&#x9634;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x8DDF;&#x9633;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x7684;&#x7EC6;&#x80DE;&#x6BD4;&#x4F8B;&#x5DEE;&#x5F02;&#xFF0C;&#x6211;&#x4EEC;&#x8BA1;&#x7B97;&#x4E86;&#x4E24;&#x79CD;&#x6DCB;&#x5DF4;&#x7ED3;&#x4E2D;&#x7684;&#x7EC6;&#x80DE;&#x6BD4;&#x4F8B;</p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x9633;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;</span>
<span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np
<span class="hljs-keyword">import</span> matplotlib.pyplot <span class="hljs-keyword">as</span> plt
<span class="hljs-keyword">from</span> pylab <span class="hljs-keyword">import</span> *
<span class="hljs-comment">#mpl.rcParams[&apos;font.sans-serif&apos;]=[&apos;SimHei&apos;]</span>
<span class="hljs-comment">#mpl.rcParams[&apos;axes.unicode_minus&apos;]=False</span>
labels=(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>].obs[<span class="hljs-string">&apos;B_celltype&apos;</span>].value_counts()/len(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>].obs)).index.tolist()
data=(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>].obs[<span class="hljs-string">&apos;B_celltype&apos;</span>].value_counts()/len(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>].obs)).values
b_color=pd.DataFrame(labels)[<span class="hljs-number">0</span>].map(type_color_B).values
explodes=[i <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> [<span class="hljs-number">0.1</span>]*len(labels)]

fig, ax = plt.subplots(figsize=(<span class="hljs-number">2</span>,<span class="hljs-number">2</span>)) 
<span class="hljs-comment">#plt.axes(aspect=0.5)</span>
plt.pie(x=data,labels=labels,autopct=<span class="hljs-string">&quot;%0.1f%%&quot;</span>,labeldistance=<span class="hljs-keyword">None</span>,
        explode=explodes,shadow=<span class="hljs-keyword">True</span>,colors=b_color)
plt.title(<span class="hljs-string">&apos;Pos Lym&apos;</span>)
plt.savefig(<span class="hljs-string">&quot;prop/pos_prop.png&quot;</span>,dpi=<span class="hljs-number">300</span>,bbox_inches = <span class="hljs-string">&apos;tight&apos;</span>)
</code></pre>
<p><img src="book-3.assets/image-20221019005254018-6119981.png" alt="image-20221019005254018" style="zoom:50%;"></p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x9634;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;</span>
<span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np
<span class="hljs-keyword">import</span> matplotlib.pyplot <span class="hljs-keyword">as</span> plt
<span class="hljs-keyword">from</span> pylab <span class="hljs-keyword">import</span> *
labels=(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Neg&apos;</span>].obs[<span class="hljs-string">&apos;B_celltype&apos;</span>].value_counts()/len(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Neg&apos;</span>].obs)).index.tolist()
<span class="hljs-comment">#labels=[i for i in [&apos; &apos;]*5]</span>
data=(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Neg&apos;</span>].obs[<span class="hljs-string">&apos;B_celltype&apos;</span>].value_counts()/len(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Neg&apos;</span>].obs)).values
b_color=pd.DataFrame(labels)[<span class="hljs-number">0</span>].map(type_color_B).values
explodes=[i <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> [<span class="hljs-number">0.1</span>]*len(labels)]

fig, ax = plt.subplots(figsize=(<span class="hljs-number">2</span>,<span class="hljs-number">2</span>)) 
<span class="hljs-comment">#plt.axes(aspect=0.5)</span>
plt.pie(x=data,labels=labels,autopct=<span class="hljs-string">&quot;%0.1f%%&quot;</span>,labeldistance=<span class="hljs-keyword">None</span>,
        explode=explodes,shadow=<span class="hljs-keyword">True</span>,colors=b_color)
plt.legend(bbox_to_anchor=(<span class="hljs-number">1.8</span>, <span class="hljs-number">0.7</span>), ncol=<span class="hljs-number">3</span>,fontsize=<span class="hljs-number">10</span>)
plt.title(<span class="hljs-string">&apos;Neg Lym&apos;</span>)
plt.grid(<span class="hljs-keyword">False</span>)
plt.axis(<span class="hljs-string">&apos;off&apos;</span>)
plt.savefig(<span class="hljs-string">&quot;prop/neg_prop.png&quot;</span>,dpi=<span class="hljs-number">300</span>,bbox_inches = <span class="hljs-string">&apos;tight&apos;</span>)
</code></pre>
<p><img src="book-3.assets/image-20221019005328419-6119984.png" alt="image-20221019005328419" style="zoom:50%;"></p>
<p>&#x9664;&#x4E86;&#x4F7F;&#x7528;&#x6247;&#x5F62;&#x56FE;&#x5916;&#xFF0C;&#x6211;&#x4EEC;&#x8FD8;&#x80FD;&#x7528;&#x67F1;&#x72B6;&#x56FE;&#x66F4;&#x52A0;&#x76F4;&#x89C2;&#x7684;&#x611F;&#x53D7;&#x7EC6;&#x80DE;&#x7C7B;&#x578B;&#x7684;&#x6BD4;&#x4F8B;&#xFF0C;&#x6211;&#x4EEC;&#x5148;&#x6784;&#x5EFA;&#x4E00;&#x4E2A;pandas&#x8868;&#x683C;&#x7528;&#x4E8E;&#x7ED8;&#x5236;&#x67F1;&#x72B6;&#x56FE;</p>
<pre><code class="lang-python">prop_pd1=pd.DataFrame(index=list(set(adata.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>])))
prop_pd1[<span class="hljs-string">&apos;Cell&apos;</span>]=prop_pd1.index.tolist()
prop_pd1[<span class="hljs-string">&apos;Values&apos;</span>]=<span class="hljs-number">0</span>
prop_pd1[<span class="hljs-string">&apos;Type&apos;</span>]=<span class="hljs-string">&apos;Neg&apos;</span>
data=(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Neg&apos;</span>].obs[<span class="hljs-string">&apos;B_celltype&apos;</span>].value_counts()/len(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Neg&apos;</span>].obs))
prop_pd1.loc[data.index.tolist(),<span class="hljs-string">&apos;Values&apos;</span>]=[round(i,<span class="hljs-number">2</span>) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> data.values]


prop_pd2=pd.DataFrame(index=list(set(adata.obs[<span class="hljs-string">&apos;B_celltype&apos;</span>])))
prop_pd2[<span class="hljs-string">&apos;Cell&apos;</span>]=prop_pd2.index.tolist()
prop_pd2[<span class="hljs-string">&apos;Values&apos;</span>]=<span class="hljs-number">0</span>
prop_pd2[<span class="hljs-string">&apos;Type&apos;</span>]=<span class="hljs-string">&apos;Pos&apos;</span>
data=(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>].obs[<span class="hljs-string">&apos;B_celltype&apos;</span>].value_counts()/len(adata[adata.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>].obs))
prop_pd2.loc[data.index.tolist(),<span class="hljs-string">&apos;Values&apos;</span>]=[round(i,<span class="hljs-number">2</span>) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> data.values]

prop_pd=pd.concat([prop_pd1,prop_pd2])
prop_pd.index=range(len(prop_pd))
prop_pd
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right"></th>
<th style="text-align:right">Cell</th>
<th style="text-align:right">Values</th>
<th style="text-align:right">Type</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">0</td>
<td style="text-align:right">B-cell (MALAT1+)</td>
<td style="text-align:right">0.13</td>
<td style="text-align:right">Neg</td>
</tr>
<tr>
<td style="text-align:right">1</td>
<td style="text-align:right">Pre-B cells</td>
<td style="text-align:right">0.27</td>
<td style="text-align:right">Neg</td>
</tr>
<tr>
<td style="text-align:right">2</td>
<td style="text-align:right">nai&#x308;ve B cells</td>
<td style="text-align:right">0.21</td>
<td style="text-align:right">Neg</td>
</tr>
<tr>
<td style="text-align:right">3</td>
<td style="text-align:right">Plasma B cells</td>
<td style="text-align:right">0.01</td>
<td style="text-align:right">Neg</td>
</tr>
<tr>
<td style="text-align:right">4</td>
<td style="text-align:right">activated B cells</td>
<td style="text-align:right">0.38</td>
<td style="text-align:right">Neg</td>
</tr>
<tr>
<td style="text-align:right">5</td>
<td style="text-align:right">B-cell (MALAT1+)</td>
<td style="text-align:right">0.06</td>
<td style="text-align:right">Pos</td>
</tr>
<tr>
<td style="text-align:right">6</td>
<td style="text-align:right">Pre-B cells</td>
<td style="text-align:right">0.37</td>
<td style="text-align:right">Pos</td>
</tr>
<tr>
<td style="text-align:right">7</td>
<td style="text-align:right">nai&#x308;ve B cells</td>
<td style="text-align:right">0.40</td>
<td style="text-align:right">Pos</td>
</tr>
<tr>
<td style="text-align:right">8</td>
<td style="text-align:right">Plasma B cells</td>
<td style="text-align:right">0.00</td>
<td style="text-align:right">Pos</td>
</tr>
<tr>
<td style="text-align:right">9</td>
<td style="text-align:right">activated B cells</td>
<td style="text-align:right">0.16</td>
<td style="text-align:right">Pos</td>
</tr>
</tbody>
</table>
<p>&#x968F;&#x540E;&#x6211;&#x4EEC;&#x5206;&#x522B;&#x4E3A;Neg&#x4E0E;Pos&#x6307;&#x5B9A;&#x989C;&#x8272;&#x540E;&#x7ED8;&#x5236;&#x67F1;&#x72B6;&#x56FE;</p>
<pre><code class="lang-python">type_color={
    <span class="hljs-string">&apos;Pos&apos;</span>:<span class="hljs-string">&apos;#9B7170&apos;</span>,
    <span class="hljs-string">&apos;Neg&apos;</span>:<span class="hljs-string">&apos;#C65A50&apos;</span>
}

fig, ax = plt.subplots(figsize=(<span class="hljs-number">2</span>,<span class="hljs-number">4</span>)) 

sns.barplot(
    data=prop_pd, 
    x=<span class="hljs-string">&quot;Values&quot;</span>, y=<span class="hljs-string">&quot;Cell&quot;</span>, hue=<span class="hljs-string">&quot;Type&quot;</span>,
    alpha=<span class="hljs-number">.6</span>,palette=[type_color[<span class="hljs-string">&apos;Neg&apos;</span>],type_color[<span class="hljs-string">&apos;Pos&apos;</span>]]
)
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> ax.containers:
    ax.bar_label(i,)
sns.move_legend(ax, <span class="hljs-string">&quot;upper left&quot;</span>, bbox_to_anchor=(<span class="hljs-number">1</span>, <span class="hljs-number">1</span>))

plt.savefig(<span class="hljs-string">&quot;prop/prop.png&quot;</span>,dpi=<span class="hljs-number">300</span>,bbox_inches = <span class="hljs-string">&apos;tight&apos;</span>)
</code></pre>
<p><img src="book-3.assets/image-20221019005554337-6119988.png" alt="image-20221019005554337" style="zoom:50%;"></p>
<p>&#x9664;&#x6B64;&#x4E4B;&#x5916;&#xFF0C;&#x6211;&#x4EEC;&#x8FD8;&#x53EF;&#x4EE5;&#x5206;&#x6790;&#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;&#x7684;&#x6F14;&#x53D8;&#x8FC7;&#x7A0B;</p>
<pre><code class="lang-python">scv.tl.paga(adata, groups=<span class="hljs-string">&apos;B_celltype&apos;</span>,use_time_prior=<span class="hljs-keyword">False</span>)
scv.pl.paga(adata, basis=<span class="hljs-string">&apos;umap&apos;</span>, size=<span class="hljs-number">50</span>, alpha=<span class="hljs-number">.1</span>,
            min_edge_width=<span class="hljs-number">2</span>, node_size_scale=<span class="hljs-number">1.5</span>)
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">running PAGA 
    finished (0:00:00) --&gt; added
    &apos;paga/connectivities&apos;, connectivities adjacency (adata.uns)
    &apos;paga/connectivities_tree&apos;, connectivities subtree (adata.uns)
    &apos;paga/transitions_confidence&apos;, velocity transitions (adata.uns)
</code></pre>
<p><img src="book-3.assets/image-20221019005740026-6119990.png" alt="image-20221019005740026" style="zoom:50%;"></p>
<p>&#x5230;&#x6B64;&#xFF0C;&#x7EC6;&#x80DE;&#x4E9A;&#x7FA4;&#x6CE8;&#x91CA;&#x53CA;&#x76F8;&#x5173;&#x7684;&#x56FE;&#x8868;&#xFF0C;&#x6211;&#x4EEC;&#x5C31;&#x7ED8;&#x5236;&#x5B8C;&#x6210;&#x4E86;&#xFF0C;&#x6211;&#x4EEC;&#x4FDD;&#x5B58;&#x4E00;&#x4E0B;</p>
<pre><code class="lang-python">adata.write_h5ad(<span class="hljs-string">&apos;B_cell_anno_new.h5ad&apos;</span>,compression=<span class="hljs-string">&apos;gzip&apos;</span>)
</code></pre>
<footer class="page-footer"><span class="copyright">&#x7248;&#x6743;&#x6240;&#x6709; &#xA9; Starlitnightly</span><span class="footer-modification">&#x6700;&#x540E;&#x4FEE;&#x8BA2;&#x65F6;&#x95F4;&#xFF1A;
2022-10-17
</span></footer>
						
					</section>
					
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

				</div>
			</div>
			
		</div>

		
		
		<a href="book-2.html" title="上一页" class="navigation navigation-prev " aria-label="Previous page: 6-2. GLUE多组学整合">
			<i class="fa fa-angle-left"></i>
		</a>
		
		
		<a href="book-4.html" title="下一页" class="navigation navigation-next " aria-label="Next page: 6-4. 差异表达分析">
			<i class="fa fa-angle-right"></i>
		</a>
		
		
		
	</div>
	<script>
		var gitbook = gitbook || [];
		gitbook.push(function() {
			gitbook.page.hasChanged({"page":{"title":"6-3. 细胞亚群分析","level":"1.9.3","depth":2,"next":{"title":"6-4. 差异表达分析","level":"1.9.4","depth":2,"path":"MultiOMIC/book-4.md","ref":"MultiOMIC/book-4.md","articles":[]},"previous":{"title":"6-2. GLUE多组学整合","level":"1.9.2","depth":2,"path":"MultiOMIC/book-2.md","ref":"MultiOMIC/book-2.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{"themeIvoryhza":{"nav":[{"url":"http://...","target":"_blank","name":"HOME"}]}},"plugins":["expandable-chapters","-sharing","-fontsettings","back-to-top-button","copy-code-button","cuav-chapters","heading-anchors","theme-ivoryhza"],"pluginsConfig":{"search":{},"theme-ivoryhza":{"search-placeholder":"搜索","logo":"./logo.png","appleTouchIconPrecomposed152":"./logo.png","copyright":"版权所有 © Starlitnightly","modify_label":"最后修订时间：","modify_format":"2022-10-17"},"cuav-chapters":{"summaryMaxSize":20,"useLimitExpanded":false},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"heading-anchors":{},"highlight":{},"back-to-top-button":{},"copy-code-button":{},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false},"expandable-chapters":{}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"MultiOMIC/book-3.md","mtime":"2022-10-18T19:06:32.073Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2022-11-22T15:18:44.125Z"},"basePath":"..","book":{"language":""}});
		});
	</script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-back-to-top-button/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-copy-code-button/toggle.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-cuav-chapters/cuav-chapters.js"></script>
        
    
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/1.2.1/anchor.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-heading-anchors/anchor-style.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-theme-ivoryhza/ivory.js"></script>
        
    

    </body>
</html>

