
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>6-5. 细胞通讯分析 · GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-back-to-top-button/plugin.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-cuav-chapters/cuav-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-theme-ivoryhza/ivory.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-theme-ivoryhza/footer.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="book-6.html" />
    
    
    <link rel="prev" href="book-4.html" />
    

    </head>
    <body>
        
<div class="book">
	<div class="header-inner">
		<!-- LOGO -->
		<div class="header-inner-left">
			<div class="logo"></div>
			<span class="title"></span>
		</div>

		<!-- Nav -->
		<div class="header-inner-right">
			<ul class="header-nav">
				<li>
					<a href="http://..." target="_blank">HOME</a>
				</li>
			</ul>
			<!-- Search -->
			
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

		</div>


	</div>

	<div class="book-summary">
		<div class="book-summary-title">目录</div>
		
		
		<nav role="navigation">
			


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Bioinformatics Tutorial
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../">
            
                <a href="../">
            
                    
                    目录
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../CONFIG/">
            
                <a href="../CONFIG/">
            
                    
                    配置：基础环境配置
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../CONFIG/config_1.html">
            
                <a href="../CONFIG/config_1.html">
            
                    
                    配置1：Python环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../CONFIG/config_2.html">
            
                <a href="../CONFIG/config_2.html">
            
                    
                    配置2：Linux环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../CONFIG/config_3.html">
            
                <a href="../CONFIG/config_3.html">
            
                    
                    配置3：R语言环境
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../RNASEQUP/">
            
                <a href="../RNASEQUP/">
            
                    
                    分析1：RNA-seq上游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../RNASEQUP/rnasequp_1.html">
            
                <a href="../RNASEQUP/rnasequp_1.html">
            
                    
                    1-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../RNASEQUP/rnasequp_2.html">
            
                <a href="../RNASEQUP/rnasequp_2.html">
            
                    
                    1-2. 上游分析全教程
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../RNASEQ/">
            
                <a href="../RNASEQ/">
            
                    
                    分析2：RNA-seq下游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="../RNASEQ/rnaseq_1.html">
            
                <a href="../RNASEQ/rnaseq_1.html">
            
                    
                    2-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="../RNASEQ/rnaseq_2.html">
            
                <a href="../RNASEQ/rnaseq_2.html">
            
                    
                    2-2. 差异表达基因(DEG)分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="../RNASEQ/rnaseq_3.html">
            
                <a href="../RNASEQ/rnaseq_3.html">
            
                    
                    2-3. 通路富集分析(GSEA)
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../SCGLUE/">
            
                <a href="../SCGLUE/">
            
                    
                    分析3：单细胞样本对齐
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" data-path="../SCGLUE/scglue_1.html">
            
                <a href="../SCGLUE/scglue_1.html">
            
                    
                    3-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.2" data-path="../SCGLUE/scglue_2.html">
            
                <a href="../SCGLUE/scglue_2.html">
            
                    
                    3-2. 数据预处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.3" data-path="../SCGLUE/scglue_3.html">
            
                <a href="../SCGLUE/scglue_3.html">
            
                    
                    3-3. 模型准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.4" data-path="../SCGLUE/scglue_4.html">
            
                <a href="../SCGLUE/scglue_4.html">
            
                    
                    3-4. 模型训练
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.5" data-path="../SCGLUE/scglue_5.html">
            
                <a href="../SCGLUE/scglue_5.html">
            
                    
                    3-5. 对齐细胞
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="../MOFA/">
            
                <a href="../MOFA/">
            
                    
                    分析4：MOFA单细胞多组学因子分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.1" data-path="../MOFA/mofa_1.html">
            
                <a href="../MOFA/mofa_1.html">
            
                    
                    4-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2" data-path="../MOFA/mofa_2.html">
            
                <a href="../MOFA/mofa_2.html">
            
                    
                    4-2. 单细胞样本对齐
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.3" data-path="../MOFA/mofa_3.html">
            
                <a href="../MOFA/mofa_3.html">
            
                    
                    4-3. scRNA-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.4" data-path="../MOFA/mofa_4.html">
            
                <a href="../MOFA/mofa_4.html">
            
                    
                    4-4. scATAC-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.5" data-path="../MOFA/mofa_5.html">
            
                <a href="../MOFA/mofa_5.html">
            
                    
                    4-5. MOFA模型构建
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.6" data-path="../MOFA/mofa_6.html">
            
                <a href="../MOFA/mofa_6.html">
            
                    
                    4-6. MOFA下游分析-R语言环节
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.7" data-path="../MOFA/mofa_7.html">
            
                <a href="../MOFA/mofa_7.html">
            
                    
                    4-7. MOFA下游分析-Python环节
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.8" data-path="../TMT/">
            
                <a href="../TMT/">
            
                    
                    分析5：TMT蛋白组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.8.1" data-path="../TMT/tmt_1.html">
            
                <a href="../TMT/tmt_1.html">
            
                    
                    5-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8.2" data-path="../TMT/tmt_2.html">
            
                <a href="../TMT/tmt_2.html">
            
                    
                    5-2. TMT蛋白组学上游分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8.3" data-path="../TMT/tmt_3.html">
            
                <a href="../TMT/tmt_3.html">
            
                    
                    5-3. TMT蛋白组学下游分析
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.9" data-path="./">
            
                <a href="./">
            
                    
                    分析6：多组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.9.1" data-path="book-1.html">
            
                <a href="book-1.html">
            
                    
                    6-1. 数据准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.2" data-path="book-2.html">
            
                <a href="book-2.html">
            
                    
                    6-2. GLUE多组学整合
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.3" data-path="book-3.html">
            
                <a href="book-3.html">
            
                    
                    6-3. 细胞亚群分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.4" data-path="book-4.html">
            
                <a href="book-4.html">
            
                    
                    6-4. 差异表达分析
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.9.5" data-path="book-5.html">
            
                <a href="book-5.html">
            
                    
                    6-5. 细胞通讯分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.6" data-path="book-6.html">
            
                <a href="book-6.html">
            
                    
                    6-6. 动态调控分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.7" data-path="book-7.html">
            
                <a href="book-7.html">
            
                    
                    6-7. 转录因子网络
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.10" data-path="../PLOT/">
            
                <a href="../PLOT/">
            
                    
                    绘图：Python可视化
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.10.1" data-path="../PLOT/plot_1.html">
            
                <a href="../PLOT/plot_1.html">
            
                    
                    1-1. 火山图
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.10.2" data-path="../PLOT/plot_2.html">
            
                <a href="../PLOT/plot_2.html">
            
                    
                    1-2. 箱线图
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


		</nav>
		
		
	</div>

	<!-- Content nav -->
	<div class="book-anchor">
		<div class="book-anchor-title">大纲</div>
		<div class="book-anchor-body">

		</div>
	</div>

	<div class="book-body">
		
		<div class="body-inner">
			
			

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >6-5. 细胞通讯分析</a>
    </h1>
</div>




			<div class="page-wrapper" tabindex="-1" role="main">
				<div class="page-inner">
					
<div id="book-search-results">
    <div class="search-noresults">
    
					<section class="normal markdown-section">
						
						<h3 id="43-&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x5206;&#x6790;">4.3 &#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x5206;&#x6790;</h3>
<p>&#x5728;&#x5206;&#x6790;&#x5B8C;&#x9633;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x8DDF;&#x9634;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x7684;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x60C5;&#x51B5;&#x540E;&#xFF0C;&#x6211;&#x4EEC;&#x8FD8;&#x6709;&#x4E00;&#x79CD;&#x5F88;&#x76F4;&#x89C2;&#x7684;&#x601D;&#x8DEF;&#x5C31;&#x662F;&#xFF0C;&#x6211;&#x4EEC;&#x53EF;&#x4EE5;&#x6BD4;&#x8F83;&#x9633;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x8DDF;&#x9634;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x4E2D;&#xFF0C;&#x8FD9;5&#x79CD;&#x4E0D;&#x540C;&#x7684;B&#x7EC6;&#x80DE;&#x4E4B;&#x95F4;&#x7684;&#x901A;&#x8BAF;&#xFF0C;&#x662F;&#x5426;&#x4E5F;&#x53D1;&#x751F;&#x4E86;&#x76F8;&#x5E94;&#x7684;&#x53D8;&#x5316;&#x3002;</p>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x7528;&#x5230;&#x7684;&#x4E00;&#x4E2A;&#x65B0;&#x7684;&#x5206;&#x6790;&#x65B9;&#x6CD5;&#xFF0C;&#x53EB;&#x505A;CellphoneDB&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x5206;&#x6790;&#xFF0C;&#x987E;&#x540D;&#x601D;&#x4E49;&#xFF0C;&#x7EC6;&#x80DE;&#x4E4B;&#x95F4;&#x6253;&#x7535;&#x8BDD;&#xFF0C;&#x76F8;&#x5173;&#x5305;&#x7684;&#x4ECB;&#x7ECD;&#x6211;&#x4EEC;&#x53EF;&#x4EE5;&#x53C2;&#x7167;&#x5B98;&#x65B9;&#x8BF4;&#x660E;&#x6587;&#x6863;&#xFF0C;&#x8FD9;&#x91CC;&#x4E0D;&#x505A;&#x8FC7;&#x591A;&#x89E3;&#x91CA;&#xFF0C;&#x90A3;&#x4E48;&#xFF0C;&#x5728;&#x6211;&#x4EEC;&#x7684;&#x6DCB;&#x5DF4;&#x7ED3;&#x4E2D;&#xFF0C;&#x8BE5;&#x5982;&#x4F55;&#x5E94;&#x7528;&#x8FD9;&#x4E2A;&#x5206;&#x6790;&#x5462;&#xFF1F;</p>
<h4 id="431-&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x5206;&#x6790;&#x6570;&#x636E;&#x51C6;&#x5907;">4.3.1 &#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x5206;&#x6790;&#x6570;&#x636E;&#x51C6;&#x5907;</h4>
<p>&#x6211;&#x4EEC;&#x9996;&#x5148;&#x8FD8;&#x662F;&#x5BFC;&#x5165;&#x5305;&#x4EE5;&#x53CA;B&#x7EC6;&#x80DE;&#x6570;&#x636E;</p>
<pre><code class="lang-python"><span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np
<span class="hljs-keyword">import</span> pandas <span class="hljs-keyword">as</span> pd
<span class="hljs-keyword">import</span> scanpy <span class="hljs-keyword">as</span> sc
<span class="hljs-keyword">import</span> anndata
<span class="hljs-keyword">import</span> os
<span class="hljs-keyword">import</span> sys
<span class="hljs-keyword">from</span> scipy <span class="hljs-keyword">import</span> sparse
sc.settings.verbosity = <span class="hljs-number">1</span>  <span class="hljs-comment"># verbosity: errors (0), warnings (1), info (2), hints (3)</span>

adata=sc.read(<span class="hljs-string">&apos;B_cell_anno_new.h5ad&apos;</span>)
adata
</code></pre>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x5C06;&#x8981;&#x7528;&#x5230;&#x7684;&#x662F;&#x539F;&#x59CB;&#x8868;&#x8FBE;&#x77E9;&#x9635;&#xFF0C;&#x4E5F;&#x5C31;&#x662F;&#x6CA1;&#x6709;&#x5F52;&#x4E00;&#x5316;&#x7684;&#x6570;&#x636E;&#xFF0C;&#x6240;&#x4EE5;&#x6211;&#x4EEC;&#x91C7;&#x7528;GLUE&#x5206;&#x6790;&#x4E2D;&#x4EA7;&#x751F;&#x7684;&#x4E2D;&#x95F4;&#x6587;&#x4EF6;<code>rna-emb.h5ad</code>&#x6765;&#x505A;&#x5206;&#x6790;&#xFF0C;&#x9996;&#x5148;&#x662F;B&#x7EC6;&#x80DE;&#x4E2D;&#x7684;&#x9633;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;</p>
<pre><code class="lang-python">adata_raw=sc.read(<span class="hljs-string">&apos;../glue_true/rna-emb.h5ad&apos;</span>)
adata_raw=adata_raw.raw.to_adata()[adata.obs.index]
adata_raw=adata_raw[adata_raw.obs[<span class="hljs-string">&apos;Type&apos;</span>]==<span class="hljs-string">&apos;Pos&apos;</span>]

sc.pp.filter_cells(adata_raw, min_genes=<span class="hljs-number">200</span>)
sc.pp.filter_genes(adata_raw, min_cells=<span class="hljs-number">3</span>)
sc.pp.normalize_total(adata_raw)
adata_raw=anndata.AnnData(adata_raw.X,obs=pd.DataFrame(index=adata_raw.obs.index),
                          var=pd.DataFrame(index=adata_raw.var.index))
adata_raw.write_h5ad(<span class="hljs-string">&apos;Bcell_pos_raw.h5ad&apos;</span>,compression=<span class="hljs-string">&apos;gzip&apos;</span>)
adata_raw.to_df().T.to_csv(<span class="hljs-string">&apos;Bcell_pos_counts.tsv&apos;</span>,sep=<span class="hljs-string">&apos;\t&apos;</span>)


df_meta = pd.DataFrame(data={<span class="hljs-string">&apos;Cell&apos;</span>:list(adata[adata_raw.obs.index].obs.index),
                             <span class="hljs-string">&apos;cell_type&apos;</span>:[ i <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> adata[adata_raw.obs.index].obs[<span class="hljs-string">&apos;B_celltype&apos;</span>]]
                            })
df_meta.set_index(<span class="hljs-string">&apos;Cell&apos;</span>, inplace=<span class="hljs-keyword">True</span>)
df_meta.to_csv(<span class="hljs-string">&apos;Bcell_pos_meta.tsv&apos;</span>, sep = <span class="hljs-string">&apos;\t&apos;</span>)

adata_raw
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">AnnData object with n_obs &#xD7; n_vars = 2684 &#xD7; 12951
</code></pre>
<p>&#x901A;&#x8FC7;&#x8FD0;&#x884C;&#x8BE5;&#x4EE3;&#x7801;&#xFF0C;&#x6211;&#x4EEC;&#x4E00;&#x5171;&#x5F97;&#x5230;&#x4E86;<code>Bcell_pos_counts.tsv</code>&#x4E0E;<code>Bcell_pos_meta.tsv</code>&#x4E24;&#x4E2A;&#x6587;&#x4EF6;&#xFF0C;&#x540C;&#x7406;&#x6211;&#x4EEC;&#x5BF9;&#x9634;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;&#x8FDB;&#x884C;&#x76F8;&#x540C;&#x7684;&#x6570;&#x636E;&#x51C6;&#x5907;&#xFF0C;&#x5C06;&#x5F97;&#x5230;<code>Bcell_neg_counts.tsv</code>&#x4E0E;<code>Bcell_neg_meta.tsv</code>&#x4E24;&#x4E2A;&#x6587;&#x4EF6;&#x3002;</p>
<h4 id="432-&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x5206;&#x6790;">4.3.2 &#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x5206;&#x6790;</h4>
<p>&#x6211;&#x4EEC;&#x5728;&#x7EC8;&#x7AEF;<code>terminal</code>&#x4E2D;&#x8F93;&#x5165;&#x4EE5;&#x4E0B;&#x4EE3;&#x7801;&#xFF0C;&#x7B49;&#x5F85;1&#x4E2A;&#x5C0F;&#x65F6;&#x5DE6;&#x53F3;&#x4FBF;&#x80FD;&#x5F97;&#x5230;&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x5206;&#x6790;&#x7684;&#x8FD0;&#x884C;&#x7ED3;&#x679C;&#x4E86;</p>
<pre><code class="lang-shell">cellphonedb method statistical_analysis /content/Bcell_neg_meta.tsv \
/content/Bcell_neg_counts.tsv --output-path /content/Bcell_neg --counts-data hgnc_symbol

cellphonedb method statistical_analysis /content/Bcell_pos_meta.tsv \
/content/Bcell_pos_counts.tsv --output-path /content/Bcell_pos --counts-data hgnc_symbol
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">[ ][CORE][09/10/22-18:07:53][INFO] Initializing SqlAlchemy CellPhoneDB Core
[ ][CORE][09/10/22-18:07:53][INFO] Using custom database at /root/.cpdb/releases/v4.0.0/cellphone.db
[ ][APP][09/10/22-18:07:53][INFO] Launching Method cpdb_statistical_analysis_local_method_launcher
[ ][APP][09/10/22-18:07:53][INFO] Launching Method _set_paths
[ ][APP][09/10/22-18:07:53][INFO] Launching Method _load_meta_counts
[ ][APP][09/10/22-18:08:00][INFO] Launching Method _check_counts_data
[ ][CORE][09/10/22-18:08:00][INFO] Launching Method cpdb_statistical_analysis_launcher
[ ][CORE][09/10/22-18:08:00][INFO] Launching Method _counts_validations
[ ][CORE][09/10/22-18:08:00][INFO] Launching Method get_interactions_genes_complex
[ ][CORE][09/10/22-18:08:00][INFO] [Cluster Statistical Analysis] Threshold:0.1 Iterations:1000 Debug-seed:-1 Threads:4 Precision:3
[ ][CORE][09/10/22-18:08:02][INFO] Running Real Analysis
[ ][CORE][09/10/22-18:08:02][INFO] Running Statistical Analysis
[ ][CORE][09/10/22-18:39:47][INFO] Building Pvalues result
[ ][CORE][09/10/22-18:39:48][INFO] Building results
[ ][CORE][09/10/22-18:39:51][INFO] Initializing SqlAlchemy CellPhoneDB Core
[ ][CORE][09/10/22-18:39:51][INFO] Using custom database at /root/.cpdb/releases/v4.0.0/cellphone.db
[ ][APP][09/10/22-18:39:51][INFO] Launching Method cpdb_statistical_analysis_local_method_launcher
[ ][APP][09/10/22-18:39:51][INFO] Launching Method _set_paths
[ ][APP][09/10/22-18:39:51][INFO] Launching Method _load_meta_counts
[ ][APP][09/10/22-18:39:57][INFO] Launching Method _check_counts_data
[ ][CORE][09/10/22-18:39:57][INFO] Launching Method cpdb_statistical_analysis_launcher
[ ][CORE][09/10/22-18:39:57][INFO] Launching Method _counts_validations
[ ][CORE][09/10/22-18:39:57][INFO] Launching Method get_interactions_genes_complex
[ ][CORE][09/10/22-18:39:57][INFO] [Cluster Statistical Analysis] Threshold:0.1 Iterations:1000 Debug-seed:-1 Threads:4 Precision:3
[ ][CORE][09/10/22-18:39:59][INFO] Running Real Analysis
[ ][CORE][09/10/22-18:39:59][INFO] Running Statistical Analysis
[ ][CORE][09/10/22-19:06:40][INFO] Building Pvalues result
[ ][CORE][09/10/22-19:06:40][INFO] Building results
</code></pre>
<p>&#x6211;&#x4EEC;&#x5C06;&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x7684;&#x7ED3;&#x679C;&#x5206;&#x522B;&#x5B58;&#x653E;&#x5230;&#x4E86;<code>Bcell_pos</code>,<code>Bcell_neg</code>&#x4E24;&#x4E2A;&#x6587;&#x4EF6;&#x5939;&#x4E2D;&#xFF0C;&#x6211;&#x4EEC;&#x8FD9;&#x91CC;&#x6F14;&#x793A;&#x5982;&#x4F55;&#x5BF9;&#x9633;&#x6027;&#x6DCB;&#x5DF4;&#x7ED3;(Pos)&#x8FDB;&#x884C;&#x5206;&#x6790;</p>
<h4 id="433-&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x53EF;&#x89C6;&#x5316;-&#x7EC6;&#x80DE;&#x5173;&#x7CFB;">4.3.3 &#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x53EF;&#x89C6;&#x5316;-&#x7EC6;&#x80DE;&#x5173;&#x7CFB;</h4>
<p>&#x9996;&#x5148;&#x4F9D;&#x7136;&#x662F;&#x5BFC;&#x5165;&#x57FA;&#x7840;&#x7684;&#x5305;</p>
<pre><code class="lang-python"><span class="hljs-keyword">import</span> pandas <span class="hljs-keyword">as</span> pd
<span class="hljs-keyword">import</span> numpy <span class="hljs-keyword">as</span> np
<span class="hljs-keyword">import</span> networkx <span class="hljs-keyword">as</span> nx
<span class="hljs-keyword">import</span> matplotlib.pyplot <span class="hljs-keyword">as</span> plt
<span class="hljs-keyword">import</span> seaborn <span class="hljs-keyword">as</span> sns
<span class="hljs-keyword">import</span> scanpy <span class="hljs-keyword">as</span> sc

sc.logging.print_header()
sc.settings.set_figure_params(dpi=<span class="hljs-number">80</span>, facecolor=<span class="hljs-string">&apos;white&apos;</span>)

sc_color=[<span class="hljs-string">&apos;#7CBB5F&apos;</span>,<span class="hljs-string">&apos;#368650&apos;</span>,<span class="hljs-string">&apos;#A499CC&apos;</span>,<span class="hljs-string">&apos;#5E4D9A&apos;</span>,<span class="hljs-string">&apos;#78C2ED&apos;</span>,<span class="hljs-string">&apos;#866017&apos;</span>,<span class="hljs-string">&apos;#9F987F&apos;</span>, <span class="hljs-string">&apos;#E0DFED&apos;</span>, <span class="hljs-string">&apos;#EF7B77&apos;</span>, <span class="hljs-string">&apos;#279AD7&apos;</span>,
 <span class="hljs-string">&apos;#F0EEF0&apos;</span>, <span class="hljs-string">&apos;#1F577B&apos;</span>, <span class="hljs-string">&apos;#A56BA7&apos;</span>, <span class="hljs-string">&apos;#E0A7C8&apos;</span>, <span class="hljs-string">&apos;#E069A6&apos;</span>, <span class="hljs-string">&apos;#941456&apos;</span>, <span class="hljs-string">&apos;#FCBC10&apos;</span>, <span class="hljs-string">&apos;#EAEFC5&apos;</span>, <span class="hljs-string">&apos;#01A0A7&apos;</span>, <span class="hljs-string">&apos;#75C8CC&apos;</span>, 
<span class="hljs-string">&apos;#F0D7BC&apos;</span>, <span class="hljs-string">&apos;#D5B26C&apos;</span>, <span class="hljs-string">&apos;#D5DA48&apos;</span>, <span class="hljs-string">&apos;#B6B812&apos;</span>,<span class="hljs-string">&apos;#9DC3C3&apos;</span>, <span class="hljs-string">&apos;#A89C92&apos;</span>, <span class="hljs-string">&apos;#FEE00C&apos;</span>,<span class="hljs-string">&apos;#FEF2A1&apos;</span>]

type_color_B={
    <span class="hljs-string">&apos;na&#xEF;ve B cells&apos;</span>:sc_color[<span class="hljs-number">13</span>],
    <span class="hljs-string">&apos;Follicular B cells&apos;</span>:sc_color[<span class="hljs-number">20</span>],
    <span class="hljs-string">&apos;Pre-B cells&apos;</span>:sc_color[<span class="hljs-number">14</span>],
    <span class="hljs-string">&apos;activated B cells&apos;</span>:sc_color[<span class="hljs-number">15</span>],
    <span class="hljs-string">&apos;IgA+ Plasma B cells&apos;</span>:sc_color[<span class="hljs-number">8</span>],
    <span class="hljs-string">&apos;IgG+ Plasma B cells&apos;</span>:sc_color[<span class="hljs-number">12</span>],
    <span class="hljs-string">&apos;Plasma B cells&apos;</span>:sc_color[<span class="hljs-number">12</span>],
    <span class="hljs-string">&apos;B-cell&apos;</span>:sc_color[<span class="hljs-number">13</span>],
    <span class="hljs-string">&apos;B-cell (MALAT1+)&apos;</span>:<span class="hljs-string">&apos;#a3007d&apos;</span>,

}
</code></pre>
<p>&#x63A5;&#x7740;&#x6211;&#x4EEC;&#x8BFB;&#x53D6;&#x9700;&#x8981;&#x7528;&#x5230;&#x7684;&#x6587;&#x4EF6;</p>
<pre><code class="lang-python">sig_means=pd.read_csv(<span class="hljs-string">&apos;../cellphonedb/Bcell_pos/significant_means.txt&apos;</span>,sep=<span class="hljs-string">&apos;\t&apos;</span>)
sig_means.head()
</code></pre>
<p><img src="book-5.assets/image-20221019014049480-6120013.png" alt="image-20221019014049480" style="zoom:50%;"></p>
<pre><code class="lang-python">de=pd.read_csv(<span class="hljs-string">&apos;../cellphonedb/Bcell_pos/deconvoluted.txt&apos;</span>,sep=<span class="hljs-string">&apos;\t&apos;</span>)
de.columns[:]
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;</p>
<pre><code class="lang-shell">Index([&apos;gene_name&apos;, &apos;uniprot&apos;, &apos;is_complex&apos;, &apos;protein_name&apos;, &apos;complex_name&apos;,
       &apos;id_cp_interaction&apos;, &apos;B-cell (MALAT1+)&apos;, &apos;Plasma B cells&apos;,
       &apos;Pre-B cells&apos;, &apos;activated B cells&apos;, &apos;na&#xEF;ve B cells&apos;],
      dtype=&apos;object&apos;)
</code></pre>
<p>&#x6784;&#x5EFA;&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x7F51;&#x7EDC;</p>
<pre><code class="lang-python">G=nx.Graph()
<span class="hljs-keyword">for</span> name <span class="hljs-keyword">in</span> de.columns[<span class="hljs-number">6</span>:]:
    test=sig_means.loc[:,sig_means.columns.str.contains(<span class="hljs-string">&apos;\|&apos;</span>)].sum()
    <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> test.index:
        <span class="hljs-keyword">if</span> test.loc[i]!=<span class="hljs-number">0</span>:
            a=i.split(<span class="hljs-string">&apos;|&apos;</span>)
            G.add_edge(a[<span class="hljs-number">0</span>],a[<span class="hljs-number">1</span>],weight=test.loc[i])

fig, ax = plt.subplots(figsize=(<span class="hljs-number">3</span>,<span class="hljs-number">3</span>)) 

nodes=[]
G_degree=dict(G.degree(G.nodes()))


G_nodes_dict={}
links = []
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> G.edges:
    <span class="hljs-keyword">if</span> i[<span class="hljs-number">0</span>] <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> G_nodes_dict.keys():
        G_nodes_dict[i[<span class="hljs-number">0</span>]]=<span class="hljs-number">0</span>
    <span class="hljs-keyword">if</span> i[<span class="hljs-number">1</span>] <span class="hljs-keyword">not</span> <span class="hljs-keyword">in</span> G_nodes_dict.keys():
        G_nodes_dict[i[<span class="hljs-number">1</span>]]=<span class="hljs-number">0</span>
    links.append({<span class="hljs-string">&quot;source&quot;</span>: i[<span class="hljs-number">0</span>], <span class="hljs-string">&quot;target&quot;</span>: i[<span class="hljs-number">1</span>]})
    weight=G.get_edge_data(i[<span class="hljs-number">0</span>],i[<span class="hljs-number">1</span>])[<span class="hljs-string">&apos;weight&apos;</span>]
    G_nodes_dict[i[<span class="hljs-number">0</span>]]+=weight
    G_nodes_dict[i[<span class="hljs-number">1</span>]]+=weight


pos = nx.spring_layout(G)
p=dict(G.nodes)
nodesize=[G_nodes_dict[u]*<span class="hljs-number">30</span> <span class="hljs-keyword">for</span> u <span class="hljs-keyword">in</span> G.nodes()]
nodecolos=[type_color_B[u] <span class="hljs-keyword">for</span> u <span class="hljs-keyword">in</span> G.nodes()]
nx.draw_networkx_nodes(G, pos, nodelist=p,node_size=nodesize,node_color=nodecolos)<span class="hljs-comment">#cmap=nodecmap,</span>
                       <span class="hljs-comment">#node_color=list(p.values()))</span>

edgewidth = [G.get_edge_data(u, v)[<span class="hljs-string">&apos;weight&apos;</span>] <span class="hljs-keyword">for</span> u, v <span class="hljs-keyword">in</span> G.edges()]
nx.draw_networkx_edges(G, pos,width=edgewidth)


<span class="hljs-comment">#label_options = {&quot;ec&quot;: &quot;white&quot;, &quot;fc&quot;: &quot;white&quot;, &quot;alpha&quot;: 0.6}</span>
nx.draw_networkx_labels(G, pos, font_size=<span class="hljs-number">10</span>,) <span class="hljs-comment">#bbox=label_options)</span>
plt.grid(<span class="hljs-keyword">False</span>)
plt.axis(<span class="hljs-string">&quot;off&quot;</span>)
plt.xlim(<span class="hljs-number">-2</span>,<span class="hljs-number">2</span>)
plt.ylim(<span class="hljs-number">-2</span>,<span class="hljs-number">1.5</span>)
plt.title(<span class="hljs-string">&apos;Pos Cellphone&apos;</span>,fontsize=<span class="hljs-number">12</span>)
<span class="hljs-comment">#plt.savefig(&quot;cellphonedb/pos_cellphone.png&quot;,dpi=300,bbox_inches = &apos;tight&apos;)</span>
</code></pre>
<p><img src="book-5.assets/image-20221019014313452-6120017.png" alt="image-20221019014313452" style="zoom:50%;"></p>
<h4 id="434-&#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x53EF;&#x89C6;&#x5316;-&#x53D7;&#x914D;&#x4F53;&#x5173;&#x7CFB;">4.3.4 &#x7EC6;&#x80DE;&#x901A;&#x8BAF;&#x53EF;&#x89C6;&#x5316;-&#x53D7;&#x914D;&#x4F53;&#x5173;&#x7CFB;</h4>
<p>&#x9996;&#x5148;&#x4F9D;&#x7136;&#x662F;&#x5BFC;&#x5165;&#x6570;&#x636E;</p>
<pre><code class="lang-python">sig_means1=sig_means.copy()
sig_means1.set_index(sig_means1.columns[<span class="hljs-number">1</span>],inplace=<span class="hljs-keyword">True</span>)

pvalue=pd.read_csv(<span class="hljs-string">&apos;../cellphonedb/Bcell_pos/pvalues.txt&apos;</span>,sep=<span class="hljs-string">&apos;\t&apos;</span>)
pvalue.set_index(pvalue.columns[<span class="hljs-number">1</span>],inplace=<span class="hljs-keyword">True</span>)
p=pvalue.loc[:,pvalue.columns.str.contains(<span class="hljs-string">&apos;&apos;</span>)]
p.head()

exp=sig_means1.loc[:,sig_means1.columns.str.contains(<span class="hljs-string">&apos;\|&apos;</span>)]
exp.head()
</code></pre>
<p>&#x7D27;&#x63A5;&#x7740;&#xFF0C;&#x6211;&#x4EEC;&#x751F;&#x6210;&#x9700;&#x8981;&#x5C55;&#x793A;&#x7684;&#x53D7;&#x914D;&#x4F53;DataFrame</p>
<pre><code class="lang-python">plot_data3=pd.DataFrame(columns=[<span class="hljs-string">&apos;Pair&apos;</span>,<span class="hljs-string">&apos;means&apos;</span>,<span class="hljs-string">&apos;p-value&apos;</span>,<span class="hljs-string">&apos;Type&apos;</span>])
p_pair=exp.columns.tolist()
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> list(set(exp.index.tolist())):
    plot_data4=pd.DataFrame()
    <span class="hljs-keyword">if</span> exp.loc[i,p_pair].fillna(<span class="hljs-number">0</span>).sum()==<span class="hljs-number">0</span>:
        <span class="hljs-keyword">continue</span>
    plot_data4[<span class="hljs-string">&apos;Pair&apos;</span>]=[j <span class="hljs-keyword">for</span> j <span class="hljs-keyword">in</span> [i]*len(p_pair)]
    plot_data4[<span class="hljs-string">&apos;means&apos;</span>]=exp.loc[i,p_pair].values
    plot_data4[<span class="hljs-string">&apos;p-value&apos;</span>]=-np.log10(p.loc[i,p_pair].values.astype(<span class="hljs-string">&apos;float&apos;</span>))+<span class="hljs-number">0.3</span>
    plot_data4[<span class="hljs-string">&apos;Type&apos;</span>]=p_pair

    plot_data3=pd.concat([plot_data3,plot_data4])
plot_data3=plot_data3.fillna(<span class="hljs-number">0</span>)

plot_data3[<span class="hljs-string">&apos;means&apos;</span>]=np.log2(plot_data3[<span class="hljs-string">&apos;means&apos;</span>])
plot_data3=plot_data3.replace(float(<span class="hljs-string">&apos;inf&apos;</span>),<span class="hljs-number">3</span>)
plot_data3=plot_data3.replace(float(<span class="hljs-string">&apos;-inf&apos;</span>),<span class="hljs-number">-3</span>)
plot_data3.head()
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right">Pair</th>
<th style="text-align:right">means</th>
<th style="text-align:right">p-value</th>
<th style="text-align:right">Type</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">0</td>
<td style="text-align:right">CCR10_CCL28</td>
<td style="text-align:right">-3.0</td>
<td style="text-align:right">0.3</td>
<td>B-cell (MALAT1+)\</td>
<td>B-cell (MALAT1+)</td>
</tr>
<tr>
<td style="text-align:right">1</td>
<td style="text-align:right">CCR10_CCL28</td>
<td style="text-align:right">-3.0</td>
<td style="text-align:right">0.3</td>
<td>B-cell (MALAT1+)\</td>
<td>Plasma B cells</td>
</tr>
<tr>
<td style="text-align:right">2</td>
<td style="text-align:right">CCR10_CCL28</td>
<td style="text-align:right">-3.0</td>
<td style="text-align:right">0.3</td>
<td>B-cell (MALAT1+)\</td>
<td>Pre-B cells</td>
</tr>
<tr>
<td style="text-align:right">3</td>
<td style="text-align:right">CCR10_CCL28</td>
<td style="text-align:right">-3.0</td>
<td style="text-align:right">0.3</td>
<td>B-cell (MALAT1+)\</td>
<td>activated B cells</td>
</tr>
<tr>
<td style="text-align:right">4</td>
<td style="text-align:right">CCR10_CCL28</td>
<td style="text-align:right">-3.0</td>
<td style="text-align:right">0.3</td>
<td>B-cell (MALAT1+)\</td>
<td>nai&#x308;ve B cells</td>
</tr>
</tbody>
</table>
<p>&#x57FA;&#x7840;&#x7ED8;&#x56FE;</p>
<pre><code class="lang-python"><span class="hljs-comment">#pip install python_dotplot</span>
<span class="hljs-keyword">import</span> dotplot
<span class="hljs-keyword">import</span> dotplot.utils
new_keys = {<span class="hljs-string">&apos;item_key&apos;</span>: <span class="hljs-string">&apos;Pair&apos;</span>,<span class="hljs-string">&apos;group_key&apos;</span>: <span class="hljs-string">&apos;Type&apos;</span>,<span class="hljs-string">&apos;sizes_key&apos;</span>: <span class="hljs-string">&apos;p-value&apos;</span>,<span class="hljs-string">&apos;color_key&apos;</span>: <span class="hljs-string">&apos;means&apos;</span>}
dp = dotplot.DotPlot.parse_from_tidy_data(plot_data3, **new_keys)
ax = dp.plot(size_factor=<span class="hljs-number">40</span>, cmap=<span class="hljs-string">&apos;Spectral_r&apos;</span>,
             dot_title = <span class="hljs-string">&apos;&#x2013;log10(Pvalue)&apos;</span>, colorbar_title = <span class="hljs-string">&apos;log2(means)&apos;</span>,vmin=<span class="hljs-number">-1</span>,vmax=<span class="hljs-number">1</span>)

xlabs=ax.get_axes()[<span class="hljs-number">0</span>].get_xticklabels()
ax.get_axes()[<span class="hljs-number">0</span>].set_xticklabels(xlabs,fontsize=<span class="hljs-number">12</span>)
ylabs=ax.get_axes()[<span class="hljs-number">0</span>].get_yticklabels()
ax.get_axes()[<span class="hljs-number">0</span>].set_yticklabels(ylabs,fontsize=<span class="hljs-number">12</span>)
</code></pre>
<p><img src="book-5.assets/image-20221019014538577-6120020.png" alt="image-20221019014538577" style="zoom:50%;"></p>
<p>&#x8FD9;&#x56FE;&#x770B;&#x8D77;&#x6765;&#x4E0D;&#x662F;&#x5F88;&#x7F8E;&#x89C2;&#xFF0C;&#x6211;&#x4EEC;&#x7A0D;&#x5FAE;&#x7F8E;&#x5316;&#x4E00;&#x4E0B;&#xFF0C;&#x5C06;&#x751F;&#x6210;&#x4E24;&#x5F20;&#x56FE;&#xFF0C;&#x6700;&#x540E;&#x518D;&#x7528;PS&#x5408;&#x5E76;&#xFF0C;&#x9996;&#x5148;&#x662F;&#x53D7;&#x4F53;&#x56FE;</p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x53BB;&#x9664;&#x914D;&#x4F53;</span>
[i.set_text(i.get_text().split(<span class="hljs-string">&apos;|&apos;</span>)[<span class="hljs-number">1</span>].replace(<span class="hljs-string">&apos;Vil-&apos;</span>,<span class="hljs-string">&apos;&apos;</span>)) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> xlabs]
ax.get_axes()[<span class="hljs-number">0</span>].set_xticklabels(xlabs,fontsize=<span class="hljs-number">12</span>)
<span class="hljs-comment">#&#x7ED8;&#x5236;&#x865A;&#x7EBF;</span>
[i.set_text(i.get_text().split(<span class="hljs-string">&apos;_&apos;</span>)[<span class="hljs-number">1</span>]) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> ylabs]
ax.get_axes()[<span class="hljs-number">0</span>].set_yticklabels(ylabs,fontsize=<span class="hljs-number">12</span>)
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(len(xlabs)):
    <span class="hljs-keyword">if</span> i%<span class="hljs-number">5</span>==<span class="hljs-number">0</span>:
        ax.get_axes()[<span class="hljs-number">0</span>].axvline(x=i+<span class="hljs-number">0.5</span>, ymin=<span class="hljs-number">-1</span>, ymax=<span class="hljs-number">10</span>,color=<span class="hljs-string">&apos;black&apos;</span>,linestyle=<span class="hljs-string">&quot;--&quot;</span>)
<span class="hljs-comment">#&#x4E0A;&#x8272;        </span>
ax.get_axes()[<span class="hljs-number">0</span>].tick_params(axis=<span class="hljs-string">&apos;both&apos;</span>,colors=<span class="hljs-string">&apos;red&apos;</span>, which=<span class="hljs-string">&apos;both&apos;</span>)
ax.savefig(<span class="hljs-string">&quot;cellphonedb/figure_pos_rec.png&quot;</span>,dpi=<span class="hljs-number">300</span>,bbox_inches = <span class="hljs-string">&apos;tight&apos;</span>)
ax
</code></pre>
<p><img src="book-5.assets/image-20221019014718599-6120022.png" alt="image-20221019014718599" style="zoom:50%;"></p>
<p>&#x5176;&#x6B21;&#x662F;&#x914D;&#x4F53;&#x56FE;</p>
<pre><code class="lang-python">new_keys = {<span class="hljs-string">&apos;item_key&apos;</span>: <span class="hljs-string">&apos;Pair&apos;</span>,<span class="hljs-string">&apos;group_key&apos;</span>: <span class="hljs-string">&apos;Type&apos;</span>,<span class="hljs-string">&apos;sizes_key&apos;</span>: <span class="hljs-string">&apos;p-value&apos;</span>,<span class="hljs-string">&apos;color_key&apos;</span>: <span class="hljs-string">&apos;means&apos;</span>}
dp = dotplot.DotPlot.parse_from_tidy_data(plot_data3, **new_keys)
ax = dp.plot(size_factor=<span class="hljs-number">40</span>, cmap=<span class="hljs-string">&apos;Spectral_r&apos;</span>,
             dot_title = <span class="hljs-string">&apos;&#x2013;log10(Pvalue)&apos;</span>, colorbar_title = <span class="hljs-string">&apos;log2(means)&apos;</span>,vmin=<span class="hljs-number">-1</span>,vmax=<span class="hljs-number">1</span>)

xlabs=ax.get_axes()[<span class="hljs-number">0</span>].get_xticklabels()
ax.get_axes()[<span class="hljs-number">0</span>].set_xticklabels(xlabs,fontsize=<span class="hljs-number">12</span>)
ylabs=ax.get_axes()[<span class="hljs-number">0</span>].get_yticklabels()
ax.get_axes()[<span class="hljs-number">0</span>].set_yticklabels(ylabs,fontsize=<span class="hljs-number">12</span>)
<span class="hljs-comment">#&#x53BB;&#x9664;&#x53D7;&#x4F53;</span>
[xlabs[i].set_text(<span class="hljs-string">&apos;------&apos;</span>) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(len(xlabs)) <span class="hljs-keyword">if</span> (i+<span class="hljs-number">3</span>)%<span class="hljs-number">5</span>!=<span class="hljs-number">0</span>]
[xlabs[i].set_text(<span class="hljs-string">&apos;------\n&apos;</span>+xlabs[i].get_text().replace(<span class="hljs-string">&apos;p-value_&apos;</span>,<span class="hljs-string">&apos;&apos;</span>).split(<span class="hljs-string">&apos;|&apos;</span>)[<span class="hljs-number">0</span>]) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(len(xlabs)) <span class="hljs-keyword">if</span> (i+<span class="hljs-number">3</span>)%<span class="hljs-number">5</span>==<span class="hljs-number">0</span>]
ax.get_axes()[<span class="hljs-number">0</span>].set_xticklabels(xlabs,fontsize=<span class="hljs-number">12</span>,rotation=<span class="hljs-number">0</span>)
<span class="hljs-comment">#&#x7ED8;&#x5236;&#x865A;&#x7EBF;</span>
[i.set_text(i.get_text().split(<span class="hljs-string">&apos;_&apos;</span>)[<span class="hljs-number">0</span>]+<span class="hljs-string">&apos;-&apos;</span>+<span class="hljs-string">&apos;  &apos;</span>*len(i.get_text().split(<span class="hljs-string">&apos;_&apos;</span>)[<span class="hljs-number">1</span>])) <span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> ylabs]
ax.get_axes()[<span class="hljs-number">0</span>].set_yticklabels(ylabs,fontsize=<span class="hljs-number">12</span>)

<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(len(xlabs)):
    <span class="hljs-keyword">if</span> i%<span class="hljs-number">5</span>==<span class="hljs-number">0</span>:
        ax.get_axes()[<span class="hljs-number">0</span>].axvline(x=i+<span class="hljs-number">0.5</span>, ymin=<span class="hljs-number">-1</span>, ymax=<span class="hljs-number">10</span>,color=<span class="hljs-string">&apos;black&apos;</span>,linestyle=<span class="hljs-string">&quot;--&quot;</span>)
<span class="hljs-comment">#&#x4E0A;&#x8272;</span>
ax.get_axes()[<span class="hljs-number">0</span>].tick_params(axis=<span class="hljs-string">&apos;both&apos;</span>,colors=<span class="hljs-string">&apos;#3C718D&apos;</span>, which=<span class="hljs-string">&apos;major&apos;</span>)
ax.savefig(<span class="hljs-string">&quot;cellphonedb/figure_pos_li.png&quot;</span>,dpi=<span class="hljs-number">300</span>,bbox_inches = <span class="hljs-string">&apos;tight&apos;</span>)
ax
</code></pre>
<p><img src="book-5.assets/image-20221019014912052-6120025.png" alt="image-20221019014912052" style="zoom:50%;"> </p>
<footer class="page-footer"><span class="copyright">&#x7248;&#x6743;&#x6240;&#x6709; &#xA9; Starlitnightly</span><span class="footer-modification">&#x6700;&#x540E;&#x4FEE;&#x8BA2;&#x65F6;&#x95F4;&#xFF1A;
2022-10-17
</span></footer>
						
					</section>
					
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

				</div>
			</div>
			
		</div>

		
		
		<a href="book-4.html" title="上一页" class="navigation navigation-prev " aria-label="Previous page: 6-4. 差异表达分析">
			<i class="fa fa-angle-left"></i>
		</a>
		
		
		<a href="book-6.html" title="下一页" class="navigation navigation-next " aria-label="Next page: 6-6. 动态调控分析">
			<i class="fa fa-angle-right"></i>
		</a>
		
		
		
	</div>
	<script>
		var gitbook = gitbook || [];
		gitbook.push(function() {
			gitbook.page.hasChanged({"page":{"title":"6-5. 细胞通讯分析","level":"1.9.5","depth":2,"next":{"title":"6-6. 动态调控分析","level":"1.9.6","depth":2,"path":"MultiOMIC/book-6.md","ref":"MultiOMIC/book-6.md","articles":[]},"previous":{"title":"6-4. 差异表达分析","level":"1.9.4","depth":2,"path":"MultiOMIC/book-4.md","ref":"MultiOMIC/book-4.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{"themeIvoryhza":{"nav":[{"url":"http://...","target":"_blank","name":"HOME"}]}},"plugins":["expandable-chapters","-sharing","-fontsettings","back-to-top-button","copy-code-button","cuav-chapters","heading-anchors","theme-ivoryhza"],"pluginsConfig":{"search":{},"theme-ivoryhza":{"search-placeholder":"搜索","logo":"./logo.png","appleTouchIconPrecomposed152":"./logo.png","copyright":"版权所有 © Starlitnightly","modify_label":"最后修订时间：","modify_format":"2022-10-17"},"cuav-chapters":{"summaryMaxSize":20,"useLimitExpanded":false},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"heading-anchors":{},"highlight":{},"back-to-top-button":{},"copy-code-button":{},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false},"expandable-chapters":{}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"MultiOMIC/book-5.md","mtime":"2022-10-18T19:07:07.610Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2022-11-22T15:14:39.839Z"},"basePath":"..","book":{"language":""}});
		});
	</script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-back-to-top-button/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-copy-code-button/toggle.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-cuav-chapters/cuav-chapters.js"></script>
        
    
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/1.2.1/anchor.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-heading-anchors/anchor-style.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-theme-ivoryhza/ivory.js"></script>
        
    

    </body>
</html>

