
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>2-3. 通路富集分析(GSEA) · GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-back-to-top-button/plugin.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-cuav-chapters/cuav-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-theme-ivoryhza/ivory.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-theme-ivoryhza/footer.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="../SCGLUE/" />
    
    
    <link rel="prev" href="rnaseq_2.html" />
    

    </head>
    <body>
        
<div class="book">
	<div class="header-inner">
		<!-- LOGO -->
		<div class="header-inner-left">
			<div class="logo"></div>
			<span class="title"></span>
		</div>

		<!-- Nav -->
		<div class="header-inner-right">
			<ul class="header-nav">
				<li>
					<a href="http://..." target="_blank">HOME</a>
				</li>
			</ul>
			<!-- Search -->
			
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

		</div>


	</div>

	<div class="book-summary">
		<div class="book-summary-title">目录</div>
		
		
		<nav role="navigation">
			


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Bioinformatics Tutorial
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../">
            
                <a href="../">
            
                    
                    目录
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../CONFIG/">
            
                <a href="../CONFIG/">
            
                    
                    配置：基础环境配置
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../CONFIG/config_1.html">
            
                <a href="../CONFIG/config_1.html">
            
                    
                    配置1：Python环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../CONFIG/config_2.html">
            
                <a href="../CONFIG/config_2.html">
            
                    
                    配置2：Linux环境
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../CONFIG/config_3.html">
            
                <a href="../CONFIG/config_3.html">
            
                    
                    配置3：R语言环境
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../RNASEQUP/">
            
                <a href="../RNASEQUP/">
            
                    
                    分析1：RNA-seq上游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../RNASEQUP/rnasequp_1.html">
            
                <a href="../RNASEQUP/rnasequp_1.html">
            
                    
                    1-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../RNASEQUP/rnasequp_2.html">
            
                <a href="../RNASEQUP/rnasequp_2.html">
            
                    
                    1-2. 上游分析全教程
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="./">
            
                <a href="./">
            
                    
                    分析2：RNA-seq下游分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="rnaseq_1.html">
            
                <a href="rnaseq_1.html">
            
                    
                    2-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="rnaseq_2.html">
            
                <a href="rnaseq_2.html">
            
                    
                    2-2. 差异表达基因(DEG)分析
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.5.3" data-path="rnaseq_3.html">
            
                <a href="rnaseq_3.html">
            
                    
                    2-3. 通路富集分析(GSEA)
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../SCGLUE/">
            
                <a href="../SCGLUE/">
            
                    
                    分析3：单细胞样本对齐
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" data-path="../SCGLUE/scglue_1.html">
            
                <a href="../SCGLUE/scglue_1.html">
            
                    
                    3-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.2" data-path="../SCGLUE/scglue_2.html">
            
                <a href="../SCGLUE/scglue_2.html">
            
                    
                    3-2. 数据预处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.3" data-path="../SCGLUE/scglue_3.html">
            
                <a href="../SCGLUE/scglue_3.html">
            
                    
                    3-3. 模型准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.4" data-path="../SCGLUE/scglue_4.html">
            
                <a href="../SCGLUE/scglue_4.html">
            
                    
                    3-4. 模型训练
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.5" data-path="../SCGLUE/scglue_5.html">
            
                <a href="../SCGLUE/scglue_5.html">
            
                    
                    3-5. 对齐细胞
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="../MOFA/">
            
                <a href="../MOFA/">
            
                    
                    分析4：MOFA单细胞多组学因子分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.7.1" data-path="../MOFA/mofa_1.html">
            
                <a href="../MOFA/mofa_1.html">
            
                    
                    4-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.2" data-path="../MOFA/mofa_2.html">
            
                <a href="../MOFA/mofa_2.html">
            
                    
                    4-2. 单细胞样本对齐
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.3" data-path="../MOFA/mofa_3.html">
            
                <a href="../MOFA/mofa_3.html">
            
                    
                    4-3. scRNA-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.4" data-path="../MOFA/mofa_4.html">
            
                <a href="../MOFA/mofa_4.html">
            
                    
                    4-4. scATAC-seq数据处理
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.5" data-path="../MOFA/mofa_5.html">
            
                <a href="../MOFA/mofa_5.html">
            
                    
                    4-5. MOFA模型构建
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.6" data-path="../MOFA/mofa_6.html">
            
                <a href="../MOFA/mofa_6.html">
            
                    
                    4-6. MOFA下游分析-R语言环节
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.7.7" data-path="../MOFA/mofa_7.html">
            
                <a href="../MOFA/mofa_7.html">
            
                    
                    4-7. MOFA下游分析-Python环节
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.8" data-path="../TMT/">
            
                <a href="../TMT/">
            
                    
                    分析5：TMT蛋白组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.8.1" data-path="../TMT/tmt_1.html">
            
                <a href="../TMT/tmt_1.html">
            
                    
                    5-1. 环境配置
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8.2" data-path="../TMT/tmt_2.html">
            
                <a href="../TMT/tmt_2.html">
            
                    
                    5-2. TMT蛋白组学上游分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8.3" data-path="../TMT/tmt_3.html">
            
                <a href="../TMT/tmt_3.html">
            
                    
                    5-3. TMT蛋白组学下游分析
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.9" data-path="../MultiOMIC/">
            
                <a href="../MultiOMIC/">
            
                    
                    分析6：多组学分析
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.9.1" data-path="../MultiOMIC/book-1.html">
            
                <a href="../MultiOMIC/book-1.html">
            
                    
                    6-1. 数据准备
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.2" data-path="../MultiOMIC/book-2.html">
            
                <a href="../MultiOMIC/book-2.html">
            
                    
                    6-2. GLUE多组学整合
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.3" data-path="../MultiOMIC/book-3.html">
            
                <a href="../MultiOMIC/book-3.html">
            
                    
                    6-3. 细胞亚群分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.4" data-path="../MultiOMIC/book-4.html">
            
                <a href="../MultiOMIC/book-4.html">
            
                    
                    6-4. 差异表达分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.5" data-path="../MultiOMIC/book-5.html">
            
                <a href="../MultiOMIC/book-5.html">
            
                    
                    6-5. 细胞通讯分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.6" data-path="../MultiOMIC/book-6.html">
            
                <a href="../MultiOMIC/book-6.html">
            
                    
                    6-6. 动态调控分析
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9.7" data-path="../MultiOMIC/book-7.html">
            
                <a href="../MultiOMIC/book-7.html">
            
                    
                    6-7. 转录因子网络
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.10" data-path="../PLOT/">
            
                <a href="../PLOT/">
            
                    
                    绘图：Python可视化
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.10.1" data-path="../PLOT/plot_1.html">
            
                <a href="../PLOT/plot_1.html">
            
                    
                    1-1. 火山图
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.10.2" data-path="../PLOT/plot_2.html">
            
                <a href="../PLOT/plot_2.html">
            
                    
                    1-2. 箱线图
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


		</nav>
		
		
	</div>

	<!-- Content nav -->
	<div class="book-anchor">
		<div class="book-anchor-title">大纲</div>
		<div class="book-anchor-body">

		</div>
	</div>

	<div class="book-body">
		
		<div class="body-inner">
			
			

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >2-3. 通路富集分析(GSEA)</a>
    </h1>
</div>




			<div class="page-wrapper" tabindex="-1" role="main">
				<div class="page-inner">
					
<div id="book-search-results">
    <div class="search-noresults">
    
					<section class="normal markdown-section">
						
						<h1 id="rna-seq&#x5206;&#x6790;&#xFF1A;-&#x901A;&#x8DEF;&#x5BCC;&#x96C6;&#x5206;&#x6790;gsea">RNA-seq&#x5206;&#x6790;&#xFF1A; &#x901A;&#x8DEF;&#x5BCC;&#x96C6;&#x5206;&#x6790;(GSEA)</h1>
<p>&#x5728;&#x5F97;&#x5230;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x5206;&#x6790;&#x7684;&#x7ED3;&#x679C;&#x540E;&#xFF0C;&#x6211;&#x4EEC;&#x5F80;&#x5F80;&#x8FD8;&#x4F1A;&#x5BF9;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x57FA;&#x56E0;&#x76F8;&#x5173;&#x7684;&#x529F;&#x80FD;&#x611F;&#x5174;&#x8DA3;&#xFF0C;&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x4ECB;&#x7ECD;GSEA&#x65B9;&#x6CD5;&#xFF0C;&#x8FD9;&#x662F;&#x4E00;&#x79CD;&#x6BD4;&#x8F83;&#x597D;&#x7528;&#x7684;&#x901A;&#x8DEF;&#x5BCC;&#x96C6;&#x65B9;&#x6CD5;&#xFF0C;&#x53EF;&#x4EE5;&#x6839;&#x636E;&#x57FA;&#x56E0;&#x8868;&#x8FBE;&#x7684;&#x500D;&#x6570;&#x60C5;&#x51B5;&#xFF0C;&#x6765;&#x5206;&#x6790;&#x57FA;&#x56E0;&#x76F8;&#x5173;&#x7684;&#x901A;&#x8DEF;&#x3002;</p>
<blockquote>
<p>&#x4E00;&#x822C;&#x6765;&#x8BF4;&#xFF0C;&#x505A;GSEA&#x5206;&#x6790;&#x5176;&#x5B9E;&#x662F;&#x53EF;&#x4EE5;&#x4E0D;&#x7528;&#x505A;&#x63D0;&#x53D6;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x57FA;&#x56E0;&#x5206;&#x6790;&#x7684;&#x7ED3;&#x679C;&#x7684;&#xFF0C;&#x76F4;&#x63A5;&#x7528;DESeq2&#x7684;&#x7ED3;&#x679C;&#x5C31;&#x53EF;&#x4EE5;&#x4E86;&#xFF0C;&#x4F46;&#x662F;&#x6211;&#x8FD9;&#x91CC;&#x66F4;&#x5173;&#x6CE8;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x57FA;&#x56E0;&#x76F8;&#x5173;&#x7684;&#x529F;&#x80FD;&#xFF0C;&#x6240;&#x4EE5;&#x5C31;&#x7528;&#x4E86;&#x5DEE;&#x5F02;&#x8868;&#x8FBE;&#x57FA;&#x56E0;&#x6765;&#x505A;GSEA&#x5206;&#x6790;</p>
</blockquote>
<h2 id="1-&#x6570;&#x636E;&#x9884;&#x5904;&#x7406;">1. &#x6570;&#x636E;&#x9884;&#x5904;&#x7406;</h2>
<h3 id="11-&#x52A0;&#x8F7D;&#x5305;">1.1 &#x52A0;&#x8F7D;&#x5305;</h3>
<pre><code class="lang-python"><span class="hljs-keyword">import</span> gseapy <span class="hljs-keyword">as</span> gp
</code></pre>
<h3 id="12-&#x63D0;&#x53D6;&#x6570;&#x636E;">1.2 &#x63D0;&#x53D6;&#x6570;&#x636E;</h3>
<p>&#x7531;&#x4E8E;&#x6211;&#x4EEC;&#x9700;&#x8981;&#x540C;&#x65F6;&#x8003;&#x8651;log2FoldChange&#x4E0E;p-value&#xFF0C;&#x6240;&#x4EE5;&#x6211;&#x4EEC;&#x8FD9;&#x91CC;&#x7528;&#x4E00;&#x4E2A;&#x65B0;&#x7684;&#x516C;&#x5F0F;&#x6765;&#x8BA1;&#x7B97;&#x57FA;&#x56E0;&#x500D;&#x6570;</p>
<p>metric=-sign(log2FoldChange)/-lg10(pvalue)</p>
<pre><code class="lang-python">gseada=data1.loc[data1[<span class="hljs-string">&apos;sig&apos;</span>]!=<span class="hljs-string">&apos;normal&apos;</span>]
<span class="hljs-comment">#&#x500D;&#x6570;&#x53D8;&#x5316;&#x89C4;&#x5219;</span>
gseada[<span class="hljs-string">&apos;fcsign&apos;</span>]=-np.sign(gseada[<span class="hljs-string">&apos;log2FoldChange&apos;</span>])
gseada[<span class="hljs-string">&apos;logp&apos;</span>]=-np.log10(gseada[<span class="hljs-string">&apos;pvalue&apos;</span>])
gseada[<span class="hljs-string">&apos;metric&apos;</span>]=gseada[<span class="hljs-string">&apos;logp&apos;</span>]/gseada[<span class="hljs-string">&apos;fcsign&apos;</span>]
gseada.head()
</code></pre>
<p>&#x5F97;&#x5230;&#x7ED3;&#x679C;&#x5982;&#x4E0B;</p>
<table>
<thead>
<tr>
<th style="text-align:right"></th>
<th style="text-align:right">baseMean</th>
<th style="text-align:right">log2FoldChange</th>
<th style="text-align:right">lfcSE</th>
<th style="text-align:right">stat</th>
<th style="text-align:right">pvalue</th>
<th style="text-align:right">padj</th>
<th style="text-align:right">sig</th>
<th style="text-align:right">log(padj)</th>
<th style="text-align:right">fcsign</th>
<th style="text-align:right">logp</th>
<th style="text-align:right">metric</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">ID3</td>
<td style="text-align:right">5442.355360</td>
<td style="text-align:right">-5.710849</td>
<td style="text-align:right">0.145082</td>
<td style="text-align:right">-39.362892</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">down</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">1.0</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">inf</td>
</tr>
<tr>
<td style="text-align:right">TXNIP</td>
<td style="text-align:right">3147.115496</td>
<td style="text-align:right">4.041551</td>
<td style="text-align:right">0.103068</td>
<td style="text-align:right">39.212444</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">up</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">-1.0</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">-inf</td>
</tr>
<tr>
<td style="text-align:right">MMP1</td>
<td style="text-align:right">4344.077177</td>
<td style="text-align:right">4.476829</td>
<td style="text-align:right">0.108971</td>
<td style="text-align:right">41.082873</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">up</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">-1.0</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">-inf</td>
</tr>
<tr>
<td style="text-align:right">ID1</td>
<td style="text-align:right">5526.691818</td>
<td style="text-align:right">-5.280627</td>
<td style="text-align:right">0.112286</td>
<td style="text-align:right">-47.028177</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">down</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">1.0</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">inf</td>
</tr>
<tr>
<td style="text-align:right">PTGIS</td>
<td style="text-align:right">4187.607799</td>
<td style="text-align:right">3.441545</td>
<td style="text-align:right">0.092530</td>
<td style="text-align:right">37.193624</td>
<td style="text-align:right">8.651991e-303</td>
<td style="text-align:right">3.367009e-299</td>
<td style="text-align:right">up</td>
<td style="text-align:right">298.472756</td>
<td style="text-align:right">-1.0</td>
<td style="text-align:right">302.062884</td>
<td style="text-align:right">-302.062884</td>
</tr>
</tbody>
</table>
<h3 id="13-&#x6570;&#x636E;&#x6392;&#x5E8F;">1.3 &#x6570;&#x636E;&#x6392;&#x5E8F;</h3>
<pre><code class="lang-python">gseada=gseada.sort_values(by=[<span class="hljs-string">&apos;metric&apos;</span>],ascending=<span class="hljs-keyword">False</span>)
gseada.head()
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right"></th>
<th style="text-align:right">baseMean</th>
<th style="text-align:right">log2FoldChange</th>
<th style="text-align:right">lfcSE</th>
<th style="text-align:right">stat</th>
<th style="text-align:right">pvalue</th>
<th style="text-align:right">padj</th>
<th style="text-align:right">sig</th>
<th style="text-align:right">log(padj)</th>
<th style="text-align:right">fcsign</th>
<th style="text-align:right">logp</th>
<th style="text-align:right">metric</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">ID3</td>
<td style="text-align:right">5442.355360</td>
<td style="text-align:right">-5.710849</td>
<td style="text-align:right">0.145082</td>
<td style="text-align:right">-39.362892</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">down</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">1.0</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">inf</td>
</tr>
<tr>
<td style="text-align:right">ID1</td>
<td style="text-align:right">5526.691818</td>
<td style="text-align:right">-5.280627</td>
<td style="text-align:right">0.112286</td>
<td style="text-align:right">-47.028177</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">0.000000e+00</td>
<td style="text-align:right">down</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">1.0</td>
<td style="text-align:right">inf</td>
<td style="text-align:right">inf</td>
</tr>
<tr>
<td style="text-align:right">OXTR</td>
<td style="text-align:right">4105.199915</td>
<td style="text-align:right">-2.959533</td>
<td style="text-align:right">0.100562</td>
<td style="text-align:right">-29.429803</td>
<td style="text-align:right">2.283138e-190</td>
<td style="text-align:right">2.468072e-187</td>
<td style="text-align:right">down</td>
<td style="text-align:right">186.607642</td>
<td style="text-align:right">1.0</td>
<td style="text-align:right">189.641468</td>
<td style="text-align:right">189.641468</td>
</tr>
<tr>
<td style="text-align:right">SAMD11</td>
<td style="text-align:right">1574.582498</td>
<td style="text-align:right">-6.502040</td>
<td style="text-align:right">0.235542</td>
<td style="text-align:right">-27.604533</td>
<td style="text-align:right">9.816869e-168</td>
<td style="text-align:right">8.682574e-165</td>
<td style="text-align:right">down</td>
<td style="text-align:right">164.061352</td>
<td style="text-align:right">1.0</td>
<td style="text-align:right">167.008027</td>
<td style="text-align:right">167.008027</td>
</tr>
</tbody>
</table>
<h3 id="14-rnk&#x77E9;&#x9635;&#x63D0;&#x53D6;">1.4 rnk&#x77E9;&#x9635;&#x63D0;&#x53D6;</h3>
<p>&#x6211;&#x4EEC;&#x53D1;&#x73B0;&#x6570;&#x636E;&#x4E2D;&#x6709;inf&#xFF0C;&#x8FD9;&#x5BF9;&#x540E;&#x7EED;&#x7684;&#x5206;&#x6790;&#x6709;&#x4E00;&#x5B9A;&#x7684;&#x5F71;&#x54CD;&#xFF0C;&#x6240;&#x4EE5;&#x6211;&#x4EEC;&#x5BF9;inf&#x8FDB;&#x884C;&#x8D4B;&#x503C;&#x3002;</p>
<pre><code class="lang-python">rnk=pd.DataFrame()
rnk[<span class="hljs-string">&apos;gene_name&apos;</span>]=gseada.index
rnk[<span class="hljs-string">&apos;rnk&apos;</span>]=gseada[<span class="hljs-string">&apos;metric&apos;</span>].values
k=<span class="hljs-number">1</span>
total=<span class="hljs-number">0</span>
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(len(rnk)):
    <span class="hljs-keyword">if</span> rnk.loc[i,<span class="hljs-string">&apos;rnk&apos;</span>]==np.inf: 
        total+=<span class="hljs-number">1</span>
<span class="hljs-comment">#200&#x8DDF;274&#x6839;&#x636E;&#x4F60;&#x7684;&#x6570;&#x636E;&#x8FDB;&#x884C;&#x66F4;&#x6539;&#xFF0C;&#x4FDD;&#x8BC1;inf&#x6BD4;&#x4F60;&#x6570;&#x636E;&#x6700;&#x5927;&#x7684;&#x5927;&#xFF0C;-inf&#x6BD4;&#x6570;&#x636E;&#x6700;&#x5C0F;&#x7684;&#x5C0F;&#x5C31;&#x597D;</span>
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(len(rnk)):
    <span class="hljs-keyword">if</span> rnk.loc[i,<span class="hljs-string">&apos;rnk&apos;</span>]==np.inf: 
        rnk.loc[i,<span class="hljs-string">&apos;rnk&apos;</span>]=<span class="hljs-number">200</span>+(total-k)
        k+=<span class="hljs-number">1</span>
    <span class="hljs-keyword">elif</span> rnk.loc[i,<span class="hljs-string">&apos;rnk&apos;</span>]==-np.inf: 
        rnk.loc[i,<span class="hljs-string">&apos;rnk&apos;</span>]=-(<span class="hljs-number">274</span>+k)
        k+=<span class="hljs-number">1</span>
<span class="hljs-comment">#rnk=rnk.replace(np.inf,300)</span>
<span class="hljs-comment">#rnk=rnk.replace(-np.inf,-300)</span>
rnk.head()
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right"></th>
<th style="text-align:right">gene_name</th>
<th style="text-align:right">rnk</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">0</td>
<td style="text-align:right">ID3</td>
<td style="text-align:right">201.000000</td>
</tr>
<tr>
<td style="text-align:right">1</td>
<td style="text-align:right">ID1</td>
<td style="text-align:right">200.000000</td>
</tr>
<tr>
<td style="text-align:right">2</td>
<td style="text-align:right">OXTR</td>
<td style="text-align:right">189.641468</td>
</tr>
<tr>
<td style="text-align:right">3</td>
<td style="text-align:right">SAMD11</td>
<td style="text-align:right">167.008027</td>
</tr>
<tr>
<td style="text-align:right">4</td>
<td style="text-align:right">ATOH8</td>
<td style="text-align:right">146.165895</td>
</tr>
</tbody>
</table>
<h2 id="2-gsea&#x901A;&#x8DEF;&#x5BCC;&#x96C6;">2. GSEA&#x901A;&#x8DEF;&#x5BCC;&#x96C6;</h2>
<h3 id="21-&#x57FA;&#x56E0;index&#x8F6C;&#x5927;&#x5199;">2.1 &#x57FA;&#x56E0;index&#x8F6C;&#x5927;&#x5199;</h3>
<p>&#x7531;&#x4E8E;gseapy&#x7684;&#x5305;&#x53EA;&#x80FD;&#x8BC6;&#x522B;&#x5927;&#x5199;&#x7684;&#x57FA;&#x56E0;&#x540D;&#xFF0C;&#x6240;&#x4EE5;&#x6211;&#x4EEC;&#x9700;&#x8981;&#x5148;&#x628A;&#x6240;&#x6709;&#x7684;&#x57FA;&#x56E0;&#x540D;&#x8F6C;&#x6362;&#x6210;&#x5927;&#x5199;&#x7684;&#x3002;</p>
<pre><code class="lang-python"><span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> range(len(rnk)):
    rnk.loc[i,<span class="hljs-string">&apos;gene_name&apos;</span>]=rnk.loc[i,<span class="hljs-string">&apos;gene_name&apos;</span>].upper()
rnk.head()
</code></pre>
<h3 id="22-gsea&#x901A;&#x8DEF;&#x5BCC;&#x96C6;">2.2 GSEA&#x901A;&#x8DEF;&#x5BCC;&#x96C6;</h3>
<p>&#x8FD9;&#x91CC;&#x7B80;&#x5355;&#x4ECB;&#x7ECD;&#x4EE5;&#x4E0B;&#x51FD;&#x6570;&#x7684;&#x53C2;&#x6570;</p>
<p>rnk&#xFF1A;&#x8F93;&#x5165;&#x6392;&#x5E8F;&#x77E9;&#x9635;</p>
<p>gene_sets&#xFF1A;&#x9700;&#x8981;&#x5BCC;&#x96C6;&#x5230;&#x7684;&#x901A;&#x8DEF;&#x6570;&#x636E;&#x96C6;</p>
<p>processes&#xFF1A;&#x5E76;&#x884C;&#x4F7F;&#x7528;&#x7684;&#x7EBF;&#x7A0B;&#x6570;</p>
<p>permutation_num&#xFF1A;&#x68C0;&#x9A8C;&#x7684;&#x901F;&#x5EA6;</p>
<p>outdir&#xFF1A;&#x7ED3;&#x679C;&#x8F93;&#x51FA;&#x76EE;&#x5F55;</p>
<pre><code class="lang-python"><span class="hljs-comment">#&#x6211;&#x4EEC;&#x53EF;&#x4EE5;&#x901A;&#x8FC7;&#x4EE5;&#x4E0B;&#x51FD;&#x6570;&#x89C2;&#x5BDF;&#x90FD;&#x6709;&#x54EA;&#x4E9B;&#x6570;&#x636E;&#x96C6;&#x53EF;&#x88AB;&#x7528;&#x6765;&#x5BCC;&#x96C6;</span>
names = gp.get_library_name() <span class="hljs-comment"># default: Human</span>
names[:<span class="hljs-number">10</span>]
</code></pre>
<p>&#x5728;&#x8FD9;&#x91CC;&#xFF0C;&#x6211;&#x4EEC;&#x9009;&#x62E9;&#x7684;&#x662F;KEGG_2019_Human&#x6570;&#x636E;&#x96C6;&#xFF0C;&#x56E0;&#x4E3A;NEGF&#x662F;&#x4EBA;&#x6E90;&#x7684;&#x7EC6;&#x80DE;&#x7CFB;</p>
<pre><code class="lang-python">pre_res = gp.prerank(rnk=rnk, gene_sets=<span class="hljs-string">&apos;KEGG_2019_Mouse&apos;</span>,
                     processes=<span class="hljs-number">4</span>,
                     permutation_num=<span class="hljs-number">100</span>, <span class="hljs-comment"># reduce number to speed up testing</span>
                     outdir=<span class="hljs-string">&apos;NEGF_gsea/prerank_report_kegg&apos;</span>, format=<span class="hljs-string">&apos;png&apos;</span>, seed=<span class="hljs-number">6</span>)
</code></pre>
<p>&#x8F93;&#x51FA;&#x7ED3;&#x679C;&#x88AB;&#x4FDD;&#x5B58;&#x5230;&#x4E86;NEGF_gsea&#x76EE;&#x5F55;&#x4E0B;</p>
<p>&#x6211;&#x4EEC;&#x53EF;&#x4EE5;&#x901A;&#x8FC7;index&#x67E5;&#x770B;</p>
<pre><code class="lang-python">pre_res.res2d.sort_index()
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right"></th>
<th style="text-align:right">es</th>
<th style="text-align:right">nes</th>
<th style="text-align:right">pval</th>
<th style="text-align:right">fdr</th>
<th style="text-align:right">geneset_size</th>
<th style="text-align:right">matched_size</th>
<th style="text-align:right">genes</th>
<th style="text-align:right">ledge_genes</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right">Term</td>
<td style="text-align:right"></td>
<td style="text-align:right"></td>
<td style="text-align:right"></td>
<td style="text-align:right"></td>
<td style="text-align:right"></td>
<td style="text-align:right"></td>
<td style="text-align:right"></td>
<td style="text-align:right"></td>
</tr>
<tr>
<td style="text-align:right">Chemokine signaling pathway</td>
<td style="text-align:right">-0.767073</td>
<td style="text-align:right">-1.330912</td>
<td style="text-align:right">0.045977</td>
<td style="text-align:right">0.091794</td>
<td style="text-align:right">190</td>
<td style="text-align:right">16</td>
<td style="text-align:right">CXCL11;CCL11;CX3CL1;HCK;CXCL10;CCL20;CCL8;VAV3...</td>
<td style="text-align:right">CXCL3;CXCL5;CXCL2;CXCL8;CXCL1;CXCL6</td>
</tr>
<tr>
<td style="text-align:right">Cytokine-cytokine receptor interaction</td>
<td style="text-align:right">-0.702018</td>
<td style="text-align:right">-1.351795</td>
<td style="text-align:right">0.055556</td>
<td style="text-align:right">0.094755</td>
<td style="text-align:right">294</td>
<td style="text-align:right">29</td>
<td style="text-align:right">IL18;IL1RN;CXCL11;CCL11;CX3CL1;CXCL10;IL18R1;C...</td>
<td style="text-align:right">LIF;CXCL3;CXCL5;CXCL2;ACKR4;IL6;IL1B;IL32;CXCL...</td>
</tr>
<tr>
<td style="text-align:right">IL-17 signaling pathway</td>
<td style="text-align:right">-0.801394</td>
<td style="text-align:right">-1.427337</td>
<td style="text-align:right">0.000000</td>
<td style="text-align:right">0.029611</td>
<td style="text-align:right">93</td>
<td style="text-align:right">18</td>
<td style="text-align:right">CCL11;CXCL10;CCL20;MMP9;CSF2;CSF3;CCL7;CXCL3;C...</td>
<td style="text-align:right">CXCL3;CXCL5;CXCL2;IL6;IL1B;PTGS2;TNFAIP3;CXCL8...</td>
</tr>
<tr>
<td style="text-align:right">NOD-like receptor signaling pathway</td>
<td style="text-align:right">-0.840998</td>
<td style="text-align:right">-1.518649</td>
<td style="text-align:right">0.000000</td>
<td style="text-align:right">0.000000</td>
<td style="text-align:right">178</td>
<td style="text-align:right">15</td>
<td style="text-align:right">IL18;IRF7;NOD2;GBP4;CXCL3;CXCL2;IL6;IL1B;BIRC3...</td>
<td style="text-align:right">CXCL3;CXCL2;IL6;IL1B;BIRC3;OAS1;TNFAIP3;OAS2;C...</td>
</tr>
</tbody>
</table>
<p>Term&#x5C31;&#x4EE3;&#x8868;&#x4E86;&#x5BCC;&#x96C6;&#x5230;&#x7684;&#x901A;&#x8DEF;</p>
<h3 id="23-&#x5355;&#x901A;&#x8DEF;&#x53EF;&#x89C6;&#x5316;">2.3 &#x5355;&#x901A;&#x8DEF;&#x53EF;&#x89C6;&#x5316;</h3>
<pre><code class="lang-python"><span class="hljs-keyword">from</span> gseapy.plot <span class="hljs-keyword">import</span> gseaplot

<span class="hljs-comment"># to save your figure, make sure that ofname is not None</span>
gseaplot(rank_metric=pre_res.ranking, term=terms[<span class="hljs-number">0</span>], **pre_res.results[terms[<span class="hljs-number">0</span>]])
</code></pre>
<p>&#x6211;&#x4EEC;&#x60F3;&#x89C2;&#x5BDF;&#x6BCF;&#x4E00;&#x4E2A;&#x901A;&#x8DEF;&#x5177;&#x4F53;&#x7684;&#x5BCC;&#x96C6;&#x60C5;&#x51B5;&#xFF0C;&#x53EF;&#x4EE5;&#x901A;&#x8FC7;gseaplot&#x7ED8;&#x5236;&#x76F8;&#x5173;&#x51FD;&#x6570;</p>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/NOD-like%20receptor%20signaling%20pathway.prerank.png" alt="NOD-like receptor signaling pathway.prerank" style="zoom:33%;"></p>
<h3 id="24-&#x591A;&#x901A;&#x8DEF;&#x53EF;&#x89C6;&#x5316;">2.4 &#x591A;&#x901A;&#x8DEF;&#x53EF;&#x89C6;&#x5316;</h3>
<p>&#x6211;&#x4EEC;&#x8FD8;&#x60F3;&#x89C2;&#x5BDF;&#x5BCC;&#x96C6;&#x5230;&#x7684;&#x6240;&#x6709;&#x901A;&#x8DEF;&#x7ED3;&#x679C;&#xFF0C;&#x8FD9;&#x91CC;&#x63D0;&#x4F9B;&#x4E86;&#x4E00;&#x79CD;&#x6C14;&#x6CE1;&#x56FE;&#x7684;&#x65B9;&#x5F0F;&#x8FDB;&#x884C;&#x89C2;&#x5BDF;</p>
<pre><code class="lang-python">sc_data=pd.read_csv(<span class="hljs-string">r&apos;NEGF_gsea\prerank_report_kegg\gseapy.prerank.gene_sets.report.csv&apos;</span>)
sc_data=sc_data.loc[sc_data[<span class="hljs-string">&apos;pval&apos;</span>]&lt;<span class="hljs-number">0.1</span>]
sc_data[<span class="hljs-string">&apos;logp&apos;</span>]=-np.log10(sc_data[<span class="hljs-string">&apos;pval&apos;</span>]+<span class="hljs-number">0.001</span>)
sc_data[<span class="hljs-string">&apos;com&apos;</span>]=sc_data[<span class="hljs-string">&apos;matched_size&apos;</span>]/sc_data[<span class="hljs-string">&apos;geneset_size&apos;</span>]
sc_data.head()

pp=plt.figure(figsize=(<span class="hljs-number">3</span>,<span class="hljs-number">10</span>))
a=pp.add_subplot(<span class="hljs-number">1</span>,<span class="hljs-number">1</span>,<span class="hljs-number">1</span>)

plt.scatter(np.zeros(len(sc_data))+<span class="hljs-number">1</span>,sc_data[<span class="hljs-string">&apos;Term&apos;</span>],s=sc_data[<span class="hljs-string">&apos;com&apos;</span>]*<span class="hljs-number">500</span>,alpha=<span class="hljs-number">0.6</span>,linewidth=<span class="hljs-number">2</span>,
            c=sc_data[<span class="hljs-string">&apos;logp&apos;</span>],
            cmap=<span class="hljs-string">&apos;Greens&apos;</span>)
plt.xticks([<span class="hljs-number">1</span>],[<span class="hljs-string">&apos;NHDF&apos;</span>])
plt.xticks(range(<span class="hljs-number">-1</span>,<span class="hljs-number">3</span>,<span class="hljs-number">1</span>))
<span class="hljs-comment">#plt.colorbar(a)</span>
plt.yticks(fontsize=<span class="hljs-number">12</span>)
plt.xticks(fontsize=<span class="hljs-number">12</span>)
plt.xlim(<span class="hljs-number">-0.5</span>,<span class="hljs-number">1.5</span>)
plt.savefig(<span class="hljs-string">&apos;NHDF_pathway.png&apos;</span>,dpi=<span class="hljs-number">300</span>,bbox_inches = <span class="hljs-string">&apos;tight&apos;</span>)
</code></pre>
<p><img src="https://raw.githubusercontent.com/Starlitnightly/bioinformatic_galaxy/master/img/untitled1.png" alt="untitled1"> </p>
<footer class="page-footer"><span class="copyright">&#x7248;&#x6743;&#x6240;&#x6709; &#xA9; Starlitnightly</span><span class="footer-modification">&#x6700;&#x540E;&#x4FEE;&#x8BA2;&#x65F6;&#x95F4;&#xFF1A;
2022-10-17
</span></footer>
						
					</section>
					
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

				</div>
			</div>
			
		</div>

		
		
		<a href="rnaseq_2.html" title="上一页" class="navigation navigation-prev " aria-label="Previous page: 2-2. 差异表达基因(DEG)分析">
			<i class="fa fa-angle-left"></i>
		</a>
		
		
		<a href="../SCGLUE/" title="下一页" class="navigation navigation-next " aria-label="Next page: 分析3：单细胞样本对齐">
			<i class="fa fa-angle-right"></i>
		</a>
		
		
		
	</div>
	<script>
		var gitbook = gitbook || [];
		gitbook.push(function() {
			gitbook.page.hasChanged({"page":{"title":"2-3. 通路富集分析(GSEA)","level":"1.5.3","depth":2,"next":{"title":"分析3：单细胞样本对齐","level":"1.6","depth":1,"path":"SCGLUE/README.md","ref":"SCGLUE/README.md","articles":[{"title":"3-1. 环境配置","level":"1.6.1","depth":2,"path":"SCGLUE/scglue_1.md","ref":"SCGLUE/scglue_1.md","articles":[]},{"title":"3-2. 数据预处理","level":"1.6.2","depth":2,"path":"SCGLUE/scglue_2.md","ref":"SCGLUE/scglue_2.md","articles":[]},{"title":"3-3. 模型准备","level":"1.6.3","depth":2,"path":"SCGLUE/scglue_3.md","ref":"SCGLUE/scglue_3.md","articles":[]},{"title":"3-4. 模型训练","level":"1.6.4","depth":2,"path":"SCGLUE/scglue_4.md","ref":"SCGLUE/scglue_4.md","articles":[]},{"title":"3-5. 对齐细胞","level":"1.6.5","depth":2,"path":"SCGLUE/scglue_5.md","ref":"SCGLUE/scglue_5.md","articles":[]}]},"previous":{"title":"2-2. 差异表达基因(DEG)分析","level":"1.5.2","depth":2,"path":"RNASEQ/rnaseq_2.md","ref":"RNASEQ/rnaseq_2.md","articles":[]},"dir":"ltr"},"config":{"gitbook":"*","theme":"default","variables":{"themeIvoryhza":{"nav":[{"url":"http://...","target":"_blank","name":"HOME"}]}},"plugins":["expandable-chapters","-sharing","-fontsettings","back-to-top-button","copy-code-button","cuav-chapters","heading-anchors","theme-ivoryhza"],"pluginsConfig":{"search":{},"theme-ivoryhza":{"search-placeholder":"搜索","logo":"./logo.png","appleTouchIconPrecomposed152":"./logo.png","copyright":"版权所有 © Starlitnightly","modify_label":"最后修订时间：","modify_format":"2022-10-17"},"cuav-chapters":{"summaryMaxSize":20,"useLimitExpanded":false},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"heading-anchors":{},"highlight":{},"back-to-top-button":{},"copy-code-button":{},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false},"expandable-chapters":{}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"}},"file":{"path":"RNASEQ/rnaseq_3.md","mtime":"2022-10-18T18:22:29.177Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2022-11-22T15:14:39.839Z"},"basePath":"..","book":{"language":""}});
		});
	</script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-back-to-top-button/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-copy-code-button/toggle.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-cuav-chapters/cuav-chapters.js"></script>
        
    
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/1.2.1/anchor.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-heading-anchors/anchor-style.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-theme-ivoryhza/ivory.js"></script>
        
    

    </body>
</html>

